{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 149,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-05-11T05:59:54.713726Z",
     "start_time": "2024-05-11T05:59:54.533203600Z"
    }
   },
   "outputs": [],
   "source": [
    "from multiprocessing import Pool\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import allel\n",
    "import pysam\n",
    "import json\n",
    "import os\n",
    "from datetime import datetime\n",
    "pd.set_option('display.max_columns', None)\n",
    "from urllib import request\n",
    "from tqdm import tqdm\n",
    "from autocnv import settings\n",
    "from gtfparse import read_gtf\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 150,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-05-11T05:59:55.023899400Z",
     "start_time": "2024-05-11T05:59:54.545171300Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "raw_data_dir: /staging/software/lib/python3.9/site-packages/autocnv-1-py3.9.egg/raw_data\n",
      "result_data_dir: /staging/software/lib/python3.9/site-packages/autocnv-1-py3.9.egg/data\n",
      "version_date_dir: /staging/software/lib/python3.9/site-packages/autocnv-1-py3.9.egg/raw_data/2021-10-28\n"
     ]
    }
   ],
   "source": [
    "work_dir = settings.BASE_DIR\n",
    "raw_data_dir = os.path.join(work_dir, 'raw_data')\n",
    "result_data_dir = os.path.join(work_dir, 'data')\n",
    "if not os.path.exists(raw_data_dir):\n",
    "    os.mkdir(raw_data_dir)\n",
    "if not os.path.exists(result_data_dir):\n",
    "    os.mkdir(result_data_dir)\n",
    "\n",
    "# version_date = datetime.today().strftime('%Y-%m-%d')\n",
    "version_date = '2021-10-28'\n",
    "version_date_dir = os.path.join(raw_data_dir, version_date)\n",
    "if not os.path.exists(version_date_dir):\n",
    "    os.mkdir(version_date_dir)\n",
    "print(\n",
    "        f\"raw_data_dir: {raw_data_dir}\",\n",
    "        f\"result_data_dir: {result_data_dir}\",\n",
    "        f\"version_date_dir: {version_date_dir}\",\n",
    "        sep=\"\\n\"\n",
    "      )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 151,
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    },
    "ExecuteTime": {
     "end_time": "2024-05-11T05:59:55.023899400Z",
     "start_time": "2024-05-11T05:59:54.557139400Z"
    }
   },
   "outputs": [],
   "source": [
    "exon_url = 'https://hgdownload.soe.ucsc.edu/goldenPath/hg19/bigZips/genes/hg19.refGene.gtf.gz'\n",
    "cyto_band_url = 'https://hgdownload.soe.ucsc.edu/goldenPath/hg19/database/cytoBand.txt.gz'\n",
    "gene_info_url = 'https://ftp.ncbi.nih.gov/refseq/H_sapiens/Homo_sapiens.gene_info.gz'\n",
    "ref_gene_url = 'https://hgdownload.soe.ucsc.edu/goldenPath/hg19/database/refGene.txt.gz'\n",
    "clingen_gene_curation_url = 'https://ftp.clinicalgenome.org/ClinGen_gene_curation_list_GRCh37.tsv'\n",
    "clingene_region_curation_url = 'https://ftp.clinicalgenome.org/ClinGen_region_curation_list_GRCh37.tsv'\n",
    "clinvar_url = 'https://ftp.ncbi.nlm.nih.gov/pub/clinvar/vcf_GRCh37/clinvar.vcf.gz'\n",
    "hi_pred_url = 'https://www.deciphergenomics.org/files/downloads/HI_Predictions_Version3.bed.gz'\n",
    "gnomad_lof_url = 'https://datasetgnomad.blob.core.windows.net/dataset/release/2.1.1/constraint/gnomad.v2.1.1.lof_metrics.by_gene.txt.bgz'\n",
    "gnomad_control_only_url = 'https://datasetgnomad.blob.core.windows.net/dataset/papers/2019-sv/gnomad_v2.1_sv.controls_only.sites.bed.gz'\n",
    "hgnc_gene_fam_url = 'ftp://ftp.ebi.ac.uk/pub/databases/genenames/hgnc/csv/genefamily_db_tables/family.csv'\n",
    "# ori file\n",
    "a=[exon_url, cyto_band_url, gene_info_url, ref_gene_url, clingen_gene_curation_url, clingene_region_curation_url, clinvar_url, hi_pred_url, gnomad_lof_url, gnomad_control_only_url, hgnc_gene_fam_url]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 152,
   "outputs": [],
   "source": [
    "exon_ori_file = os.path.join(version_date_dir, 'hg19.refGene.gtf.gz ')\n",
    "cyto_band_ori_file = os.path.join(version_date_dir, 'cytoBand.txt.gz')\n",
    "gene_info_ori_file = os.path.join(version_date_dir, 'Homo_sapiens.gene_info.gz')\n",
    "ref_gene_ori_file = os.path.join(version_date_dir, 'refGene.txt.gz')\n",
    "clingen_gene_ori_file = os.path.join(version_date_dir, 'clingen_gene_hg19.tsv')\n",
    "clingen_region_ori_file = os.path.join(version_date_dir, 'clingen_region_hg19.tsv')\n",
    "clinvar_ori_vcf_file = os.path.join(version_date_dir, 'clinvar.vcf.gz')\n",
    "hi_pred_ori_file = os.path.join(version_date_dir, 'HI_Predictions_Version3.bed.gz')\n",
    "gnomad_lof_ori_file = os.path.join(version_date_dir, 'gnomad.v2.1.1.lof_metrics.by_gene.txt.bgz')\n",
    "gnomad_control_ori_file = os.path.join(version_date_dir, 'gnomad_v2.1_sv.controls_only.sites.bed.gz')\n",
    "hgnc_gene_fam_file = os.path.join(version_date_dir, 'family.csv')\n",
    "# result file\n",
    "b = [exon_ori_file, cyto_band_ori_file, gene_info_ori_file, ref_gene_ori_file, clingen_gene_ori_file, clingen_region_ori_file, clinvar_ori_vcf_file, hi_pred_ori_file, gnomad_lof_ori_file, gnomad_control_ori_file, hgnc_gene_fam_file]"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2024-05-11T05:59:55.107668500Z",
     "start_time": "2024-05-11T05:59:54.568139400Z"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 153,
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "--2024-05-11 13:59:54--  https://hgdownload.soe.ucsc.edu/goldenPath/hg19/bigZips/genes/hg19.refGene.gtf.gz\n",
      "Connecting to 192.168.134.12:1080... connected.\n",
      "Proxy request sent, awaiting response... 200 OK\n",
      "Length: 21846291 (21M) [application/x-gzip]\n",
      "Saving to: ‘/staging/software/lib/python3.9/site-packages/autocnv-1-py3.9.egg/raw_data/2021-10-28/hg19.refGene.gtf.gz ’\n",
      "\n",
      "     0K .......... .......... .......... .......... ..........  0% 27.2K 13m2s\n",
      "    50K .......... .......... .......... .......... ..........  0% 37.1K 11m16s\n",
      "   100K .......... .......... .......... .......... ..........  0% 32.5K 11m7s\n",
      "   150K .......... .......... .......... .......... ..........  0% 45.0K 10m17s\n",
      "   200K .......... .......... .......... .......... ..........  1% 78.4K 9m6s\n",
      "   250K .......... .......... .......... .......... ..........  1% 76.7K 8m20s\n",
      "   300K .......... .......... .......... .......... ..........  1% 50.5K 8m6s\n",
      "   350K .......... .......... .......... .......... ..........  1% 52.5K 7m55s\n",
      "   400K .......... .......... .......... .......... ..........  2% 95.2K 7m25s\n",
      "   450K .......... .......... .......... .......... ..........  2%  106K 6m59s\n",
      "   500K .......... .......... .......... .......... ..........  2% 89.0K 6m42s\n",
      "   550K .......... .......... .......... .......... ..........  2%  108K 6m23s\n",
      "   600K .......... .......... .......... .......... ..........  3%  125K 6m6s\n",
      "   650K .......... .......... .......... .......... ..........  3% 98.1K 5m54s\n",
      "   700K .......... .......... .......... .......... ..........  3% 99.3K 5m43s\n",
      "   750K .......... .......... .......... .......... ..........  3%  237K 5m26s\n",
      "   800K .......... .......... .......... .......... ..........  3% 57.1K 5m27s\n",
      "   850K .......... .......... .......... .......... ..........  4%  702K 5m10s\n",
      "   900K .......... .......... .......... .......... ..........  4%  135K 5m1s\n",
      "   950K .......... .......... .......... .......... ..........  4% 82.2K 4m58s\n",
      "  1000K .......... .......... .......... .......... ..........  4% 1.03M 4m44s\n",
      "  1050K .......... .......... .......... .......... ..........  5%  124K 4m38s\n",
      "  1100K .......... .......... .......... .......... ..........  5%  171K 4m30s\n",
      "  1150K .......... .......... .......... .......... ..........  5%  228K 4m22s\n",
      "  1200K .......... .......... .......... .......... ..........  5%  126K 4m17s\n",
      "  1250K .......... .......... .......... .......... ..........  6%  208K 4m10s\n",
      "  1300K .......... .......... .......... .......... ..........  6%  206K 4m4s\n",
      "  1350K .......... .......... .......... .......... ..........  6% 98.9K 4m2s\n",
      "  1400K .......... .......... .......... .......... ..........  6% 1.17M 3m53s\n",
      "  1450K .......... .......... .......... .......... ..........  7%  284K 3m47s\n",
      "  1500K .......... .......... .......... .......... ..........  7%  349K 3m41s\n",
      "  1550K .......... .......... .......... .......... ..........  7%  125K 3m39s\n",
      "  1600K .......... .......... .......... .......... ..........  7%  274K 3m34s\n",
      "  1650K .......... .......... .......... .......... ..........  7%  114K 3m32s\n",
      "  1700K .......... .......... .......... .......... ..........  8%  939K 3m26s\n",
      "  1750K .......... .......... .......... .......... ..........  8%  125K 3m24s\n",
      "  1800K .......... .......... .......... .......... ..........  8%  137K 3m22s\n",
      "  1850K .......... .......... .......... .......... ..........  8% 1010K 3m17s\n",
      "  1900K .......... .......... .......... .......... ..........  9%  348K 3m13s\n",
      "  1950K .......... .......... .......... .......... ..........  9%  454K 3m8s\n",
      "  2000K .......... .......... .......... .......... ..........  9%  299K 3m5s\n",
      "  2050K .......... .......... .......... .......... ..........  9% 84.6K 3m5s\n",
      "  2100K .......... .......... .......... .......... .......... 10% 1.14M 3m1s\n",
      "  2150K .......... .......... .......... .......... .......... 10%  195K 2m59s\n",
      "  2200K .......... .......... .......... .......... .......... 10% 1.39M 2m55s\n",
      "  2250K .......... .......... .......... .......... .......... 10%  288K 2m52s\n",
      "  2300K .......... .......... .......... .......... .......... 11%  255K 2m49s\n",
      "  2350K .......... .......... .......... .......... .......... 11%  111K 2m49s\n",
      "  2400K .......... .......... .......... .......... .......... 11% 7.65M 2m45s\n",
      "  2450K .......... .......... .......... .......... .......... 11%  148K 2m44s\n",
      "  2500K .......... .......... .......... .......... .......... 11%  161K 2m42s\n",
      "  2550K .......... .......... .......... .......... .......... 12% 1.77M 2m39s\n",
      "  2600K .......... .......... .......... .......... .......... 12%  359K 2m37s\n",
      "  2650K .......... .......... .......... .......... .......... 12%  428K 2m34s\n",
      "  2700K .......... .......... .......... .......... .......... 12%  454K 2m32s\n",
      "  2750K .......... .......... .......... .......... .......... 13%  222K 2m30s\n",
      "  2800K .......... .......... .......... .......... .......... 13%  170K 2m29s\n",
      "  2850K .......... .......... .......... .......... .......... 13%  411K 2m27s\n",
      "  2900K .......... .......... .......... .......... .......... 13%  474K 2m25s\n",
      "  2950K .......... .......... .......... .......... .......... 14%  504K 2m22s\n",
      "  3000K .......... .......... .......... .......... .......... 14%  186K 2m21s\n",
      "  3050K .......... .......... .......... .......... .......... 14%  176K 2m20s\n",
      "  3100K .......... .......... .......... .......... .......... 14% 1.56M 2m18s\n",
      "  3150K .......... .......... .......... .......... .......... 14%  301K 2m16s\n",
      "  3200K .......... .......... .......... .......... .......... 15% 1.55M 2m14s\n",
      "  3250K .......... .......... .......... .......... .......... 15% 87.8K 2m15s\n",
      "  3300K .......... .......... .......... .......... .......... 15% 1.79M 2m12s\n",
      "  3350K .......... .......... .......... .......... .......... 15% 2.18M 2m10s\n",
      "  3400K .......... .......... .......... .......... .......... 16%  563K 2m8s\n",
      "  3450K .......... .......... .......... .......... .......... 16% 2.20M 2m6s\n",
      "  3500K .......... .......... .......... .......... .......... 16% 2.62M 2m4s\n",
      "  3550K .......... .......... .......... .......... .......... 16% 2.44M 2m2s\n",
      "  3600K .......... .......... .......... .......... .......... 17% 2.14M 2m0s\n",
      "  3650K .......... .......... .......... .......... .......... 17%  766K 1m59s\n",
      "  3700K .......... .......... .......... .......... .......... 17%  158K 1m58s\n",
      "  3750K .......... .......... .......... .......... .......... 17% 2.87M 1m57s\n",
      "  3800K .......... .......... .......... .......... .......... 18%  303K 1m55s\n",
      "  3850K .......... .......... .......... .......... .......... 18% 1.40M 1m54s\n",
      "  3900K .......... .......... .......... .......... .......... 18%  352K 1m53s\n",
      "  3950K .......... .......... .......... .......... .......... 18%  584K 1m51s\n",
      "  4000K .......... .......... .......... .......... .......... 18% 3.07M 1m50s\n",
      "  4050K .......... .......... .......... .......... .......... 19%  289K 1m49s\n",
      "  4100K .......... .......... .......... .......... .......... 19% 2.13M 1m47s\n",
      "  4150K .......... .......... .......... .......... .......... 19% 3.16M 1m46s\n",
      "  4200K .......... .......... .......... .......... .......... 19%  501K 1m45s\n",
      "  4250K .......... .......... .......... .......... .......... 20%  208K 1m44s\n",
      "  4300K .......... .......... .......... .......... .......... 20%  395K 1m43s\n",
      "  4350K .......... .......... .......... .......... .......... 20% 2.71M 1m42s\n",
      "  4400K .......... .......... .......... .......... .......... 20% 1.35M 1m40s\n",
      "  4450K .......... .......... .......... .......... .......... 21%  833K 99s\n",
      "  4500K .......... .......... .......... .......... .......... 21%  746K 98s\n",
      "  4550K .......... .......... .......... .......... .......... 21%  492K 97s\n",
      "  4600K .......... .......... .......... .......... .......... 21%  157K 97s\n",
      "  4650K .......... .......... .......... .......... .......... 22% 2.79M 96s\n",
      "  4700K .......... .......... .......... .......... .......... 22% 2.28M 94s\n",
      "  4750K .......... .......... .......... .......... .......... 22% 88.8K 95s\n",
      "  4800K .......... .......... .......... .......... .......... 22% 2.07M 94s\n",
      "  4850K .......... .......... .......... .......... .......... 22% 3.24M 93s\n",
      "  4900K .......... .......... .......... .......... .......... 23% 2.63M 92s\n",
      "  4950K .......... .......... .......... .......... .......... 23% 4.23M 90s\n",
      "  5000K .......... .......... .......... .......... .......... 23% 3.84M 89s\n",
      "  5050K .......... .......... .......... .......... .......... 23% 1.48M 88s\n",
      "  5100K .......... .......... .......... .......... .......... 24%  868K 87s\n",
      "  5150K .......... .......... .......... .......... .......... 24%  554K 86s\n",
      "  5200K .......... .......... .......... .......... .......... 24%  115K 87s\n",
      "  5250K .......... .......... .......... .......... .......... 24% 3.43M 86s\n",
      "  5300K .......... .......... .......... .......... .......... 25% 2.91M 85s\n",
      "  5350K .......... .......... .......... .......... .......... 25%  896K 84s\n",
      "  5400K .......... .......... .......... .......... .......... 25%  658K 83s\n",
      "  5450K .......... .......... .......... .......... .......... 25%  788K 82s\n",
      "  5500K .......... .......... .......... .......... .......... 26%  169K 82s\n",
      "  5550K .......... .......... .......... .......... .......... 26% 3.42M 81s\n",
      "  5600K .......... .......... .......... .......... .......... 26% 2.86M 80s\n",
      "  5650K .......... .......... .......... .......... .......... 26% 3.49M 79s\n",
      "  5700K .......... .......... .......... .......... .......... 26%  341K 79s\n",
      "  5750K .......... .......... .......... .......... .......... 27%  578K 78s\n",
      "  5800K .......... .......... .......... .......... .......... 27% 1.57M 77s\n",
      "  5850K .......... .......... .......... .......... .......... 27%  698K 76s\n",
      "  5900K .......... .......... .......... .......... .......... 27% 3.42M 76s\n",
      "  5950K .......... .......... .......... .......... .......... 28% 1.28M 75s\n",
      "  6000K .......... .......... .......... .......... .......... 28% 2.03M 74s\n",
      "  6050K .......... .......... .......... .......... .......... 28%  159K 74s\n",
      "  6100K .......... .......... .......... .......... .......... 28% 2.03M 73s\n",
      "  6150K .......... .......... .......... .......... .......... 29%  409K 73s\n",
      "  6200K .......... .......... .......... .......... .......... 29%  192K 72s\n",
      "  6250K .......... .......... .......... .......... .......... 29% 3.55M 72s\n",
      "  6300K .......... .......... .......... .......... .......... 29% 3.47M 71s\n",
      "  6350K .......... .......... .......... .......... .......... 29% 2.87M 70s\n",
      "  6400K .......... .......... .......... .......... .......... 30% 1022K 69s\n",
      "  6450K .......... .......... .......... .......... .......... 30% 3.74M 69s\n",
      "  6500K .......... .......... .......... .......... .......... 30% 3.39M 68s\n",
      "  6550K .......... .......... .......... .......... .......... 30%  659K 67s\n",
      "  6600K .......... .......... .......... .......... .......... 31%  209K 67s\n",
      "  6650K .......... .......... .......... .......... .......... 31%  768K 67s\n",
      "  6700K .......... .......... .......... .......... .......... 31%  874K 66s\n",
      "  6750K .......... .......... .......... .......... .......... 31% 4.30M 65s\n",
      "  6800K .......... .......... .......... .......... .......... 32%  297K 65s\n",
      "  6850K .......... .......... .......... .......... .......... 32% 1.19M 64s\n",
      "  6900K .......... .......... .......... .......... .......... 32%  532K 64s\n",
      "  6950K .......... .......... .......... .......... .......... 32% 3.03M 63s\n",
      "  7000K .......... .......... .......... .......... .......... 33% 3.39M 63s\n",
      "  7050K .......... .......... .......... .......... .......... 33%  243K 62s\n",
      "  7100K .......... .......... .......... .......... .......... 33% 3.05M 62s\n",
      "  7150K .......... .......... .......... .......... .......... 33%  454K 61s\n",
      "  7200K .......... .......... .......... .......... .......... 33% 2.90M 61s\n",
      "  7250K .......... .......... .......... .......... .......... 34%  319K 60s\n",
      "  7300K .......... .......... .......... .......... .......... 34% 1.20M 60s\n",
      "  7350K .......... .......... .......... .......... .......... 34% 18.5M 59s\n",
      "  7400K .......... .......... .......... .......... .......... 34% 1.58M 59s\n",
      "  7450K .......... .......... .......... .......... .......... 35% 4.43M 58s\n",
      "  7500K .......... .......... .......... .......... .......... 35%  195K 58s\n",
      "  7550K .......... .......... .......... .......... .......... 35% 3.68M 57s\n",
      "  7600K .......... .......... .......... .......... .......... 35%  558K 57s\n",
      "  7650K .......... .......... .......... .......... .......... 36% 3.35M 56s\n",
      "  7700K .......... .......... .......... .......... .......... 36%  371K 56s\n",
      "  7750K .......... .......... .......... .......... .......... 36% 3.53M 56s\n",
      "  7800K .......... .......... .......... .......... .......... 36% 3.48M 55s\n",
      "  7850K .......... .......... .......... .......... .......... 37%  250K 55s\n",
      "  7900K .......... .......... .......... .......... .......... 37% 4.31M 54s\n",
      "  7950K .......... .......... .......... .......... .......... 37%  435K 54s\n",
      "  8000K .......... .......... .......... .......... .......... 37%  895K 53s\n",
      "  8050K .......... .......... .......... .......... .......... 37%  810K 53s\n",
      "  8100K .......... .......... .......... .......... .......... 38% 4.99M 53s\n",
      "  8150K .......... .......... .......... .......... .......... 38%  386K 52s\n",
      "  8200K .......... .......... .......... .......... .......... 38% 1.45M 52s\n",
      "  8250K .......... .......... .......... .......... .......... 38% 1015K 51s\n",
      "  8300K .......... .......... .......... .......... .......... 39% 4.35M 51s\n",
      "  8350K .......... .......... .......... .......... .......... 39% 1.18M 50s\n",
      "  8400K .......... .......... .......... .......... .......... 39%  289K 50s\n",
      "  8450K .......... .......... .......... .......... .......... 39%  831K 50s\n",
      "  8500K .......... .......... .......... .......... .......... 40% 5.79M 49s\n",
      "  8550K .......... .......... .......... .......... .......... 40% 4.29M 49s\n",
      "  8600K .......... .......... .......... .......... .......... 40% 3.83M 48s\n",
      "  8650K .......... .......... .......... .......... .......... 40%  724K 48s\n",
      "  8700K .......... .......... .......... .......... .......... 41% 1.22M 48s\n",
      "  8750K .......... .......... .......... .......... .......... 41%  633K 47s\n",
      "  8800K .......... .......... .......... .......... .......... 41%  329K 47s\n",
      "  8850K .......... .......... .......... .......... .......... 41%  355K 47s\n",
      "  8900K .......... .......... .......... .......... .......... 41% 1.32M 46s\n",
      "  8950K .......... .......... .......... .......... .......... 42%  302K 46s\n",
      "  9000K .......... .......... .......... .......... .......... 42%  862K 46s\n",
      "  9050K .......... .......... .......... .......... .......... 42%  117K 46s\n",
      "  9100K .......... .......... .......... .......... .......... 42%  203K 46s\n",
      "  9150K .......... .......... .......... .......... .......... 43% 1.03M 45s\n",
      "  9200K .......... .......... .......... .......... .......... 43%  742K 45s\n",
      "  9250K .......... .......... .......... .......... .......... 43% 1.33M 45s\n",
      "  9300K .......... .......... .......... .......... .......... 43% 3.58M 44s\n",
      "  9350K .......... .......... .......... .......... .......... 44% 1.20M 44s\n",
      "  9400K .......... .......... .......... .......... .......... 44% 3.03M 44s\n",
      "  9450K .......... .......... .......... .......... .......... 44% 2.53M 43s\n",
      "  9500K .......... .......... .......... .......... .......... 44%  180K 43s\n",
      "  9550K .......... .......... .......... .......... .......... 44% 3.44M 43s\n",
      "  9600K .......... .......... .......... .......... .......... 45% 2.94M 42s\n",
      "  9650K .......... .......... .......... .......... .......... 45%  119K 42s\n",
      "  9700K .......... .......... .......... .......... .......... 45% 3.01M 42s\n",
      "  9750K .......... .......... .......... .......... .......... 45% 3.36M 42s\n",
      "  9800K .......... .......... .......... .......... .......... 46% 3.54M 41s\n",
      "  9850K .......... .......... .......... .......... .......... 46% 3.35M 41s\n",
      "  9900K .......... .......... .......... .......... .......... 46% 4.43M 41s\n",
      "  9950K .......... .......... .......... .......... .......... 46% 1.49M 40s\n",
      " 10000K .......... .......... .......... .......... .......... 47% 2.46M 40s\n",
      " 10050K .......... .......... .......... .......... .......... 47% 19.7M 40s\n",
      " 10100K .......... .......... .......... .......... .......... 47%  977K 39s\n",
      " 10150K .......... .......... .......... .......... .......... 47% 4.48M 39s\n",
      " 10200K .......... .......... .......... .......... .......... 48% 4.15M 38s\n",
      " 10250K .......... .......... .......... .......... .......... 48% 5.28M 38s\n",
      " 10300K .......... .......... .......... .......... .......... 48% 2.43M 38s\n",
      " 10350K .......... .......... .......... .......... .......... 48%  820K 38s\n",
      " 10400K .......... .......... .......... .......... .......... 48%  211K 37s\n",
      " 10450K .......... .......... .......... .......... .......... 49%  825K 37s\n",
      " 10500K .......... .......... .......... .......... .......... 49% 5.01M 37s\n",
      " 10550K .......... .......... .......... .......... .......... 49% 5.84M 36s\n",
      " 10600K .......... .......... .......... .......... .......... 49% 4.82M 36s\n",
      " 10650K .......... .......... .......... .......... .......... 50%  238K 36s\n",
      " 10700K .......... .......... .......... .......... .......... 50%  578K 36s\n",
      " 10750K .......... .......... .......... .......... .......... 50% 5.19M 35s\n",
      " 10800K .......... .......... .......... .......... .......... 50% 5.76M 35s\n",
      " 10850K .......... .......... .......... .......... .......... 51%  325K 35s\n",
      " 10900K .......... .......... .......... .......... .......... 51% 4.54M 35s\n",
      " 10950K .......... .......... .......... .......... .......... 51%  509K 34s\n",
      " 11000K .......... .......... .......... .......... .......... 51% 2.14M 34s\n",
      " 11050K .......... .......... .......... .......... .......... 52%  589K 34s\n",
      " 11100K .......... .......... .......... .......... .......... 52%  105K 34s\n",
      " 11150K .......... .......... .......... .......... .......... 52% 1.12M 34s\n",
      " 11200K .......... .......... .......... .......... .......... 52% 5.24M 33s\n",
      " 11250K .......... .......... .......... .......... .......... 52% 5.12M 33s\n",
      " 11300K .......... .......... .......... .......... .......... 53% 5.67M 33s\n",
      " 11350K .......... .......... .......... .......... .......... 53%  234K 33s\n",
      " 11400K .......... .......... .......... .......... .......... 53% 3.45M 32s\n",
      " 11450K .......... .......... .......... .......... .......... 53% 5.41M 32s\n",
      " 11500K .......... .......... .......... .......... .......... 54% 24.1M 32s\n",
      " 11550K .......... .......... .......... .......... .......... 54% 34.4M 31s\n",
      " 11600K .......... .......... .......... .......... .......... 54%  401K 31s\n",
      " 11650K .......... .......... .......... .......... .......... 54% 30.1M 31s\n",
      " 11700K .......... .......... .......... .......... .......... 55% 15.6M 31s\n",
      " 11750K .......... .......... .......... .......... .......... 55% 5.38M 30s\n",
      " 11800K .......... .......... .......... .......... .......... 55% 4.44M 30s\n",
      " 11850K .......... .......... .......... .......... .......... 55% 4.91M 30s\n",
      " 11900K .......... .......... .......... .......... .......... 56% 4.52M 30s\n",
      " 11950K .......... .......... .......... .......... .......... 56%  321K 29s\n",
      " 12000K .......... .......... .......... .......... .......... 56% 5.71M 29s\n",
      " 12050K .......... .......... .......... .......... .......... 56%  533K 29s\n",
      " 12100K .......... .......... .......... .......... .......... 56%  646K 29s\n",
      " 12150K .......... .......... .......... .......... .......... 57% 4.98M 28s\n",
      " 12200K .......... .......... .......... .......... .......... 57% 5.08M 28s\n",
      " 12250K .......... .......... .......... .......... .......... 57%  440K 28s\n",
      " 12300K .......... .......... .......... .......... .......... 57% 4.22M 28s\n",
      " 12350K .......... .......... .......... .......... .......... 58%  408K 28s\n",
      " 12400K .......... .......... .......... .......... .......... 58% 1.22M 27s\n",
      " 12450K .......... .......... .......... .......... .......... 58% 6.49M 27s\n",
      " 12500K .......... .......... .......... .......... .......... 58%  513K 27s\n",
      " 12550K .......... .......... .......... .......... .......... 59% 4.67M 27s\n",
      " 12600K .......... .......... .......... .......... .......... 59%  832K 26s\n",
      " 12650K .......... .......... .......... .......... .......... 59% 1.45M 26s\n",
      " 12700K .......... .......... .......... .......... .......... 59%  465K 26s\n",
      " 12750K .......... .......... .......... .......... .......... 59% 1.29M 26s\n",
      " 12800K .......... .......... .......... .......... .......... 60% 1003K 26s\n",
      " 12850K .......... .......... .......... .......... .......... 60% 2.00M 25s\n",
      " 12900K .......... .......... .......... .......... .......... 60%  864K 25s\n",
      " 12950K .......... .......... .......... .......... .......... 60% 5.55M 25s\n",
      " 13000K .......... .......... .......... .......... .......... 61%  482K 25s\n",
      " 13050K .......... .......... .......... .......... .......... 61%  685K 24s\n",
      " 13100K .......... .......... .......... .......... .......... 61% 1.46M 24s\n",
      " 13150K .......... .......... .......... .......... .......... 61% 3.02M 24s\n",
      " 13200K .......... .......... .......... .......... .......... 62% 6.49M 24s\n",
      " 13250K .......... .......... .......... .......... .......... 62% 2.72M 24s\n",
      " 13300K .......... .......... .......... .......... .......... 62%  315K 23s\n",
      " 13350K .......... .......... .......... .......... .......... 62% 3.71M 23s\n",
      " 13400K .......... .......... .......... .......... .......... 63% 5.08M 23s\n",
      " 13450K .......... .......... .......... .......... .......... 63%  391K 23s\n",
      " 13500K .......... .......... .......... .......... .......... 63%  963K 23s\n",
      " 13550K .......... .......... .......... .......... .......... 63% 4.20M 22s\n",
      " 13600K .......... .......... .......... .......... .......... 63% 5.53M 22s\n",
      " 13650K .......... .......... .......... .......... .......... 64%  376K 22s\n",
      " 13700K .......... .......... .......... .......... .......... 64% 30.4M 22s\n",
      " 13750K .......... .......... .......... .......... .......... 64%  847K 22s\n",
      " 13800K .......... .......... .......... .......... .......... 64% 5.08M 21s\n",
      " 13850K .......... .......... .......... .......... .......... 65%  429K 21s\n",
      " 13900K .......... .......... .......... .......... .......... 65% 2.27M 21s\n",
      " 13950K .......... .......... .......... .......... .......... 65% 1.23M 21s\n",
      " 14000K .......... .......... .......... .......... .......... 65% 1.59M 21s\n",
      " 14050K .......... .......... .......... .......... .......... 66% 1.22M 20s\n",
      " 14100K .......... .......... .......... .......... .......... 66%  887K 20s\n",
      " 14150K .......... .......... .......... .......... .......... 66%  802K 20s\n",
      " 14200K .......... .......... .......... .......... .......... 66% 3.86M 20s\n",
      " 14250K .......... .......... .......... .......... .......... 67%  595K 20s\n",
      " 14300K .......... .......... .......... .......... .......... 67% 4.69M 20s\n",
      " 14350K .......... .......... .......... .......... .......... 67% 5.14M 19s\n",
      " 14400K .......... .......... .......... .......... .......... 67% 4.56M 19s\n",
      " 14450K .......... .......... .......... .......... .......... 67%  271K 19s\n",
      " 14500K .......... .......... .......... .......... .......... 68% 5.28M 19s\n",
      " 14550K .......... .......... .......... .......... .......... 68%  603K 19s\n",
      " 14600K .......... .......... .......... .......... .......... 68% 3.86M 18s\n",
      " 14650K .......... .......... .......... .......... .......... 68% 5.22M 18s\n",
      " 14700K .......... .......... .......... .......... .......... 69%  513K 18s\n",
      " 14750K .......... .......... .......... .......... .......... 69% 5.12M 18s\n",
      " 14800K .......... .......... .......... .......... .......... 69% 5.30M 18s\n",
      " 14850K .......... .......... .......... .......... .......... 69% 4.38M 18s\n",
      " 14900K .......... .......... .......... .......... .......... 70%  282K 17s\n",
      " 14950K .......... .......... .......... .......... .......... 70%  199K 17s\n",
      " 15000K .......... .......... .......... .......... .......... 70% 1.09M 17s\n",
      " 15050K .......... .......... .......... .......... .......... 70% 2.29M 17s\n",
      " 15100K .......... .......... .......... .......... .......... 71% 5.10M 17s\n",
      " 15150K .......... .......... .......... .......... .......... 71% 7.21M 17s\n",
      " 15200K .......... .......... .......... .......... .......... 71% 5.23M 16s\n",
      " 15250K .......... .......... .......... .......... .......... 71% 6.84M 16s\n",
      " 15300K .......... .......... .......... .......... .......... 71% 5.77M 16s\n",
      " 15350K .......... .......... .......... .......... .......... 72% 5.17M 16s\n",
      " 15400K .......... .......... .......... .......... .......... 72% 5.89M 16s\n",
      " 15450K .......... .......... .......... .......... .......... 72%  445K 16s\n",
      " 15500K .......... .......... .......... .......... .......... 72%  799K 15s\n",
      " 15550K .......... .......... .......... .......... .......... 73% 2.44M 15s\n",
      " 15600K .......... .......... .......... .......... .......... 73% 5.63M 15s\n",
      " 15650K .......... .......... .......... .......... .......... 73%  183K 15s\n",
      " 15700K .......... .......... .......... .......... .......... 73%  997K 15s\n",
      " 15750K .......... .......... .......... .......... .......... 74% 32.3M 15s\n",
      " 15800K .......... .......... .......... .......... .......... 74% 33.8M 14s\n",
      " 15850K .......... .......... .......... .......... .......... 74% 32.6M 14s\n",
      " 15900K .......... .......... .......... .......... .......... 74% 30.0M 14s\n",
      " 15950K .......... .......... .......... .......... .......... 74% 24.5M 14s\n",
      " 16000K .......... .......... .......... .......... .......... 75% 26.2M 14s\n",
      " 16050K .......... .......... .......... .......... .......... 75%  570K 14s\n",
      " 16100K .......... .......... .......... .......... .......... 75%  837K 13s\n",
      " 16150K .......... .......... .......... .......... .......... 75%  230K 13s\n",
      " 16200K .......... .......... .......... .......... .......... 76% 4.20M 13s\n",
      " 16250K .......... .......... .......... .......... .......... 76% 9.27M 13s\n",
      " 16300K .......... .......... .......... .......... .......... 76% 8.72M 13s\n",
      " 16350K .......... .......... .......... .......... .......... 76% 5.14M 13s\n",
      " 16400K .......... .......... .......... .......... .......... 77% 6.03M 12s\n",
      " 16450K .......... .......... .......... .......... .......... 77% 5.65M 12s\n",
      " 16500K .......... .......... .......... .......... .......... 77% 8.30M 12s\n",
      " 16550K .......... .......... .......... .......... .......... 77% 7.82M 12s\n",
      " 16600K .......... .......... .......... .......... .......... 78%  394K 12s\n",
      " 16650K .......... .......... .......... .......... .......... 78%  754K 12s\n",
      " 16700K .......... .......... .......... .......... .......... 78% 1.81M 12s\n",
      " 16750K .......... .......... .......... .......... .......... 78%  215K 11s\n",
      " 16800K .......... .......... .......... .......... .......... 78% 7.30M 11s\n",
      " 16850K .......... .......... .......... .......... .......... 79% 5.76M 11s\n",
      " 16900K .......... .......... .......... .......... .......... 79% 6.70M 11s\n",
      " 16950K .......... .......... .......... .......... .......... 79% 5.93M 11s\n",
      " 17000K .......... .......... .......... .......... .......... 79% 6.28M 11s\n",
      " 17050K .......... .......... .......... .......... .......... 80% 7.85M 11s\n",
      " 17100K .......... .......... .......... .......... .......... 80% 1.68M 10s\n",
      " 17150K .......... .......... .......... .......... .......... 80% 1.62M 10s\n",
      " 17200K .......... .......... .......... .......... .......... 80% 6.32M 10s\n",
      " 17250K .......... .......... .......... .......... .......... 81%  487K 10s\n",
      " 17300K .......... .......... .......... .......... .......... 81% 6.84M 10s\n",
      " 17350K .......... .......... .......... .......... .......... 81%  768K 10s\n",
      " 17400K .......... .......... .......... .......... .......... 81% 1.14M 10s\n",
      " 17450K .......... .......... .......... .......... .......... 82% 1.29M 9s\n",
      " 17500K .......... .......... .......... .......... .......... 82% 6.86M 9s\n",
      " 17550K .......... .......... .......... .......... .......... 82% 8.47M 9s\n",
      " 17600K .......... .......... .......... .......... .......... 82% 1.53M 9s\n",
      " 17650K .......... .......... .......... .......... .......... 82% 3.82M 9s\n",
      " 17700K .......... .......... .......... .......... .......... 83%  292K 9s\n",
      " 17750K .......... .......... .......... .......... .......... 83% 22.1M 9s\n",
      " 17800K .......... .......... .......... .......... .......... 83% 37.3M 8s\n",
      " 17850K .......... .......... .......... .......... .......... 83%  194K 8s\n",
      " 17900K .......... .......... .......... .......... .......... 84% 8.83M 8s\n",
      " 17950K .......... .......... .......... .......... .......... 84% 3.19M 8s\n",
      " 18000K .......... .......... .......... .......... .......... 84%  497K 8s\n",
      " 18050K .......... .......... .......... .......... .......... 84% 6.39M 8s\n",
      " 18100K .......... .......... .......... .......... .......... 85% 4.96M 8s\n",
      " 18150K .......... .......... .......... .......... .......... 85%  669K 8s\n",
      " 18200K .......... .......... .......... .......... .......... 85% 5.44M 7s\n",
      " 18250K .......... .......... .......... .......... .......... 85% 6.52M 7s\n",
      " 18300K .......... .......... .......... .......... .......... 86% 8.40M 7s\n",
      " 18350K .......... .......... .......... .......... .......... 86% 1.02M 7s\n",
      " 18400K .......... .......... .......... .......... .......... 86% 18.0M 7s\n",
      " 18450K .......... .......... .......... .......... .......... 86% 26.0M 7s\n",
      " 18500K .......... .......... .......... .......... .......... 86% 35.7M 7s\n",
      " 18550K .......... .......... .......... .......... .......... 87% 30.7M 6s\n",
      " 18600K .......... .......... .......... .......... .......... 87% 1021K 6s\n",
      " 18650K .......... .......... .......... .......... .......... 87% 1.33M 6s\n",
      " 18700K .......... .......... .......... .......... .......... 87% 2.72M 6s\n",
      " 18750K .......... .......... .......... .......... .......... 88% 25.8M 6s\n",
      " 18800K .......... .......... .......... .......... .......... 88%  267K 6s\n",
      " 18850K .......... .......... .......... .......... .......... 88% 1.13M 6s\n",
      " 18900K .......... .......... .......... .......... .......... 88% 7.91M 6s\n",
      " 18950K .......... .......... .......... .......... .......... 89% 7.02M 5s\n",
      " 19000K .......... .......... .......... .......... .......... 89% 9.12M 5s\n",
      " 19050K .......... .......... .......... .......... .......... 89% 8.40M 5s\n",
      " 19100K .......... .......... .......... .......... .......... 89%  419K 5s\n",
      " 19150K .......... .......... .......... .......... .......... 89% 1.45M 5s\n",
      " 19200K .......... .......... .......... .......... .......... 90% 4.62M 5s\n",
      " 19250K .......... .......... .......... .......... .......... 90%  850K 5s\n",
      " 19300K .......... .......... .......... .......... .......... 90%  243K 5s\n",
      " 19350K .......... .......... .......... .......... .......... 90% 1.22M 4s\n",
      " 19400K .......... .......... .......... .......... .......... 91% 2.65M 4s\n",
      " 19450K .......... .......... .......... .......... .......... 91% 1.18M 4s\n",
      " 19500K .......... .......... .......... .......... .......... 91% 1.64M 4s\n",
      " 19550K .......... .......... .......... .......... .......... 91% 1.63M 4s\n",
      " 19600K .......... .......... .......... .......... .......... 92%  995K 4s\n",
      " 19650K .......... .......... .......... .......... .......... 92% 3.55M 4s\n",
      " 19700K .......... .......... .......... .......... .......... 92%  120K 4s\n",
      " 19750K .......... .......... .......... .......... .......... 92% 7.48M 3s\n",
      " 19800K .......... .......... .......... .......... .......... 93% 7.70M 3s\n",
      " 19850K .......... .......... .......... .......... .......... 93%  565K 3s\n",
      " 19900K .......... .......... .......... .......... .......... 93% 6.77M 3s\n",
      " 19950K .......... .......... .......... .......... .......... 93% 5.92M 3s\n",
      " 20000K .......... .......... .......... .......... .......... 93% 7.19M 3s\n",
      " 20050K .......... .......... .......... .......... .......... 94% 7.42M 3s\n",
      " 20100K .......... .......... .......... .......... .......... 94% 1.04M 3s\n",
      " 20150K .......... .......... .......... .......... .......... 94% 37.4M 3s\n",
      " 20200K .......... .......... .......... .......... .......... 94% 32.3M 2s\n",
      " 20250K .......... .......... .......... .......... .......... 95% 39.7M 2s\n",
      " 20300K .......... .......... .......... .......... .......... 95% 6.51M 2s\n",
      " 20350K .......... .......... .......... .......... .......... 95% 2.55M 2s\n",
      " 20400K .......... .......... .......... .......... .......... 95% 33.5M 2s\n",
      " 20450K .......... .......... .......... .......... .......... 96%  508K 2s\n",
      " 20500K .......... .......... .......... .......... .......... 96% 1.43M 2s\n",
      " 20550K .......... .......... .......... .......... .......... 96% 4.74M 2s\n",
      " 20600K .......... .......... .......... .......... .......... 96% 14.6M 2s\n",
      " 20650K .......... .......... .......... .......... .......... 97% 23.8M 1s\n",
      " 20700K .......... .......... .......... .......... .......... 97%  373K 1s\n",
      " 20750K .......... .......... .......... .......... .......... 97% 1.72M 1s\n",
      " 20800K .......... .......... .......... .......... .......... 97% 1.31M 1s\n",
      " 20850K ........."
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/staging/software/lib/python3.9/site-packages/autocnv-1-py3.9.egg/raw_data/2021-10-28/hg19.refGene.gtf.gz  done\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      ". .......... .......... .......... .......... 97%  242K 1s\n",
      " 20900K .......... .......... .......... .......... .......... 98% 1.55M 1s\n",
      " 20950K .......... .......... .......... .......... .......... 98% 5.85M 1s\n",
      " 21000K .......... .......... .......... .......... .......... 98% 8.40M 1s\n",
      " 21050K .......... .......... .......... .......... .......... 98% 9.14M 1s\n",
      " 21100K .......... .......... .......... .......... .......... 99% 7.20M 0s\n",
      " 21150K .......... .......... .......... .......... .......... 99% 1.38M 0s\n",
      " 21200K .......... .......... .......... .......... .......... 99% 5.22M 0s\n",
      " 21250K .......... .......... .......... .......... .......... 99% 19.8M 0s\n",
      " 21300K .......... .......... .......... ....                 100% 36.3M=46s\n",
      "\n",
      "2024-05-11 14:00:42 (464 KB/s) - ‘/staging/software/lib/python3.9/site-packages/autocnv-1-py3.9.egg/raw_data/2021-10-28/hg19.refGene.gtf.gz ’ saved [21846291/21846291]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Download data\n",
    "import subprocess\n",
    "\n",
    "for i, j in zip(b, a):\n",
    "    if not os.path.exists(i):\n",
    "        # print(f'downloading {i} from {j}')\n",
    "        # request.urlretrieve(exon_url, exon_ori_file)\n",
    "        commands = f'wget -c -P {version_date_dir} {j}'\n",
    "        task = subprocess.run(\n",
    "                [\"wget\", \"-c\", \"-O\", i, j],\n",
    "                encoding='utf-8',\n",
    "        )\n",
    "        if task.returncode == 0:\n",
    "            print(f'{i} done')\n",
    "        else:\n",
    "            print(f'Rerun: {commands}')"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2024-05-11T06:00:42.184827400Z",
     "start_time": "2024-05-11T05:59:54.658868500Z"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 154,
   "outputs": [],
   "source": [
    "dgv_ori_file = os.path.join(raw_data_dir, 'DGV.GS.March2016.50percent.GainLossSep.Final.hg19.gff3')"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2024-05-11T06:00:42.192805Z",
     "start_time": "2024-05-11T06:00:42.188815700Z"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 155,
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "--2024-05-11 14:00:42--  http://dgv.tcag.ca/dgv/docs/DGV.GS.March2016.50percent.GainLossSep.Final.hg19.gff3\n",
      "Connecting to 192.168.134.12:1080... connected.\n",
      "Proxy request sent, awaiting response... 416 Requested Range Not Satisfiable\n",
      "\n",
      "    The file is already fully retrieved; nothing to do.\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": "0"
     },
     "execution_count": 155,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "os.system(f'wget -c -P {raw_data_dir} http://dgv.tcag.ca/dgv/docs/DGV.GS.March2016.50percent.GainLossSep.Final.hg19.gff3')"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2024-05-11T06:00:43.033572900Z",
     "start_time": "2024-05-11T06:00:42.279600200Z"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 156,
   "outputs": [],
   "source": [
    "gene_file = os.path.join(result_data_dir, 'gene.sorted.bed')\n",
    "omim_gene_file = os.path.join(result_data_dir, 'omim-gene.sorted.bed')\n",
    "clinvar_file = os.path.join(result_data_dir, 'clinvar-pathogenic.sorted.vcf')\n",
    "decipher_gene_file = os.path.join(result_data_dir, 'decipher-gene.sorted.bed')\n",
    "dgv_gain_file = os.path.join(result_data_dir, 'dgv-gain.sorted.bed')\n",
    "dgv_loss_file = os.path.join(result_data_dir, 'dgv-loss.sorted.bed')\n",
    "func_region_file = os.path.join(result_data_dir, 'func-region.sorted')\n",
    "gnomad_del_file = os.path.join(result_data_dir, 'gnomad-del.sorted.bed')\n",
    "gnomad_dup_file = os.path.join(result_data_dir, 'gnomad-dup.sorted.bed')\n",
    "hi_cds_file = os.path.join(result_data_dir, 'hi-cds.sorted.bed')\n",
    "hi_exon_file = os.path.join(result_data_dir, 'hi-exon.sorted.bed')\n",
    "hi_gene_file = os.path.join(result_data_dir, 'hi-gene.sorted.bed')\n",
    "hi_region_file = os.path.join(result_data_dir, 'hi-region.sorted.bed')\n",
    "ts_gene_file = os.path.join(result_data_dir, 'ts-gene.sorted.bed')\n",
    "ts_region_file = os.path.join(result_data_dir, 'ts-region.sorted.bed')\n",
    "uhi_gene_file = os.path.join(result_data_dir, 'uhi-gene.sorted.bed')\n",
    "uhi_region_file = os.path.join(result_data_dir, 'uhi-region.sorted.bed')\n",
    "uts_gene_file = os.path.join(result_data_dir, 'uts-gene.sorted.bed')\n",
    "uts_region_file = os.path.join(result_data_dir, 'uts-region.sorted.bed')\n"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2024-05-11T06:00:43.049530Z",
     "start_time": "2024-05-11T06:00:43.038204700Z"
    }
   }
  },
  {
   "cell_type": "markdown",
   "source": [
    "# Download require data\n",
    "\n",
    "if not os.path.exists(exon_ori_file):\n",
    "    print(f'downloading gene exon file')\n",
    "    request.urlretrieve(exon_url, exon_ori_file)\n",
    "    print('done')\n",
    "\n",
    "if not os.path.exists(cyto_band_ori_file):\n",
    "    print(f'downloading cytoband conf file to {cyto_band_ori_file}')\n",
    "    request.urlretrieve(cyto_band_url, cyto_band_ori_file)\n",
    "    print('done!')\n",
    "\n",
    "if not os.path.exists(gene_info_ori_file):\n",
    "    print(f'downloading gene info to {gene_info_ori_file}')\n",
    "    request.urlretrieve(gene_info_url, gene_info_ori_file)\n",
    "    print('done!')\n",
    "\n",
    "if not os.path.exists(ref_gene_ori_file):\n",
    "    print(f'downloading ref gene to {ref_gene_ori_file}')\n",
    "    request.urlretrieve(ref_gene_url, ref_gene_ori_file)\n",
    "    print('done!')\n",
    "\n",
    "if not os.path.exists(clingen_gene_ori_file):\n",
    "    print(f'downloading clingen gene list to {clingen_gene_ori_file}')\n",
    "    request.urlretrieve(clingen_gene_curation_url, clingen_gene_ori_file)\n",
    "    print('done!')\n",
    "\n",
    "if not os.path.exists(clingen_region_ori_file):\n",
    "    print(f'downloading clingen region file to {clingen_region_ori_file}')\n",
    "    request.urlretrieve(clingene_region_curation_url, clingen_region_ori_file)\n",
    "    print('done!')\n",
    "\n",
    "if not os.path.exists(clinvar_ori_vcf_file):\n",
    "    print(f'downloading clingen region file to {clinvar_ori_vcf_file}')\n",
    "    request.urlretrieve(clinvar_url, clinvar_ori_vcf_file)\n",
    "    print('done!')\n",
    "\n",
    "if not os.path.exists(hi_pred_ori_file):\n",
    "    print(f'downloading hi prediction file to {hi_pred_ori_file}')\n",
    "    request.urlretrieve(hi_pred_url, hi_pred_ori_file)\n",
    "    print('done!')\n",
    "if not os.path.exists(gnomad_lof_ori_file):\n",
    "    print(f'downloading pLoF file from GnomAD to {gnomad_lof_ori_file}')\n",
    "    request.urlretrieve(gnomad_lof_url, gnomad_lof_ori_file)\n",
    "    print('done!')\n",
    "if not os.path.exists(gnomad_control_ori_file):\n",
    "    print(f'downloading gnomad control only file to {gnomad_control_ori_file}')\n",
    "    request.urlretrieve(gnomad_control_only_url, gnomad_control_ori_file)\n",
    "    print('done!')\n",
    "\n",
    "if not os.path.exists(hgnc_gene_fam_file):\n",
    "    print(f'downloading gene family file to {hgnc_gene_fam_file}')\n",
    "    request.urlretrieve(hgnc_gene_fam_url, hgnc_gene_fam_file)\n",
    "    print('done!')"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": 157,
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    },
    "ExecuteTime": {
     "end_time": "2024-05-11T06:00:43.084436400Z",
     "start_time": "2024-05-11T06:00:43.058505800Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/staging/software/lib/python3.9/site-packages/autocnv-1-py3.9.egg/data/cyto-band.bed\n"
     ]
    }
   ],
   "source": [
    "cyto_band_df = pd.read_csv(\n",
    "    cyto_band_ori_file, sep='\\t',\n",
    "    names=['#chrom', 'start', 'end', 'name', 'gieStain']\n",
    ")\n",
    "cyto_band_df.to_csv(settings.CYTO_BAND_FILE.strip('.gz'), sep='\\t', index=False)\n",
    "print(settings.CYTO_BAND_FILE.strip('.gz'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 158,
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/staging/software/lib/python3.9/site-packages/autocnv-1-py3.9.egg/data/cyto-band.bed.gz\n"
     ]
    }
   ],
   "source": [
    "!bgzip -cf {settings.CYTO_BAND_FILE.strip(\".gz\")} > {settings.CYTO_BAND_FILE}\n",
    "!tabix -fp bed {settings.CYTO_BAND_FILE}\n",
    "!rm {settings.CYTO_BAND_FILE.strip(\".gz\")}\n",
    "print(settings.CYTO_BAND_FILE)"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    },
    "ExecuteTime": {
     "end_time": "2024-05-11T06:00:45.061014900Z",
     "start_time": "2024-05-11T06:00:43.087428400Z"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 159,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-05-11T06:00:45.815357300Z",
     "start_time": "2024-05-11T06:00:45.067987200Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/staging/software/lib/python3.9/site-packages/autocnv-1-py3.9.egg/raw_data/2021-10-28/refGene.txt.gz\n"
     ]
    },
    {
     "data": {
      "text/plain": "       bin          name  chrom strand    txStart      txEnd   cdsStart  \\\n10573   26  NM_001351365   chr1      +  144146811  146467744  144146846   \n10459   26  NM_001278267   chr1      +  144146810  146467744  144158383   \n18617   12     NM_000109   chrX      -   31137340   33357505   31140035   \n29979    2  NM_001351274  chr11      -   83166055   85339417   83170860   \n68033    9     NM_033225   chr8      -    2792882    4852436    2796106   \n\n          cdsEnd  exonCount  \\\n10573  146466121         93   \n10459  146466121        131   \n18617   33357382         79   \n29979   85337661         27   \n68033    4851938         70   \n\n                                              exonStarts  \\\n10573  144146811,144148789,144149726,144150981,144151...   \n10459  144146810,144148789,144149726,144150981,144151...   \n18617  31137340,31144758,31152218,31164407,31165391,3...   \n29979  83166055,83173044,83177750,83180243,83182668,8...   \n68033  2792882,2799993,2806820,2807752,2808635,281174...   \n\n                                                exonEnds  score   name2  \\\n10573  144147021,144148892,144149941,144151054,144151...      0  NBPF19   \n10459  144147021,144148892,144149941,144151054,144151...      0  NBPF20   \n18617  31140047,31144790,31152311,31164531,31165635,3...      0     DMD   \n29979  83170967,83173136,83177860,83180416,83182770,8...      0    DLG2   \n68033  2796266,2800126,2806908,2807865,2808797,281179...      0   CSMD1   \n\n      cdsStartStat cdsEndStat  \\\n10573         cmpl       cmpl   \n10459         cmpl       cmpl   \n18617         cmpl       cmpl   \n29979         cmpl       cmpl   \n68033         cmpl       cmpl   \n\n                                              ExonFrames   length  \n10573  0,1,2,1,2,1,2,1,2,1,2,1,2,1,2,1,2,1,2,1,2,1,2,...  2319275  \n10459  -1,-1,-1,-1,-1,-1,-1,-1,0,2,1,2,1,2,1,2,1,2,1,...  2307738  \n18617  0,1,1,0,2,2,2,2,2,0,2,0,1,2,1,1,2,1,0,0,1,0,2,...  2217347  \n29979  1,2,0,1,1,1,0,2,2,1,0,2,2,2,1,0,1,2,0,0,0,0,0,...  2166801  \n68033  2,1,0,1,1,1,1,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,...  2055832  ",
      "text/html": "<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>bin</th>\n      <th>name</th>\n      <th>chrom</th>\n      <th>strand</th>\n      <th>txStart</th>\n      <th>txEnd</th>\n      <th>cdsStart</th>\n      <th>cdsEnd</th>\n      <th>exonCount</th>\n      <th>exonStarts</th>\n      <th>exonEnds</th>\n      <th>score</th>\n      <th>name2</th>\n      <th>cdsStartStat</th>\n      <th>cdsEndStat</th>\n      <th>ExonFrames</th>\n      <th>length</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>10573</th>\n      <td>26</td>\n      <td>NM_001351365</td>\n      <td>chr1</td>\n      <td>+</td>\n      <td>144146811</td>\n      <td>146467744</td>\n      <td>144146846</td>\n      <td>146466121</td>\n      <td>93</td>\n      <td>144146811,144148789,144149726,144150981,144151...</td>\n      <td>144147021,144148892,144149941,144151054,144151...</td>\n      <td>0</td>\n      <td>NBPF19</td>\n      <td>cmpl</td>\n      <td>cmpl</td>\n      <td>0,1,2,1,2,1,2,1,2,1,2,1,2,1,2,1,2,1,2,1,2,1,2,...</td>\n      <td>2319275</td>\n    </tr>\n    <tr>\n      <th>10459</th>\n      <td>26</td>\n      <td>NM_001278267</td>\n      <td>chr1</td>\n      <td>+</td>\n      <td>144146810</td>\n      <td>146467744</td>\n      <td>144158383</td>\n      <td>146466121</td>\n      <td>131</td>\n      <td>144146810,144148789,144149726,144150981,144151...</td>\n      <td>144147021,144148892,144149941,144151054,144151...</td>\n      <td>0</td>\n      <td>NBPF20</td>\n      <td>cmpl</td>\n      <td>cmpl</td>\n      <td>-1,-1,-1,-1,-1,-1,-1,-1,0,2,1,2,1,2,1,2,1,2,1,...</td>\n      <td>2307738</td>\n    </tr>\n    <tr>\n      <th>18617</th>\n      <td>12</td>\n      <td>NM_000109</td>\n      <td>chrX</td>\n      <td>-</td>\n      <td>31137340</td>\n      <td>33357505</td>\n      <td>31140035</td>\n      <td>33357382</td>\n      <td>79</td>\n      <td>31137340,31144758,31152218,31164407,31165391,3...</td>\n      <td>31140047,31144790,31152311,31164531,31165635,3...</td>\n      <td>0</td>\n      <td>DMD</td>\n      <td>cmpl</td>\n      <td>cmpl</td>\n      <td>0,1,1,0,2,2,2,2,2,0,2,0,1,2,1,1,2,1,0,0,1,0,2,...</td>\n      <td>2217347</td>\n    </tr>\n    <tr>\n      <th>29979</th>\n      <td>2</td>\n      <td>NM_001351274</td>\n      <td>chr11</td>\n      <td>-</td>\n      <td>83166055</td>\n      <td>85339417</td>\n      <td>83170860</td>\n      <td>85337661</td>\n      <td>27</td>\n      <td>83166055,83173044,83177750,83180243,83182668,8...</td>\n      <td>83170967,83173136,83177860,83180416,83182770,8...</td>\n      <td>0</td>\n      <td>DLG2</td>\n      <td>cmpl</td>\n      <td>cmpl</td>\n      <td>1,2,0,1,1,1,0,2,2,1,0,2,2,2,1,0,1,2,0,0,0,0,0,...</td>\n      <td>2166801</td>\n    </tr>\n    <tr>\n      <th>68033</th>\n      <td>9</td>\n      <td>NM_033225</td>\n      <td>chr8</td>\n      <td>-</td>\n      <td>2792882</td>\n      <td>4852436</td>\n      <td>2796106</td>\n      <td>4851938</td>\n      <td>70</td>\n      <td>2792882,2799993,2806820,2807752,2808635,281174...</td>\n      <td>2796266,2800126,2806908,2807865,2808797,281179...</td>\n      <td>0</td>\n      <td>CSMD1</td>\n      <td>cmpl</td>\n      <td>cmpl</td>\n      <td>2,1,0,1,1,1,1,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,...</td>\n      <td>2055832</td>\n    </tr>\n  </tbody>\n</table>\n</div>"
     },
     "execution_count": 159,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cols = [\n",
    "    'bin', 'name', 'chrom', 'strand', 'txStart', 'txEnd', 'cdsStart', 'cdsEnd', 'exonCount', 'exonStarts', 'exonEnds',\n",
    "    'score', 'name2', 'cdsStartStat', 'cdsEndStat', 'ExonFrames']\n",
    "refgene = pd.read_csv(ref_gene_ori_file, sep='\\t', names=cols)\n",
    "print(ref_gene_ori_file)\n",
    "refgene = refgene[~refgene['chrom'].str.match(r'.*fix$')]\n",
    "refgene['length'] = refgene['cdsEnd'] - refgene['cdsStart']\n",
    "refgene = refgene.sort_values('length', ascending=False)\n",
    "refgene = refgene.drop_duplicates('name2', keep='first')\n",
    "refgene.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 160,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-05-11T06:00:46.521331900Z",
     "start_time": "2024-05-11T06:00:45.818349600Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": "   #tax_id  GeneID Symbol LocusTag                   Synonyms  \\\n0     9606       1   A1BG        -       A1B|ABG|GAB|HYST2477   \n1     9606       2    A2M        -  A2MD|CPAMD5|FWP007|S863-7   \n2     9606       3  A2MP1        -                       A2MP   \n3     9606       9   NAT1        -       AAC1|MNAT|NAT-1|NATI   \n4     9606      10   NAT2        -            AAC2|NAT-2|PNAT   \n\n                                             dbXrefs chromosome map_location  \\\n0  MIM:138670|HGNC:HGNC:5|Ensembl:ENSG00000121410...         19     19q13.43   \n1  MIM:103950|HGNC:HGNC:7|Ensembl:ENSG00000175899...         12     12p13.31   \n2  HGNC:HGNC:8|Ensembl:ENSG00000291190|AllianceGe...         12     12p13.31   \n3  MIM:108345|HGNC:HGNC:7645|Ensembl:ENSG00000171...          8         8p22   \n4  MIM:612182|HGNC:HGNC:7646|Ensembl:ENSG00000156...          8         8p22   \n\n                          description    type_of_gene  \\\n0              alpha-1-B glycoprotein  protein-coding   \n1               alpha-2-macroglobulin  protein-coding   \n2  alpha-2-macroglobulin pseudogene 1          pseudo   \n3               N-acetyltransferase 1  protein-coding   \n4               N-acetyltransferase 2  protein-coding   \n\n  Symbol_from_nomenclature_authority Full_name_from_nomenclature_authority  \\\n0                               A1BG                alpha-1-B glycoprotein   \n1                                A2M                 alpha-2-macroglobulin   \n2                              A2MP1    alpha-2-macroglobulin pseudogene 1   \n3                               NAT1                 N-acetyltransferase 1   \n4                               NAT2                 N-acetyltransferase 2   \n\n  Nomenclature_status                                 Other_designations  \\\n0                   O  alpha-1B-glycoprotein|HEL-S-163pA|epididymis s...   \n1                   O  alpha-2-macroglobulin|C3 and PZP-like alpha-2-...   \n2                   O                  pregnancy-zone protein pseudogene   \n3                   O  arylamine N-acetyltransferase 1|N-acetyltransf...   \n4                   O  arylamine N-acetyltransferase 2|N-acetyltransf...   \n\n   Modification_date Feature_type  \n0           20240503            -  \n1           20240505            -  \n2           20231010            -  \n3           20240403            -  \n4           20240502            -  ",
      "text/html": "<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>#tax_id</th>\n      <th>GeneID</th>\n      <th>Symbol</th>\n      <th>LocusTag</th>\n      <th>Synonyms</th>\n      <th>dbXrefs</th>\n      <th>chromosome</th>\n      <th>map_location</th>\n      <th>description</th>\n      <th>type_of_gene</th>\n      <th>Symbol_from_nomenclature_authority</th>\n      <th>Full_name_from_nomenclature_authority</th>\n      <th>Nomenclature_status</th>\n      <th>Other_designations</th>\n      <th>Modification_date</th>\n      <th>Feature_type</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>0</th>\n      <td>9606</td>\n      <td>1</td>\n      <td>A1BG</td>\n      <td>-</td>\n      <td>A1B|ABG|GAB|HYST2477</td>\n      <td>MIM:138670|HGNC:HGNC:5|Ensembl:ENSG00000121410...</td>\n      <td>19</td>\n      <td>19q13.43</td>\n      <td>alpha-1-B glycoprotein</td>\n      <td>protein-coding</td>\n      <td>A1BG</td>\n      <td>alpha-1-B glycoprotein</td>\n      <td>O</td>\n      <td>alpha-1B-glycoprotein|HEL-S-163pA|epididymis s...</td>\n      <td>20240503</td>\n      <td>-</td>\n    </tr>\n    <tr>\n      <th>1</th>\n      <td>9606</td>\n      <td>2</td>\n      <td>A2M</td>\n      <td>-</td>\n      <td>A2MD|CPAMD5|FWP007|S863-7</td>\n      <td>MIM:103950|HGNC:HGNC:7|Ensembl:ENSG00000175899...</td>\n      <td>12</td>\n      <td>12p13.31</td>\n      <td>alpha-2-macroglobulin</td>\n      <td>protein-coding</td>\n      <td>A2M</td>\n      <td>alpha-2-macroglobulin</td>\n      <td>O</td>\n      <td>alpha-2-macroglobulin|C3 and PZP-like alpha-2-...</td>\n      <td>20240505</td>\n      <td>-</td>\n    </tr>\n    <tr>\n      <th>2</th>\n      <td>9606</td>\n      <td>3</td>\n      <td>A2MP1</td>\n      <td>-</td>\n      <td>A2MP</td>\n      <td>HGNC:HGNC:8|Ensembl:ENSG00000291190|AllianceGe...</td>\n      <td>12</td>\n      <td>12p13.31</td>\n      <td>alpha-2-macroglobulin pseudogene 1</td>\n      <td>pseudo</td>\n      <td>A2MP1</td>\n      <td>alpha-2-macroglobulin pseudogene 1</td>\n      <td>O</td>\n      <td>pregnancy-zone protein pseudogene</td>\n      <td>20231010</td>\n      <td>-</td>\n    </tr>\n    <tr>\n      <th>3</th>\n      <td>9606</td>\n      <td>9</td>\n      <td>NAT1</td>\n      <td>-</td>\n      <td>AAC1|MNAT|NAT-1|NATI</td>\n      <td>MIM:108345|HGNC:HGNC:7645|Ensembl:ENSG00000171...</td>\n      <td>8</td>\n      <td>8p22</td>\n      <td>N-acetyltransferase 1</td>\n      <td>protein-coding</td>\n      <td>NAT1</td>\n      <td>N-acetyltransferase 1</td>\n      <td>O</td>\n      <td>arylamine N-acetyltransferase 1|N-acetyltransf...</td>\n      <td>20240403</td>\n      <td>-</td>\n    </tr>\n    <tr>\n      <th>4</th>\n      <td>9606</td>\n      <td>10</td>\n      <td>NAT2</td>\n      <td>-</td>\n      <td>AAC2|NAT-2|PNAT</td>\n      <td>MIM:612182|HGNC:HGNC:7646|Ensembl:ENSG00000156...</td>\n      <td>8</td>\n      <td>8p22</td>\n      <td>N-acetyltransferase 2</td>\n      <td>protein-coding</td>\n      <td>NAT2</td>\n      <td>N-acetyltransferase 2</td>\n      <td>O</td>\n      <td>arylamine N-acetyltransferase 2|N-acetyltransf...</td>\n      <td>20240502</td>\n      <td>-</td>\n    </tr>\n  </tbody>\n</table>\n</div>"
     },
     "execution_count": 160,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_info = pd.read_csv(gene_info_ori_file, sep='\\t')\n",
    "gene_info.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 161,
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    },
    "ExecuteTime": {
     "end_time": "2024-05-11T06:00:46.658887300Z",
     "start_time": "2024-05-11T06:00:46.526319700Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": "   bin          name  chrom strand    txStart      txEnd   cdsStart  \\\n0   26  NM_001351365   chr1      +  144146811  146467744  144146846   \n1   26  NM_001278267   chr1      +  144146810  146467744  144158383   \n2   12     NM_000109   chrX      -   31137340   33357505   31140035   \n3    2  NM_001351274  chr11      -   83166055   85339417   83170860   \n4    9     NM_033225   chr8      -    2792882    4852436    2796106   \n\n      cdsEnd  exonCount                                         exonStarts  \\\n0  146466121         93  144146811,144148789,144149726,144150981,144151...   \n1  146466121        131  144146810,144148789,144149726,144150981,144151...   \n2   33357382         79  31137340,31144758,31152218,31164407,31165391,3...   \n3   85337661         27  83166055,83173044,83177750,83180243,83182668,8...   \n4    4851938         70  2792882,2799993,2806820,2807752,2808635,281174...   \n\n                                            exonEnds  score   name2  \\\n0  144147021,144148892,144149941,144151054,144151...      0  NBPF19   \n1  144147021,144148892,144149941,144151054,144151...      0  NBPF20   \n2  31140047,31144790,31152311,31164531,31165635,3...      0     DMD   \n3  83170967,83173136,83177860,83180416,83182770,8...      0    DLG2   \n4  2796266,2800126,2806908,2807865,2808797,281179...      0   CSMD1   \n\n  cdsStartStat cdsEndStat                                         ExonFrames  \\\n0         cmpl       cmpl  0,1,2,1,2,1,2,1,2,1,2,1,2,1,2,1,2,1,2,1,2,1,2,...   \n1         cmpl       cmpl  -1,-1,-1,-1,-1,-1,-1,-1,0,2,1,2,1,2,1,2,1,2,1,...   \n2         cmpl       cmpl  0,1,1,0,2,2,2,2,2,0,2,0,1,2,1,1,2,1,0,0,1,0,2,...   \n3         cmpl       cmpl  1,2,0,1,1,1,0,2,2,1,0,2,2,2,1,0,1,2,0,0,0,0,0,...   \n4         cmpl       cmpl  2,1,0,1,1,1,1,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,...   \n\n    length  #tax_id     GeneID  Symbol LocusTag  \\\n0  2319275     9606  101060226  NBPF19        -   \n1  2307738     9606  100288142  NBPF20        -   \n2  2217347     9606       1756     DMD        -   \n3  2166801     9606       1740    DLG2        -   \n4  2055832     9606      64478   CSMD1        -   \n\n                                            Synonyms  \\\n0                                                  -   \n1                                                  -   \n2  BMD|CMD3B|DXS142|DXS164|DXS206|DXS230|DXS239|D...   \n3                   PPP1R58|PSD-93|PSD93|chapsyn-110   \n4                                            PPP1R24   \n\n                                             dbXrefs chromosome  map_location  \\\n0  MIM:614006|HGNC:HGNC:31999|Ensembl:ENSG0000027...          1        1q21.2   \n1  MIM:614007|HGNC:HGNC:32000|Ensembl:ENSG0000016...          1        1q21.1   \n2  MIM:300377|HGNC:HGNC:2928|Ensembl:ENSG00000198...          X  Xp21.2-p21.1   \n3  MIM:603583|HGNC:HGNC:2901|Ensembl:ENSG00000150...         11       11q14.1   \n4  MIM:608397|HGNC:HGNC:14026|Ensembl:ENSG0000018...          8        8p23.2   \n\n                            description    type_of_gene  \\\n0                        NBPF member 19  protein-coding   \n1                        NBPF member 20  protein-coding   \n2                            dystrophin  protein-coding   \n3  discs large MAGUK scaffold protein 2  protein-coding   \n4      CUB and Sushi multiple domains 1  protein-coding   \n\n  Symbol_from_nomenclature_authority Full_name_from_nomenclature_authority  \\\n0                             NBPF19                        NBPF member 19   \n1                             NBPF20                        NBPF member 20   \n2                                DMD                            dystrophin   \n3                               DLG2  discs large MAGUK scaffold protein 2   \n4                              CSMD1      CUB and Sushi multiple domains 1   \n\n  Nomenclature_status                                 Other_designations  \\\n0                   O          neuroblastoma breakpoint family member 19   \n1                   O          neuroblastoma breakpoint family member 20   \n2                   O                       dystrophin|mutant dystrophin   \n3                   O  disks large homolog 2|channel-associated prote...   \n4                   O  CUB and sushi domain-containing protein 1|prot...   \n\n   Modification_date Feature_type  \n0           20240407            -  \n1           20240403            -  \n2           20240506            -  \n3           20240411            -  \n4           20240407            -  ",
      "text/html": "<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>bin</th>\n      <th>name</th>\n      <th>chrom</th>\n      <th>strand</th>\n      <th>txStart</th>\n      <th>txEnd</th>\n      <th>cdsStart</th>\n      <th>cdsEnd</th>\n      <th>exonCount</th>\n      <th>exonStarts</th>\n      <th>exonEnds</th>\n      <th>score</th>\n      <th>name2</th>\n      <th>cdsStartStat</th>\n      <th>cdsEndStat</th>\n      <th>ExonFrames</th>\n      <th>length</th>\n      <th>#tax_id</th>\n      <th>GeneID</th>\n      <th>Symbol</th>\n      <th>LocusTag</th>\n      <th>Synonyms</th>\n      <th>dbXrefs</th>\n      <th>chromosome</th>\n      <th>map_location</th>\n      <th>description</th>\n      <th>type_of_gene</th>\n      <th>Symbol_from_nomenclature_authority</th>\n      <th>Full_name_from_nomenclature_authority</th>\n      <th>Nomenclature_status</th>\n      <th>Other_designations</th>\n      <th>Modification_date</th>\n      <th>Feature_type</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>0</th>\n      <td>26</td>\n      <td>NM_001351365</td>\n      <td>chr1</td>\n      <td>+</td>\n      <td>144146811</td>\n      <td>146467744</td>\n      <td>144146846</td>\n      <td>146466121</td>\n      <td>93</td>\n      <td>144146811,144148789,144149726,144150981,144151...</td>\n      <td>144147021,144148892,144149941,144151054,144151...</td>\n      <td>0</td>\n      <td>NBPF19</td>\n      <td>cmpl</td>\n      <td>cmpl</td>\n      <td>0,1,2,1,2,1,2,1,2,1,2,1,2,1,2,1,2,1,2,1,2,1,2,...</td>\n      <td>2319275</td>\n      <td>9606</td>\n      <td>101060226</td>\n      <td>NBPF19</td>\n      <td>-</td>\n      <td>-</td>\n      <td>MIM:614006|HGNC:HGNC:31999|Ensembl:ENSG0000027...</td>\n      <td>1</td>\n      <td>1q21.2</td>\n      <td>NBPF member 19</td>\n      <td>protein-coding</td>\n      <td>NBPF19</td>\n      <td>NBPF member 19</td>\n      <td>O</td>\n      <td>neuroblastoma breakpoint family member 19</td>\n      <td>20240407</td>\n      <td>-</td>\n    </tr>\n    <tr>\n      <th>1</th>\n      <td>26</td>\n      <td>NM_001278267</td>\n      <td>chr1</td>\n      <td>+</td>\n      <td>144146810</td>\n      <td>146467744</td>\n      <td>144158383</td>\n      <td>146466121</td>\n      <td>131</td>\n      <td>144146810,144148789,144149726,144150981,144151...</td>\n      <td>144147021,144148892,144149941,144151054,144151...</td>\n      <td>0</td>\n      <td>NBPF20</td>\n      <td>cmpl</td>\n      <td>cmpl</td>\n      <td>-1,-1,-1,-1,-1,-1,-1,-1,0,2,1,2,1,2,1,2,1,2,1,...</td>\n      <td>2307738</td>\n      <td>9606</td>\n      <td>100288142</td>\n      <td>NBPF20</td>\n      <td>-</td>\n      <td>-</td>\n      <td>MIM:614007|HGNC:HGNC:32000|Ensembl:ENSG0000016...</td>\n      <td>1</td>\n      <td>1q21.1</td>\n      <td>NBPF member 20</td>\n      <td>protein-coding</td>\n      <td>NBPF20</td>\n      <td>NBPF member 20</td>\n      <td>O</td>\n      <td>neuroblastoma breakpoint family member 20</td>\n      <td>20240403</td>\n      <td>-</td>\n    </tr>\n    <tr>\n      <th>2</th>\n      <td>12</td>\n      <td>NM_000109</td>\n      <td>chrX</td>\n      <td>-</td>\n      <td>31137340</td>\n      <td>33357505</td>\n      <td>31140035</td>\n      <td>33357382</td>\n      <td>79</td>\n      <td>31137340,31144758,31152218,31164407,31165391,3...</td>\n      <td>31140047,31144790,31152311,31164531,31165635,3...</td>\n      <td>0</td>\n      <td>DMD</td>\n      <td>cmpl</td>\n      <td>cmpl</td>\n      <td>0,1,1,0,2,2,2,2,2,0,2,0,1,2,1,1,2,1,0,0,1,0,2,...</td>\n      <td>2217347</td>\n      <td>9606</td>\n      <td>1756</td>\n      <td>DMD</td>\n      <td>-</td>\n      <td>BMD|CMD3B|DXS142|DXS164|DXS206|DXS230|DXS239|D...</td>\n      <td>MIM:300377|HGNC:HGNC:2928|Ensembl:ENSG00000198...</td>\n      <td>X</td>\n      <td>Xp21.2-p21.1</td>\n      <td>dystrophin</td>\n      <td>protein-coding</td>\n      <td>DMD</td>\n      <td>dystrophin</td>\n      <td>O</td>\n      <td>dystrophin|mutant dystrophin</td>\n      <td>20240506</td>\n      <td>-</td>\n    </tr>\n    <tr>\n      <th>3</th>\n      <td>2</td>\n      <td>NM_001351274</td>\n      <td>chr11</td>\n      <td>-</td>\n      <td>83166055</td>\n      <td>85339417</td>\n      <td>83170860</td>\n      <td>85337661</td>\n      <td>27</td>\n      <td>83166055,83173044,83177750,83180243,83182668,8...</td>\n      <td>83170967,83173136,83177860,83180416,83182770,8...</td>\n      <td>0</td>\n      <td>DLG2</td>\n      <td>cmpl</td>\n      <td>cmpl</td>\n      <td>1,2,0,1,1,1,0,2,2,1,0,2,2,2,1,0,1,2,0,0,0,0,0,...</td>\n      <td>2166801</td>\n      <td>9606</td>\n      <td>1740</td>\n      <td>DLG2</td>\n      <td>-</td>\n      <td>PPP1R58|PSD-93|PSD93|chapsyn-110</td>\n      <td>MIM:603583|HGNC:HGNC:2901|Ensembl:ENSG00000150...</td>\n      <td>11</td>\n      <td>11q14.1</td>\n      <td>discs large MAGUK scaffold protein 2</td>\n      <td>protein-coding</td>\n      <td>DLG2</td>\n      <td>discs large MAGUK scaffold protein 2</td>\n      <td>O</td>\n      <td>disks large homolog 2|channel-associated prote...</td>\n      <td>20240411</td>\n      <td>-</td>\n    </tr>\n    <tr>\n      <th>4</th>\n      <td>9</td>\n      <td>NM_033225</td>\n      <td>chr8</td>\n      <td>-</td>\n      <td>2792882</td>\n      <td>4852436</td>\n      <td>2796106</td>\n      <td>4851938</td>\n      <td>70</td>\n      <td>2792882,2799993,2806820,2807752,2808635,281174...</td>\n      <td>2796266,2800126,2806908,2807865,2808797,281179...</td>\n      <td>0</td>\n      <td>CSMD1</td>\n      <td>cmpl</td>\n      <td>cmpl</td>\n      <td>2,1,0,1,1,1,1,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,...</td>\n      <td>2055832</td>\n      <td>9606</td>\n      <td>64478</td>\n      <td>CSMD1</td>\n      <td>-</td>\n      <td>PPP1R24</td>\n      <td>MIM:608397|HGNC:HGNC:14026|Ensembl:ENSG0000018...</td>\n      <td>8</td>\n      <td>8p23.2</td>\n      <td>CUB and Sushi multiple domains 1</td>\n      <td>protein-coding</td>\n      <td>CSMD1</td>\n      <td>CUB and Sushi multiple domains 1</td>\n      <td>O</td>\n      <td>CUB and sushi domain-containing protein 1|prot...</td>\n      <td>20240407</td>\n      <td>-</td>\n    </tr>\n  </tbody>\n</table>\n</div>"
     },
     "execution_count": 161,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "refgene_info = refgene.merge(gene_info, left_on='name2', right_on='Symbol')\n",
    "refgene_info.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 162,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-05-11T06:00:46.690951600Z",
     "start_time": "2024-05-11T06:00:46.601874100Z"
    }
   },
   "outputs": [],
   "source": [
    "gene_fam = pd.read_csv(hgnc_gene_fam_file, sep=',')\n",
    "gene_fam = gene_fam[gene_fam['typical_gene'].notna()].rename(columns={'id': 'fam'})\n",
    "gene_fam_grp = gene_fam.groupby('typical_gene').agg({'fam': set})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 163,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-05-11T06:00:46.698924100Z",
     "start_time": "2024-05-11T06:00:46.645887400Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": "      chrom  txStart   txEnd  GeneID   name2          name strand\n18195  chr1    69090   70008   79501   OR4F5  NM_001005484      +\n18011  chr1   367658  368597  729759  OR4F29  NM_001005221      +\n18008  chr1   621095  622034   81399  OR4F16  NM_001005277      -\n9159   chr1   859302  879954  148398  SAMD11  NM_001385641      +\n10490  chr1   879582  894636   26155   NOC2L     NM_015658      -",
      "text/html": "<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>chrom</th>\n      <th>txStart</th>\n      <th>txEnd</th>\n      <th>GeneID</th>\n      <th>name2</th>\n      <th>name</th>\n      <th>strand</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>18195</th>\n      <td>chr1</td>\n      <td>69090</td>\n      <td>70008</td>\n      <td>79501</td>\n      <td>OR4F5</td>\n      <td>NM_001005484</td>\n      <td>+</td>\n    </tr>\n    <tr>\n      <th>18011</th>\n      <td>chr1</td>\n      <td>367658</td>\n      <td>368597</td>\n      <td>729759</td>\n      <td>OR4F29</td>\n      <td>NM_001005221</td>\n      <td>+</td>\n    </tr>\n    <tr>\n      <th>18008</th>\n      <td>chr1</td>\n      <td>621095</td>\n      <td>622034</td>\n      <td>81399</td>\n      <td>OR4F16</td>\n      <td>NM_001005277</td>\n      <td>-</td>\n    </tr>\n    <tr>\n      <th>9159</th>\n      <td>chr1</td>\n      <td>859302</td>\n      <td>879954</td>\n      <td>148398</td>\n      <td>SAMD11</td>\n      <td>NM_001385641</td>\n      <td>+</td>\n    </tr>\n    <tr>\n      <th>10490</th>\n      <td>chr1</td>\n      <td>879582</td>\n      <td>894636</td>\n      <td>26155</td>\n      <td>NOC2L</td>\n      <td>NM_015658</td>\n      <td>-</td>\n    </tr>\n  </tbody>\n</table>\n</div>"
     },
     "execution_count": 163,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene = refgene_info.loc[\n",
    "    refgene_info['type_of_gene'] == 'protein-coding',\n",
    "    ['chrom', 'txStart', 'txEnd', 'GeneID', 'name2', 'name', 'strand']\n",
    "].sort_values(['chrom', 'txStart', 'txEnd'])\n",
    "gene.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 164,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-05-11T06:00:46.871901300Z",
     "start_time": "2024-05-11T06:00:46.690951600Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/staging/software/lib/python3.9/site-packages/autocnv-1-py3.9.egg/data/gene.sorted.bed\n"
     ]
    }
   ],
   "source": [
    "gene.rename(columns={\n",
    "    'chrom': '#chrom', 'txStart': 'start', 'txEnd': 'end',\n",
    "    'GeneID': 'gene_id', 'name2': 'symbol', 'name': 'transcript'\n",
    "}).to_csv(gene_file, index=False, sep='\\t')\n",
    "print(gene_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 165,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-05-11T06:00:48.896496200Z",
     "start_time": "2024-05-11T06:00:46.811035600Z"
    }
   },
   "outputs": [],
   "source": [
    "!bgzip -cf {gene_file} > {gene_file}.gz\n",
    "!tabix -fp bed {gene_file}.gz\n",
    "!rm {gene_file}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 166,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-05-11T06:00:52.199607200Z",
     "start_time": "2024-05-11T06:00:48.990333100Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--2024-05-11 14:00:49--  https://www.omim.org/static/omim/data/mim2gene.txt\r\n",
      "Connecting to 192.168.134.12:1080... connected.\r\n",
      "Proxy request sent, awaiting response... 416 Requested Range Not Satisfiable\r\n",
      "\r\n",
      "    The file is already fully retrieved; nothing to do.\r\n",
      "\r\n"
     ]
    }
   ],
   "source": [
    "# # prep an external omim gene list\n",
    "# omim_gene_list_file = os.path.join(raw_data_dir, 'gene-list-key-lte3.xlsx')\n",
    "# raw_omim_df = pd.read_excel(omim_gene_list_file)\n",
    "## can not download gene-list-key-lt3.xlsx\n",
    "## use https://www.omim.org/static/omim/data/mim2gene.txt to extract omim_gene instead\n",
    "!wget -c -P {raw_data_dir} \"https://www.omim.org/static/omim/data/mim2gene.txt\"\n",
    "omim_ori_file = os.path.join(raw_data_dir, 'mim2gene.txt')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 167,
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['# MIM Number',\n",
      "       'MIM Entry Type (see FAQ 1.3 at https://omim.org/help/faq)',\n",
      "       'Entrez Gene ID (NCBI)', 'Approved Gene Symbol (HGNC)',\n",
      "       'Ensembl Gene ID (Ensembl)'],\n",
      "      dtype='object')\n"
     ]
    },
    {
     "data": {
      "text/plain": "   # MIM Number MIM Entry Type (see FAQ 1.3 at https://omim.org/help/faq)  \\\n0        100050                           predominantly phenotypes          \n1        100070                                          phenotype          \n2        100100                                          phenotype          \n3        100200                           predominantly phenotypes          \n4        100300                                          phenotype          \n\n   Entrez Gene ID (NCBI) Approved Gene Symbol (HGNC) Ensembl Gene ID (Ensembl)  \n0                    NaN                         NaN                       NaN  \n1            100329167.0                         NaN                       NaN  \n2                    NaN                         NaN                       NaN  \n3                    NaN                         NaN                       NaN  \n4                    NaN                         NaN                       NaN  ",
      "text/html": "<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th># MIM Number</th>\n      <th>MIM Entry Type (see FAQ 1.3 at https://omim.org/help/faq)</th>\n      <th>Entrez Gene ID (NCBI)</th>\n      <th>Approved Gene Symbol (HGNC)</th>\n      <th>Ensembl Gene ID (Ensembl)</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>0</th>\n      <td>100050</td>\n      <td>predominantly phenotypes</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n    </tr>\n    <tr>\n      <th>1</th>\n      <td>100070</td>\n      <td>phenotype</td>\n      <td>100329167.0</td>\n      <td>NaN</td>\n      <td>NaN</td>\n    </tr>\n    <tr>\n      <th>2</th>\n      <td>100100</td>\n      <td>phenotype</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n    </tr>\n    <tr>\n      <th>3</th>\n      <td>100200</td>\n      <td>predominantly phenotypes</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n    </tr>\n    <tr>\n      <th>4</th>\n      <td>100300</td>\n      <td>phenotype</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n    </tr>\n  </tbody>\n</table>\n</div>"
     },
     "execution_count": 167,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "raw_omim_df = pd.read_csv(omim_ori_file, sep='\\t', skiprows=4)\n",
    "print(raw_omim_df.columns)\n",
    "raw_omim_df.head()"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2024-05-11T06:00:52.275763300Z",
     "start_time": "2024-05-11T06:00:52.206613900Z"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 168,
   "outputs": [
    {
     "data": {
      "text/plain": "{'TMEM160',\n 'MSTO1',\n 'H3-4',\n 'MPZL3',\n 'DDX42',\n 'NPR2',\n 'SMR3A',\n 'SPINK2',\n 'DDAH2',\n 'CRIPTO',\n 'MON1B',\n 'SCGB2B2',\n 'NOS2',\n 'PRSS16',\n 'MAST3',\n 'SLC35E3',\n 'ULBP2',\n 'MAGI2',\n 'LY75',\n 'UGT1A9',\n 'TENM4',\n 'SCT',\n 'ZNF703',\n 'RNGTT',\n 'COX6B1',\n 'NBPF5P',\n 'ARHGEF3',\n 'CPPED1',\n 'HMGCLL1',\n 'PTGDR2',\n 'NOD2',\n 'RHOU',\n 'PLA2G4B',\n 'ZNF19',\n 'B2M',\n 'MCU',\n 'DNA2',\n 'LSM5',\n 'MIR520F',\n 'C2orf69',\n 'DCAF1',\n 'IRF6',\n 'TNIP2',\n 'MSH4',\n 'FABP6',\n 'PCDHB5',\n 'EFEMP2',\n 'C1QTNF3',\n 'APOL3',\n 'SFXN2',\n 'BSG',\n 'GOLPH3L',\n 'SHISA9',\n 'PARP9',\n 'STOX1',\n 'UQCR11',\n 'TNFRSF6B',\n 'FAM83G',\n 'NRK',\n 'SNORD114-1',\n 'LTBP1',\n 'INPP5A',\n 'CD300H',\n 'GMFB',\n 'ITGAV',\n 'HPX',\n 'GATM',\n 'RAB3C',\n 'ZNF668',\n 'CFAP44',\n 'CRHBP',\n 'SLC6A15',\n 'ISG20',\n 'MBOAT1',\n 'HMGCS1',\n 'SGK1',\n 'CDC37',\n 'FAAP100',\n 'USP6',\n 'ANKFY1',\n 'SGMS1',\n 'NFAM1',\n 'SLC23A1',\n 'SBF1',\n 'EPM2AIP1',\n 'BDNF',\n 'AKT1',\n 'ATP2A2',\n 'GABRR2',\n 'COP1',\n 'RBM45',\n 'PPP6R3',\n 'TACC2',\n 'UTP20',\n 'TMEM175',\n 'FBXL15',\n 'CRELD1',\n 'NCLN',\n 'ZNF14',\n 'RXFP4',\n 'MAP1S',\n 'TPD52',\n 'TSPEAR',\n 'ZIC2',\n 'HOXD9',\n 'MTOR',\n 'MFSD1',\n 'EDN2',\n 'PMP2',\n 'GAB3',\n 'MRPL43',\n 'KLHL15',\n 'PLCXD2',\n 'HFE',\n 'PCGF5',\n 'PRSS36',\n 'DNPEP',\n 'SUN3',\n 'GANAB',\n 'EMC3',\n 'SHFL',\n 'ARPC2',\n 'RETSAT',\n 'CEP76',\n 'TESMIN',\n 'FADS2',\n 'IL17C',\n 'OSR1',\n 'ARHGAP36',\n 'PARP10',\n 'LCE1A',\n 'SIDT1',\n 'EMSY',\n 'TPRA1',\n 'HS2ST1',\n 'SLC16A6',\n 'ARPC3',\n 'PDE6H',\n 'DSP',\n 'DDX19B',\n 'SLC22A20P',\n 'CDX2',\n 'TBC1D22B',\n 'DHX40',\n 'ZNF184',\n 'PSMB7',\n 'EMCN',\n 'PEDS1',\n 'PTK7',\n 'DHRS3',\n 'HPGD',\n 'PDCD1LG2',\n 'ZNF318',\n 'MC1R',\n 'TMC8',\n 'CPTP',\n 'SLC25A1',\n 'TM6SF2',\n 'SNAPC4',\n 'SCARNA12',\n 'DUOX2',\n 'UNK',\n 'AKAP10',\n 'ADIG',\n 'TIPARP',\n 'ADAM8',\n 'RENBP',\n 'TLCD3B',\n 'MPPE1',\n 'PLEKHA5',\n 'TRX-CAT2-1',\n 'CKMT2',\n 'EMID1',\n 'UGDH',\n 'TFDP1',\n 'COPG2IT1',\n 'HHIPL2',\n 'COLQ',\n 'IFI44',\n 'GPR174',\n 'KERA',\n 'ARL13B',\n 'RTL10',\n 'RPS3A',\n 'SNX7',\n 'CEBPA-DT',\n 'REPS2',\n 'H2BC4',\n 'IL36B',\n 'PRMT3',\n 'PRSS50',\n 'IFT46',\n 'COPZ2',\n 'FAM90A14',\n 'DPYS',\n 'ROR1',\n 'GATA6',\n 'GLIS3',\n 'TM6SF1',\n 'UTP18',\n 'LLPH',\n 'FES',\n 'ODF2L',\n 'RALGAPB',\n 'ADAMTS3',\n 'NPC1L1',\n 'HSPA7',\n 'STBD1',\n 'GALC',\n 'GGA2',\n 'STAC',\n 'FKBP11',\n 'TNFSF15',\n 'REG1B',\n 'UGT2B7',\n 'RBP7',\n 'ADCY8',\n 'GIPR',\n 'SERPINB8',\n 'DACT2',\n 'FOXP2',\n 'UMOD',\n 'MIR130B',\n 'AGPAT4',\n 'OBI1',\n 'VSIG2',\n 'LSINCT5',\n 'JAM3',\n 'NGB',\n 'SIRT5',\n 'RAB12',\n 'DEFB4A',\n 'TMEM266',\n 'SMARCA2',\n 'PIGL',\n 'MRPS26',\n 'GALNT4',\n 'DRC1',\n 'HID1',\n 'MIR370',\n 'MX2',\n 'MAPRE1',\n 'TMEM67',\n 'CHRM3',\n 'ESM1',\n 'LINC02381',\n 'EIF3D',\n 'NTF4',\n 'UBE4A',\n 'AGO1',\n 'CHODL',\n 'MFSD2B',\n 'DDX23',\n 'NUFIP1',\n 'USP31',\n 'PHF14',\n 'METTL4',\n 'TBL1Y',\n 'ZFYVE19',\n 'EIF3F',\n 'TRP-AGG2-6',\n 'MRPL35',\n 'SMAD6',\n 'STAT2',\n 'ARVCF',\n 'KLHL41',\n 'LPIN1',\n 'FCRL1',\n 'ATE1',\n 'ZNF407',\n 'SLF1',\n 'DDIT4',\n 'CAPN2',\n 'TAGLN3',\n 'IARS1',\n 'STAB1',\n 'ARHGEF25',\n 'ATG13',\n 'STYK1',\n 'NGLY1',\n 'CHRNA6',\n 'NRXN3',\n 'PAIP1',\n 'PTPN14',\n 'ROCK1',\n 'OXA1L',\n 'MPPED1',\n 'DLL1',\n 'RARB',\n 'FOXL2',\n 'RUNX1',\n 'FHL5',\n 'TMC1',\n 'MIR521-1',\n 'NCF1',\n 'MT-ND3',\n 'ADGRA2',\n 'CLDN12',\n 'LCN8',\n 'MMP16',\n 'RBM46',\n 'MAGEB6',\n 'ZNF207',\n 'CLEC2A',\n 'ZNF341',\n 'CARD10',\n 'H1-6',\n 'OSTN',\n 'RTBDN',\n 'ENDOD1',\n 'BEX3',\n 'F9',\n 'ERVFRD-1',\n 'GUCY1B1',\n 'TMOD1',\n 'ZNF266',\n 'EDAR',\n 'LINC00673',\n 'APOM',\n 'LYZL6',\n 'ZFHX4',\n 'ST8SIA3',\n 'PPP1R42',\n 'REG4',\n 'SLC9A9',\n 'PLPP3',\n 'PRICKLE1',\n 'LSM14B',\n 'MPI',\n 'FBXO22',\n 'PEX3',\n 'ZFAND6',\n 'EEF1AKMT1',\n 'LINCMD1',\n 'C9',\n 'PSKH2',\n 'SUB1',\n 'PAN2',\n 'DCP1B',\n 'TMED2',\n 'GKAP1',\n 'NKAIN4',\n 'HSD17B6',\n 'POLK',\n 'ZNF443',\n 'MRPS25',\n 'MIR518D',\n 'CDK20',\n 'KIFAP3',\n 'RANBP10',\n 'TRAPPC14',\n 'TIMM23',\n 'GOLT1B',\n 'TIMP2',\n 'ZNF565',\n 'CCDC115',\n 'GPR52',\n 'HSH2D',\n 'FIP1L1',\n 'SLC49A3',\n 'ARMCX5',\n 'LGI2',\n 'CRH',\n 'CRTC2',\n 'CYB5A',\n 'LAMB4',\n 'ZNF141',\n 'MDGA2',\n 'FAM234B',\n 'SPINT1',\n 'GAB2',\n 'PDE11A',\n 'MAGEB18',\n 'SPRY1',\n 'PPP1R12A',\n 'CROT',\n 'TTC4',\n 'NLRP5',\n 'MUC22',\n 'COL18A1',\n 'ZBTB24',\n 'ACSM2B',\n 'MLLT11',\n 'ZBTB7A',\n 'LYPD6',\n 'TAAR8',\n 'CXCL8',\n 'ATP4A',\n 'IFNW1',\n 'PAAF1',\n 'CYP4F8',\n 'PER3',\n 'KLK15',\n 'MFAP3L',\n 'POP4',\n 'MAGEC2',\n 'NACA2',\n 'SLC22A7',\n 'MACO1',\n 'KMT2E',\n 'BAIAP2L2',\n 'DNAJC5B',\n 'IQGAP1',\n 'ZC3H12D',\n 'ATG4A',\n 'ARMC3',\n 'KLF7',\n 'MSX2',\n 'LLCFC1',\n 'LYZL4',\n 'CKAP5',\n 'TP53BP2',\n 'CEP68',\n 'ITK',\n 'CBR4',\n 'TLE3',\n 'SLC7A2',\n 'CDK12',\n 'GSTO1',\n 'ADCY1',\n 'RNF41',\n 'GZF1',\n 'GASK1A',\n 'DDX12P',\n 'PDIK1L',\n 'HSPA1A',\n 'OSTC',\n 'DBX1',\n 'GFM1',\n 'IGHG3',\n 'KRTAP1-1',\n 'PSMA7',\n 'HEYL',\n 'UPB1',\n 'AMIGO1',\n 'AURKA',\n 'GTF2I',\n 'PLK2',\n 'RIC8B',\n 'RASA2',\n 'TONSL',\n 'PLOD2',\n 'WASHC2C',\n 'CDH13',\n 'NDUFB6',\n 'MTMR4',\n 'MIR497',\n 'CSTF1',\n 'ZNF492',\n 'CIAO1',\n 'CARD17P',\n 'MFSD2A',\n 'CRTAP',\n 'BAG3',\n 'IL2RG',\n 'AP2A2',\n 'RAB39A',\n 'CIRBP',\n 'SLC25A30',\n 'TOGARAM1',\n 'TM7SF3',\n 'ATP2A3',\n 'ZBTB48',\n 'DYNC1LI1',\n 'ADAMDEC1',\n 'CDS1',\n 'PCDHAC2',\n 'MIR1909',\n 'BLMH',\n 'IL24',\n 'ISM1',\n 'NKX2-3',\n 'PRDX1',\n 'C8A',\n 'SESN1',\n 'HKDC1',\n 'AFG3L2',\n 'ENDOV',\n 'PROM1',\n 'DRG1',\n 'H2AB2',\n 'CCT5',\n 'DNAI3',\n 'CLPS',\n 'THG1L',\n 'PCGF3',\n 'FFAR1',\n 'FRMPD3',\n 'ADAD1',\n 'MIR218-2',\n 'LRRC3',\n 'CYP11A1',\n 'BCDIN3D',\n 'MANF',\n 'MIR802',\n 'SLC17A4',\n 'TRIM73',\n 'OSBPL10',\n 'NSUN7',\n 'MRPL49',\n 'KLHL20',\n 'TRG-GCC2-6',\n 'RBM14',\n 'PEX6',\n 'WASL',\n 'PROZ',\n 'LRIT1',\n 'CPLANE1',\n 'GRB14',\n 'CCNA1',\n 'FGA',\n 'TNFRSF11B',\n 'AXIN2',\n 'HAPLN2',\n 'ELOA',\n 'PRODH2',\n 'ZNF827',\n 'TIMM17A',\n 'MIR551A',\n 'CDADC1',\n 'ODAD4',\n 'TUBGCP2',\n 'KRR1',\n 'FIBP',\n 'SEA',\n 'CD93',\n 'BTN3A1',\n 'BIRC2',\n 'H3C12',\n 'SNTN',\n 'NRTN',\n 'MDM2',\n 'YBX2',\n 'KRTAP1-3',\n 'NPVF',\n 'GARIN5A',\n 'GARIN1A',\n 'C2',\n 'PUDP',\n 'THRSP',\n 'DMAC2',\n 'LGALS13',\n 'GAP43',\n 'LILRB1',\n 'TRAPPC3L',\n 'DZIP3',\n 'SLC10A6',\n 'RTF1',\n 'VCAN',\n 'PGAM4',\n 'PAFAH1B2',\n 'TRIM61',\n 'TRMT10A',\n 'CRTAM',\n 'TRX-CAT1-2',\n 'PABPC4',\n 'CFAP251',\n 'RPP30',\n 'FGFR1OP2',\n 'LAT',\n 'N6AMT1',\n 'RUNDC1',\n 'UBL3',\n 'BSCL2',\n 'NMNAT2',\n 'ETF1',\n 'RAB6A',\n 'ZBED5',\n 'CCL20',\n 'FXN',\n 'PPM1H',\n 'MATCAP1',\n 'NAPB',\n 'ACTR10',\n 'SLC39A5',\n 'ELOVL1',\n 'CD3G',\n 'CLCN4',\n 'MAN2C1',\n 'DNAI1',\n 'ONECUT3',\n 'RPLP1',\n 'COA8',\n 'SFTPC',\n 'POP5',\n 'CSNK1G2',\n 'MRGPRF',\n 'MT-TS2',\n 'H2AC20',\n 'OSMR',\n 'ARHGEF18',\n 'KLHL6',\n 'MRPS22',\n 'VPS51',\n 'FAR2',\n 'LGR5',\n 'LCAT',\n 'ELMO1',\n 'NPNT',\n 'DSCAM',\n 'SPOCK2',\n 'C1QL1',\n 'BPGM',\n 'CASC2',\n 'KRT82',\n 'TAB1',\n 'SAMD9L',\n 'NELFA',\n 'CNOT7',\n 'CD300C',\n 'LYPD1',\n 'TRPC5',\n 'DOK2',\n 'UBE2Z',\n 'PSAT1',\n 'INSYN2A',\n 'USF2',\n 'CLEC4M',\n 'PHOX2A',\n 'SEPHS2',\n 'CRMP1',\n 'JPH2',\n 'PC',\n 'SPTSSB',\n 'RBFOX1',\n 'ERRFI1',\n 'LRATD1',\n 'TMTC4',\n 'CST9L',\n 'AKNA',\n 'KIF12',\n 'PDE10A',\n 'AGTPBP1',\n 'HPV18I1',\n 'MIR503',\n 'RPS27A',\n 'COL4A5',\n 'PKD1',\n 'HTRA1',\n 'FUT10',\n 'MEGF8',\n 'PSIP1',\n 'PSMA8',\n 'PLS1',\n 'EIF1AD',\n 'IDH3A',\n 'TMSB4X',\n 'ZNF33B',\n 'MTIF3',\n 'SLC6A14',\n 'CAMK2N2',\n 'HOXA9',\n 'DIS3L',\n 'DNAH2',\n 'RBBP5',\n 'RBM7',\n 'ANXA2',\n 'MIB1',\n 'DNASE1L1',\n 'HCAR1',\n 'POLR1H',\n 'RUSC2',\n 'SEPHS1',\n 'MAP3K13',\n 'BAIAP2L1',\n 'CAPRIN1',\n 'IQCB1',\n 'CDH22',\n 'EHBP1L1',\n 'KLF3',\n 'MIR139',\n 'GHRH',\n 'SSX2IP',\n 'AGR2',\n 'GAL3ST4',\n 'MIR20A',\n 'CDV3',\n 'BLNK',\n 'IL27RA',\n 'SIM2',\n 'SRGAP2D',\n 'CIR1',\n 'PAPOLG',\n 'MTBP',\n 'FAAP20',\n 'PLA2G4C',\n 'IFI27L1',\n 'DUSP13B',\n 'CCNQ',\n 'ZFP36L1',\n 'SPICE1',\n 'LCE5A',\n 'C5AR1',\n 'SPACA1',\n 'URGCP',\n 'DENND1A',\n 'ZNF397',\n 'TBXT',\n 'RAB4A',\n 'ZNF92',\n 'NXF2',\n 'SPC24',\n 'DIP2C',\n 'RIMS4',\n 'CRIP1',\n 'STING1',\n 'PARD3B',\n 'ADAM30',\n 'FLCN',\n 'ZNF490',\n 'NUP85',\n 'CDK5R2',\n 'NOL4',\n 'NCAPD3',\n 'DHRS4L2',\n 'COL5A1',\n 'MPHOSPH8',\n 'KANSL2',\n 'GTF2F2',\n 'LUC7L3',\n 'PRL',\n 'SF3A2',\n 'KPRP',\n 'FAM171B',\n 'IL1RL1',\n 'PPP1R12B',\n 'EPC2',\n 'SUN5',\n 'EBVS1',\n 'ATP1A2',\n 'CCDC25',\n 'MINPP1',\n 'WWC2-AS1',\n 'SLC52A3',\n 'CPN1',\n 'ACADVL',\n 'ZNF507',\n 'PNMA5',\n 'BEX5',\n 'DIO1',\n 'ZNF274',\n 'SCMH1',\n 'GDPGP1',\n 'CD68',\n 'ICAM5',\n 'PLOD3',\n 'CHAF1B',\n 'TMED1',\n 'ABI3',\n 'RNF185',\n 'KRT33A',\n 'ZNF213',\n 'HCRT',\n 'TTLL5',\n 'LCE3E',\n 'SHB',\n 'GTF3C1',\n 'TBC1D20',\n 'SSBP2',\n 'SUPT3H',\n 'ADCYAP1',\n 'UBE2L1',\n 'TOR1B',\n 'CUL9',\n 'NT5C2',\n 'PKN2',\n 'SPATA19',\n 'PRMT7',\n 'RPL15',\n 'GLYR1',\n 'CMPK2',\n 'AKAIN1',\n 'CASZ1',\n 'BTNL3',\n 'DPYSL3',\n 'MIR365A',\n 'P2RX5',\n 'LSM6',\n 'BRD1',\n 'NPHP3',\n 'VPREB1',\n 'SLITRK3',\n 'PRICKLE3',\n 'OR2M7',\n 'VSTM1',\n 'TMEM123',\n 'ACYP2',\n 'CCN5',\n 'SERPINA4',\n 'ZNF536',\n 'AQP10',\n 'DAGLB',\n 'H19',\n 'NPFF',\n 'BICRAL',\n 'KLHL21',\n 'TXLNG',\n 'KCNH6',\n 'CD40LG',\n 'RANBP3',\n 'TRBC2',\n 'PFAS',\n 'MYO5B',\n 'POLA1',\n 'WDR87',\n 'SURF1',\n 'MCL1',\n 'GALK2',\n 'CD7',\n 'DGKQ',\n 'IL6R',\n 'CD248',\n 'TTC21B',\n 'ZNF71',\n 'AP2A1',\n 'MASP1',\n 'CCL28',\n 'CEP192',\n 'TP53TG1',\n 'NR1D2',\n 'DDO',\n 'GSKIP',\n 'MTHFR',\n 'SLC7A10',\n 'TM4SF20',\n 'TEX261',\n 'MRPL14',\n 'OR2J3',\n 'PRPSAP2',\n 'TRPC1',\n 'XPO4',\n 'TBC1D4',\n 'ZNF436',\n 'TTI1',\n 'PIGF',\n 'PRKDC',\n 'CHST6',\n 'CEP70',\n 'SLC12A9',\n 'CALHM6',\n 'GEMIN8',\n 'PDCD2',\n 'CEMIP',\n 'CPB1',\n 'CALB1',\n 'CDCA5',\n 'SPRR2D',\n 'CXCL13',\n 'SCLY',\n 'PCDHB4',\n 'MRPS18A',\n 'CETN3',\n 'ECSCR',\n 'ETV1',\n 'AIRE',\n 'MRPL3',\n 'BTBD9',\n 'PEX19',\n 'ZBTB2',\n 'TAS1R3',\n 'INPP4B',\n 'SSB',\n 'SLC30A9',\n 'DPYSL4',\n 'BRCC3',\n 'MT-TM',\n 'HAX1',\n 'STK35',\n 'GDF3',\n 'VPS26B',\n 'BATF',\n 'BCAR4',\n 'PYDC2',\n 'AEN',\n 'NAP1L3',\n 'PHACTR2',\n 'DMP1',\n 'NDUFV1',\n 'TXLNGY',\n 'CFH',\n 'ATF3',\n 'SMIM3',\n 'NAPEPLD',\n 'SNORA2C',\n 'GOLGA7B',\n 'TSSK3',\n 'CCDC141',\n 'KIF25',\n 'EDEM3',\n 'RALA',\n 'SPOCD1',\n 'SS18L2',\n 'GPR33',\n 'RNMT',\n 'UGT8',\n 'GOLGB1',\n 'F2RL3',\n 'SUSD2',\n 'TFAP2A-AS2',\n 'ASB10',\n 'OR5F1',\n 'HTR2C',\n 'YY2',\n 'ADCYAP1R1',\n 'TH',\n 'VPS13C',\n 'MAST4',\n 'UTP14C',\n 'SPDYE2',\n 'ATP11A',\n 'REV1',\n 'MYG1',\n 'CHD8',\n 'PTCSC3',\n 'STIP1',\n 'LRRC32',\n 'NELL1',\n 'GLRX3',\n 'RAB40C',\n 'MID1IP1',\n 'LYNX1',\n 'SCARNA10',\n 'DISP1',\n 'IPPK',\n 'NPIPA1',\n 'NANOS3',\n 'ATOH7',\n 'TXNL4B',\n 'PPP6R1',\n 'C6orf15',\n 'PTPN11',\n 'SWI5',\n 'PCDHB15',\n 'MXD4',\n 'GPX2',\n 'CASKIN2',\n 'ALMS1',\n 'CLMP',\n 'MT-TH',\n 'FCER1G',\n 'LASP1',\n 'OBP2B',\n 'SPATA25',\n 'IFNAR1',\n 'TNFRSF8',\n 'FOXP4',\n 'PPP1R3F',\n 'HGD',\n 'OR1A2',\n 'TRPC4',\n 'AVEN',\n 'ACSM1',\n 'MAP2K2',\n 'BCL2L12',\n 'CACNG7',\n 'PSMB9',\n 'SLC1A5',\n 'CC2D1A',\n 'KRT9',\n 'CUL5',\n 'LAMA2',\n 'PTCH1',\n 'LMAN2L',\n 'CIB3',\n 'UQCC3',\n 'NR5A1',\n 'PYROXD1',\n 'MIR489',\n 'ANKRD24',\n 'AMBRA1',\n 'UBA3',\n 'ADISSP',\n 'CD83',\n 'RAP2C',\n 'AP1S1',\n 'HPS3',\n 'RHEB',\n 'CEP290',\n 'LYRM4',\n 'GNE',\n 'ST7',\n 'ALPG',\n 'TNFAIP8L1',\n 'AGO3',\n 'PLAC4',\n 'HELZ2',\n 'SYPL1',\n 'EMD',\n 'EXTL1',\n 'FBXO46',\n 'ABCC5',\n 'CR1',\n 'APOO',\n 'NCAPG2',\n 'PATJ',\n 'CLCN2',\n 'TRIB1',\n 'ACRV1',\n 'HOXB4',\n 'MAZ',\n ...}"
     },
     "execution_count": 168,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "omim_gene = set(raw_omim_df['Approved Gene Symbol (HGNC)'].dropna())\n",
    "# with open(omim_gene_list_file) as f:\n",
    "#     omim_gene = set(f.read().split('\\n'))\n",
    "omim_gene"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2024-05-11T06:00:52.336176300Z",
     "start_time": "2024-05-11T06:00:52.282850900Z"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 169,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-05-11T06:00:52.469095900Z",
     "start_time": "2024-05-11T06:00:52.323162200Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": "      chrom  txStart   txEnd  GeneID   name2          name strand\n9159   chr1   859302  879954  148398  SAMD11  NM_001385641      +\n10490  chr1   879582  894636   26155   NOC2L     NM_015658      -\n14731  chr1   895963  901099  339451  KLHL17     NM_198317      +\n14417  chr1   910577  917473   84808   PERM1  NM_001291367      -\n18199  chr1   934343  935477   57801    HES4  NM_001142467      -",
      "text/html": "<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>chrom</th>\n      <th>txStart</th>\n      <th>txEnd</th>\n      <th>GeneID</th>\n      <th>name2</th>\n      <th>name</th>\n      <th>strand</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>9159</th>\n      <td>chr1</td>\n      <td>859302</td>\n      <td>879954</td>\n      <td>148398</td>\n      <td>SAMD11</td>\n      <td>NM_001385641</td>\n      <td>+</td>\n    </tr>\n    <tr>\n      <th>10490</th>\n      <td>chr1</td>\n      <td>879582</td>\n      <td>894636</td>\n      <td>26155</td>\n      <td>NOC2L</td>\n      <td>NM_015658</td>\n      <td>-</td>\n    </tr>\n    <tr>\n      <th>14731</th>\n      <td>chr1</td>\n      <td>895963</td>\n      <td>901099</td>\n      <td>339451</td>\n      <td>KLHL17</td>\n      <td>NM_198317</td>\n      <td>+</td>\n    </tr>\n    <tr>\n      <th>14417</th>\n      <td>chr1</td>\n      <td>910577</td>\n      <td>917473</td>\n      <td>84808</td>\n      <td>PERM1</td>\n      <td>NM_001291367</td>\n      <td>-</td>\n    </tr>\n    <tr>\n      <th>18199</th>\n      <td>chr1</td>\n      <td>934343</td>\n      <td>935477</td>\n      <td>57801</td>\n      <td>HES4</td>\n      <td>NM_001142467</td>\n      <td>-</td>\n    </tr>\n  </tbody>\n</table>\n</div>"
     },
     "execution_count": 169,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "omim_gene = refgene_info.loc[\n",
    "    refgene_info['name2'].isin(omim_gene),\n",
    "    ['chrom', 'txStart', 'txEnd', 'GeneID', 'name2', 'name', 'strand']\n",
    "].sort_values(['chrom', 'txStart', 'txEnd'])\n",
    "omim_gene.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 170,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-05-11T06:00:52.469095900Z",
     "start_time": "2024-05-11T06:00:52.427098300Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/staging/software/lib/python3.9/site-packages/autocnv-1-py3.9.egg/data/omim-gene.sorted.bed\n"
     ]
    }
   ],
   "source": [
    "omim_gene.rename(columns={\n",
    "    'chrom': '#chrom', 'txStart': 'start', 'txEnd': 'end',\n",
    "    'GeneID': 'gene_id', 'name2': 'symbol', 'name': 'transcript'\n",
    "}).to_csv(omim_gene_file, index=False, sep='\\t')\n",
    "print(omim_gene_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 171,
   "outputs": [],
   "source": [
    "!bgzip -cf {omim_gene_file} > {omim_gene_file}.gz\n",
    "!tabix -fp bed {omim_gene_file}.gz\n",
    "!rm {omim_gene_file}"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    },
    "ExecuteTime": {
     "end_time": "2024-05-11T06:00:54.624682700Z",
     "start_time": "2024-05-11T06:00:52.529935600Z"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 172,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-05-11T06:00:54.682811900Z",
     "start_time": "2024-05-11T06:00:54.631666600Z"
    }
   },
   "outputs": [],
   "source": [
    "curation_gene = pd.read_csv(clingen_gene_ori_file, sep='\\t', dtype=str, skiprows=5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 173,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-05-11T06:00:54.706987900Z",
     "start_time": "2024-05-11T06:00:54.668848900Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": "      chrom   txStart     txEnd  GeneID   name2          name strand\n51     chr1   6845513   7829766   23261  CAMTA1  NM_001349609      +\n6453   chr1  17345220  17380527    6390    SDHB     NM_003000      -\n3053   chr1  27022505  27108595    8289  ARID1A     NM_139135      +\n14538  chr1  27860755  27930655   27245   AHDC1  NM_001371928      -\n6966   chr1  43391025  43424539    6513  SLC2A1     NM_006516      -",
      "text/html": "<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>chrom</th>\n      <th>txStart</th>\n      <th>txEnd</th>\n      <th>GeneID</th>\n      <th>name2</th>\n      <th>name</th>\n      <th>strand</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>51</th>\n      <td>chr1</td>\n      <td>6845513</td>\n      <td>7829766</td>\n      <td>23261</td>\n      <td>CAMTA1</td>\n      <td>NM_001349609</td>\n      <td>+</td>\n    </tr>\n    <tr>\n      <th>6453</th>\n      <td>chr1</td>\n      <td>17345220</td>\n      <td>17380527</td>\n      <td>6390</td>\n      <td>SDHB</td>\n      <td>NM_003000</td>\n      <td>-</td>\n    </tr>\n    <tr>\n      <th>3053</th>\n      <td>chr1</td>\n      <td>27022505</td>\n      <td>27108595</td>\n      <td>8289</td>\n      <td>ARID1A</td>\n      <td>NM_139135</td>\n      <td>+</td>\n    </tr>\n    <tr>\n      <th>14538</th>\n      <td>chr1</td>\n      <td>27860755</td>\n      <td>27930655</td>\n      <td>27245</td>\n      <td>AHDC1</td>\n      <td>NM_001371928</td>\n      <td>-</td>\n    </tr>\n    <tr>\n      <th>6966</th>\n      <td>chr1</td>\n      <td>43391025</td>\n      <td>43424539</td>\n      <td>6513</td>\n      <td>SLC2A1</td>\n      <td>NM_006516</td>\n      <td>-</td>\n    </tr>\n  </tbody>\n</table>\n</div>"
     },
     "execution_count": 173,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hi_genes = set(\n",
    "    curation_gene.loc[\n",
    "        curation_gene['Haploinsufficiency Score'] == '3', '#Gene Symbol'\n",
    "    ]\n",
    ")\n",
    "hi_gene = refgene_info.loc[\n",
    "    refgene_info['name2'].isin(hi_genes),\n",
    "    ['chrom', 'txStart', 'txEnd', 'GeneID', 'name2', 'name', 'strand']\n",
    "].sort_values(['chrom', 'txStart', 'txEnd'])\n",
    "hi_gene.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 174,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-05-11T06:00:54.754114500Z",
     "start_time": "2024-05-11T06:00:54.710741900Z"
    }
   },
   "outputs": [],
   "source": [
    "hi_gene.rename(columns={\n",
    "    'chrom': '#chrom', 'txStart': 'start', 'txEnd': 'end',\n",
    "    'GeneID': 'gene_id', 'name2': 'symbol', 'name': 'transcript'\n",
    "}).to_csv(hi_gene_file, index=False, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 175,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-05-11T06:00:56.758069900Z",
     "start_time": "2024-05-11T06:00:54.723706400Z"
    }
   },
   "outputs": [],
   "source": [
    "!bgzip -cf {hi_gene_file} > {hi_gene_file}.gz\n",
    "!tabix -fp bed {hi_gene_file}.gz\n",
    "!rm {hi_gene_file}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 176,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-05-11T06:00:56.814597400Z",
     "start_time": "2024-05-11T06:00:56.775834300Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": "      chrom  cdsStart    cdsEnd  GeneID   name2          name\n51     chr1   6845590   7826582   23261  CAMTA1  NM_001349609\n6453   chr1  17345375  17380514    6390    SDHB     NM_003000\n3053   chr1  27022894  27107247    8289  ARID1A     NM_139135\n14538  chr1  27873814  27878626   27245   AHDC1  NM_001371928\n6966   chr1  43392711  43424322    6513  SLC2A1     NM_006516",
      "text/html": "<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>chrom</th>\n      <th>cdsStart</th>\n      <th>cdsEnd</th>\n      <th>GeneID</th>\n      <th>name2</th>\n      <th>name</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>51</th>\n      <td>chr1</td>\n      <td>6845590</td>\n      <td>7826582</td>\n      <td>23261</td>\n      <td>CAMTA1</td>\n      <td>NM_001349609</td>\n    </tr>\n    <tr>\n      <th>6453</th>\n      <td>chr1</td>\n      <td>17345375</td>\n      <td>17380514</td>\n      <td>6390</td>\n      <td>SDHB</td>\n      <td>NM_003000</td>\n    </tr>\n    <tr>\n      <th>3053</th>\n      <td>chr1</td>\n      <td>27022894</td>\n      <td>27107247</td>\n      <td>8289</td>\n      <td>ARID1A</td>\n      <td>NM_139135</td>\n    </tr>\n    <tr>\n      <th>14538</th>\n      <td>chr1</td>\n      <td>27873814</td>\n      <td>27878626</td>\n      <td>27245</td>\n      <td>AHDC1</td>\n      <td>NM_001371928</td>\n    </tr>\n    <tr>\n      <th>6966</th>\n      <td>chr1</td>\n      <td>43392711</td>\n      <td>43424322</td>\n      <td>6513</td>\n      <td>SLC2A1</td>\n      <td>NM_006516</td>\n    </tr>\n  </tbody>\n</table>\n</div>"
     },
     "execution_count": 176,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hi_cds = refgene_info.loc[\n",
    "    (refgene_info['name2'].isin(hi_genes)) & (refgene_info['length'] != 0),\n",
    "    ['chrom', 'cdsStart', 'cdsEnd', 'GeneID', 'name2', 'name']\n",
    "].sort_values(['chrom', 'cdsStart', 'cdsEnd'])\n",
    "hi_cds.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 177,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-05-11T06:00:56.814597400Z",
     "start_time": "2024-05-11T06:00:56.811605700Z"
    }
   },
   "outputs": [],
   "source": [
    "hi_cds.rename(columns={\n",
    "    'chrom': '#chrom', 'cdsStart': 'start', 'cdsEnd': 'end',\n",
    "    'GeneID': 'gene_id', 'name2': 'symbol', 'name': 'transcript'\n",
    "}).to_csv(hi_cds_file, index=False, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 178,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-05-11T06:00:58.864513600Z",
     "start_time": "2024-05-11T06:00:56.820581600Z"
    }
   },
   "outputs": [],
   "source": [
    "!bgzip -cf {hi_cds_file} > {hi_cds_file}.gz\n",
    "!tabix -fp bed {hi_cds_file}.gz\n",
    "!rm {hi_cds_file}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 179,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-05-11T06:00:58.918661900Z",
     "start_time": "2024-05-11T06:00:58.896720500Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": "   chrom                                         exonStarts  \\\n2   chrX  31137340,31144758,31152218,31164407,31165391,3...   \n26  chr7  69063460,69364271,69583117,69599521,69900737,7...   \n31  chrX  28605562,28807436,29301054,29414374,29417271,2...   \n44  chr2  50145642,50170841,50280408,50282092,50318460,5...   \n51  chr1  6845513,6880240,6885151,7151363,7309550,752788...   \n\n                                             exonEnds  GeneID     name2  \\\n2   31140047,31144790,31152311,31164531,31165635,3...    1756       DMD   \n26  69064948,69364484,69583219,69599557,69900767,7...   26053     AUTS2   \n31  28606164,28807542,29301334,29414561,29417425,2...   11141  IL1RAPL1   \n44  50149389,50170929,50280728,50282182,50318632,5...    9378     NRXN1   \n51  6845635,6880310,6885270,7151431,7309686,752796...   23261    CAMTA1   \n\n            name strand  \n2      NM_000109      -  \n26     NM_015570      +  \n31     NM_014271      +  \n44  NM_001330084      -  \n51  NM_001349609      +  ",
      "text/html": "<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>chrom</th>\n      <th>exonStarts</th>\n      <th>exonEnds</th>\n      <th>GeneID</th>\n      <th>name2</th>\n      <th>name</th>\n      <th>strand</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>2</th>\n      <td>chrX</td>\n      <td>31137340,31144758,31152218,31164407,31165391,3...</td>\n      <td>31140047,31144790,31152311,31164531,31165635,3...</td>\n      <td>1756</td>\n      <td>DMD</td>\n      <td>NM_000109</td>\n      <td>-</td>\n    </tr>\n    <tr>\n      <th>26</th>\n      <td>chr7</td>\n      <td>69063460,69364271,69583117,69599521,69900737,7...</td>\n      <td>69064948,69364484,69583219,69599557,69900767,7...</td>\n      <td>26053</td>\n      <td>AUTS2</td>\n      <td>NM_015570</td>\n      <td>+</td>\n    </tr>\n    <tr>\n      <th>31</th>\n      <td>chrX</td>\n      <td>28605562,28807436,29301054,29414374,29417271,2...</td>\n      <td>28606164,28807542,29301334,29414561,29417425,2...</td>\n      <td>11141</td>\n      <td>IL1RAPL1</td>\n      <td>NM_014271</td>\n      <td>+</td>\n    </tr>\n    <tr>\n      <th>44</th>\n      <td>chr2</td>\n      <td>50145642,50170841,50280408,50282092,50318460,5...</td>\n      <td>50149389,50170929,50280728,50282182,50318632,5...</td>\n      <td>9378</td>\n      <td>NRXN1</td>\n      <td>NM_001330084</td>\n      <td>-</td>\n    </tr>\n    <tr>\n      <th>51</th>\n      <td>chr1</td>\n      <td>6845513,6880240,6885151,7151363,7309550,752788...</td>\n      <td>6845635,6880310,6885270,7151431,7309686,752796...</td>\n      <td>23261</td>\n      <td>CAMTA1</td>\n      <td>NM_001349609</td>\n      <td>+</td>\n    </tr>\n  </tbody>\n</table>\n</div>"
     },
     "execution_count": 179,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hi_exons = refgene_info.loc[\n",
    "    refgene_info['name2'].isin(hi_genes), ['chrom', 'exonStarts', 'exonEnds', 'GeneID', 'name2', 'name', 'strand']\n",
    "].copy()\n",
    "hi_exons.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 180,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-05-11T06:00:58.934703600Z",
     "start_time": "2024-05-11T06:00:58.916667400Z"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_136260/867614133.py:1: FutureWarning: The default value of regex will change from True to False in a future version.\n",
      "  hi_exons['exonStarts'] = hi_exons['exonStarts'].str.replace(r',$', '')\n",
      "/tmp/ipykernel_136260/867614133.py:2: FutureWarning: The default value of regex will change from True to False in a future version.\n",
      "  hi_exons['exonEnds'] = hi_exons['exonEnds'].str.replace(r',$', '')\n"
     ]
    },
    {
     "data": {
      "text/plain": "   chrom                                         exonStarts  \\\n2   chrX  31137340,31144758,31152218,31164407,31165391,3...   \n26  chr7  69063460,69364271,69583117,69599521,69900737,7...   \n31  chrX  28605562,28807436,29301054,29414374,29417271,2...   \n44  chr2  50145642,50170841,50280408,50282092,50318460,5...   \n51  chr1  6845513,6880240,6885151,7151363,7309550,752788...   \n\n                                             exonEnds  GeneID     name2  \\\n2   31140047,31144790,31152311,31164531,31165635,3...    1756       DMD   \n26  69064948,69364484,69583219,69599557,69900767,7...   26053     AUTS2   \n31  28606164,28807542,29301334,29414561,29417425,2...   11141  IL1RAPL1   \n44  50149389,50170929,50280728,50282182,50318632,5...    9378     NRXN1   \n51  6845635,6880310,6885270,7151431,7309686,752796...   23261    CAMTA1   \n\n            name strand  \n2      NM_000109      -  \n26     NM_015570      +  \n31     NM_014271      +  \n44  NM_001330084      -  \n51  NM_001349609      +  ",
      "text/html": "<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>chrom</th>\n      <th>exonStarts</th>\n      <th>exonEnds</th>\n      <th>GeneID</th>\n      <th>name2</th>\n      <th>name</th>\n      <th>strand</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>2</th>\n      <td>chrX</td>\n      <td>31137340,31144758,31152218,31164407,31165391,3...</td>\n      <td>31140047,31144790,31152311,31164531,31165635,3...</td>\n      <td>1756</td>\n      <td>DMD</td>\n      <td>NM_000109</td>\n      <td>-</td>\n    </tr>\n    <tr>\n      <th>26</th>\n      <td>chr7</td>\n      <td>69063460,69364271,69583117,69599521,69900737,7...</td>\n      <td>69064948,69364484,69583219,69599557,69900767,7...</td>\n      <td>26053</td>\n      <td>AUTS2</td>\n      <td>NM_015570</td>\n      <td>+</td>\n    </tr>\n    <tr>\n      <th>31</th>\n      <td>chrX</td>\n      <td>28605562,28807436,29301054,29414374,29417271,2...</td>\n      <td>28606164,28807542,29301334,29414561,29417425,2...</td>\n      <td>11141</td>\n      <td>IL1RAPL1</td>\n      <td>NM_014271</td>\n      <td>+</td>\n    </tr>\n    <tr>\n      <th>44</th>\n      <td>chr2</td>\n      <td>50145642,50170841,50280408,50282092,50318460,5...</td>\n      <td>50149389,50170929,50280728,50282182,50318632,5...</td>\n      <td>9378</td>\n      <td>NRXN1</td>\n      <td>NM_001330084</td>\n      <td>-</td>\n    </tr>\n    <tr>\n      <th>51</th>\n      <td>chr1</td>\n      <td>6845513,6880240,6885151,7151363,7309550,752788...</td>\n      <td>6845635,6880310,6885270,7151431,7309686,752796...</td>\n      <td>23261</td>\n      <td>CAMTA1</td>\n      <td>NM_001349609</td>\n      <td>+</td>\n    </tr>\n  </tbody>\n</table>\n</div>"
     },
     "execution_count": 180,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hi_exons['exonStarts'] = hi_exons['exonStarts'].str.replace(r',$', '')\n",
    "hi_exons['exonEnds'] = hi_exons['exonEnds'].str.replace(r',$', '')\n",
    "hi_exons.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 181,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-05-11T06:00:59.117462500Z",
     "start_time": "2024-05-11T06:00:58.973599600Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": "   row     start       end\n0    2  31137340  31140047\n1    2  31144758  31144790\n2    2  31152218  31152311\n3    2  31164407  31164531\n4    2  31165391  31165635",
      "text/html": "<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>row</th>\n      <th>start</th>\n      <th>end</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>0</th>\n      <td>2</td>\n      <td>31137340</td>\n      <td>31140047</td>\n    </tr>\n    <tr>\n      <th>1</th>\n      <td>2</td>\n      <td>31144758</td>\n      <td>31144790</td>\n    </tr>\n    <tr>\n      <th>2</th>\n      <td>2</td>\n      <td>31152218</td>\n      <td>31152311</td>\n    </tr>\n    <tr>\n      <th>3</th>\n      <td>2</td>\n      <td>31164407</td>\n      <td>31164531</td>\n    </tr>\n    <tr>\n      <th>4</th>\n      <td>2</td>\n      <td>31165391</td>\n      <td>31165635</td>\n    </tr>\n  </tbody>\n</table>\n</div>"
     },
     "execution_count": 181,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "start = hi_exons['exonStarts'].str.split(',').apply(pd.Series).stack().reset_index()\n",
    "start = start.rename(columns={'level_0': 'row', 0: 'start'})[['row', 'start']]\n",
    "start['start'] = start['start'].astype(int)\n",
    "end = hi_exons['exonEnds'].str.split(',').apply(pd.Series).stack().reset_index()\n",
    "end = end.rename(columns={0: 'end'})['end'].astype(int)\n",
    "position = start.join(end)\n",
    "position.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 182,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-05-11T06:00:59.222205800Z",
     "start_time": "2024-05-11T06:00:59.114472Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": "     row    start      end chrom  GeneID   name2          name strand\n128   51  6845513  6845635  chr1   23261  CAMTA1  NM_001349609      +\n129   51  6880240  6880310  chr1   23261  CAMTA1  NM_001349609      +\n130   51  6885151  6885270  chr1   23261  CAMTA1  NM_001349609      +\n131   51  7151363  7151431  chr1   23261  CAMTA1  NM_001349609      +\n132   51  7309550  7309686  chr1   23261  CAMTA1  NM_001349609      +",
      "text/html": "<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>row</th>\n      <th>start</th>\n      <th>end</th>\n      <th>chrom</th>\n      <th>GeneID</th>\n      <th>name2</th>\n      <th>name</th>\n      <th>strand</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>128</th>\n      <td>51</td>\n      <td>6845513</td>\n      <td>6845635</td>\n      <td>chr1</td>\n      <td>23261</td>\n      <td>CAMTA1</td>\n      <td>NM_001349609</td>\n      <td>+</td>\n    </tr>\n    <tr>\n      <th>129</th>\n      <td>51</td>\n      <td>6880240</td>\n      <td>6880310</td>\n      <td>chr1</td>\n      <td>23261</td>\n      <td>CAMTA1</td>\n      <td>NM_001349609</td>\n      <td>+</td>\n    </tr>\n    <tr>\n      <th>130</th>\n      <td>51</td>\n      <td>6885151</td>\n      <td>6885270</td>\n      <td>chr1</td>\n      <td>23261</td>\n      <td>CAMTA1</td>\n      <td>NM_001349609</td>\n      <td>+</td>\n    </tr>\n    <tr>\n      <th>131</th>\n      <td>51</td>\n      <td>7151363</td>\n      <td>7151431</td>\n      <td>chr1</td>\n      <td>23261</td>\n      <td>CAMTA1</td>\n      <td>NM_001349609</td>\n      <td>+</td>\n    </tr>\n    <tr>\n      <th>132</th>\n      <td>51</td>\n      <td>7309550</td>\n      <td>7309686</td>\n      <td>chr1</td>\n      <td>23261</td>\n      <td>CAMTA1</td>\n      <td>NM_001349609</td>\n      <td>+</td>\n    </tr>\n  </tbody>\n</table>\n</div>"
     },
     "execution_count": 182,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exon = position.merge(\n",
    "    hi_exons[['chrom', 'GeneID', 'name2', 'name', 'strand']], how='left', left_on='row', right_index=True\n",
    ")\n",
    "exon = exon.sort_values(['chrom', 'start', 'end'])\n",
    "exon.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 183,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-05-11T06:00:59.250077Z",
     "start_time": "2024-05-11T06:00:59.154403400Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": "        row      start        end chrom  GeneID   name2       name strand   +  \\\n1600   1189  154227753  154227875  chrX    2157      F8  NM_000132      -  25   \n1601   1189  154250684  154250998  chrX    2157      F8  NM_000132      -  26   \n5975  15336  154487519  154490514  chrX  116442  RAB39B  NM_171998      -   1   \n5976  15336  154493358  154493776  chrX  116442  RAB39B  NM_171998      -   2   \n6032  18569    2654895    2655723  chrY    6736     SRY  NM_003140      -   1   \n\n      -  \n1600  2  \n1601  1  \n5975  2  \n5976  1  \n6032  1  ",
      "text/html": "<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>row</th>\n      <th>start</th>\n      <th>end</th>\n      <th>chrom</th>\n      <th>GeneID</th>\n      <th>name2</th>\n      <th>name</th>\n      <th>strand</th>\n      <th>+</th>\n      <th>-</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>1600</th>\n      <td>1189</td>\n      <td>154227753</td>\n      <td>154227875</td>\n      <td>chrX</td>\n      <td>2157</td>\n      <td>F8</td>\n      <td>NM_000132</td>\n      <td>-</td>\n      <td>25</td>\n      <td>2</td>\n    </tr>\n    <tr>\n      <th>1601</th>\n      <td>1189</td>\n      <td>154250684</td>\n      <td>154250998</td>\n      <td>chrX</td>\n      <td>2157</td>\n      <td>F8</td>\n      <td>NM_000132</td>\n      <td>-</td>\n      <td>26</td>\n      <td>1</td>\n    </tr>\n    <tr>\n      <th>5975</th>\n      <td>15336</td>\n      <td>154487519</td>\n      <td>154490514</td>\n      <td>chrX</td>\n      <td>116442</td>\n      <td>RAB39B</td>\n      <td>NM_171998</td>\n      <td>-</td>\n      <td>1</td>\n      <td>2</td>\n    </tr>\n    <tr>\n      <th>5976</th>\n      <td>15336</td>\n      <td>154493358</td>\n      <td>154493776</td>\n      <td>chrX</td>\n      <td>116442</td>\n      <td>RAB39B</td>\n      <td>NM_171998</td>\n      <td>-</td>\n      <td>2</td>\n      <td>1</td>\n    </tr>\n    <tr>\n      <th>6032</th>\n      <td>18569</td>\n      <td>2654895</td>\n      <td>2655723</td>\n      <td>chrY</td>\n      <td>6736</td>\n      <td>SRY</td>\n      <td>NM_003140</td>\n      <td>-</td>\n      <td>1</td>\n      <td>1</td>\n    </tr>\n  </tbody>\n</table>\n</div>"
     },
     "execution_count": 183,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exon['+'] = exon.groupby(['name2', 'name'])['start'].rank('first', ascending=True).astype(int)\n",
    "exon['-'] = exon.groupby(['name2', 'name'])['start'].rank('first', ascending=False).astype(int)\n",
    "exon.head()\n",
    "exon.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 184,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-05-11T06:00:59.266034900Z",
     "start_time": "2024-05-11T06:00:59.195198900Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": "        row      start        end chrom  GeneID   name2       name strand   +  \\\n1600   1189  154227753  154227875  chrX    2157      F8  NM_000132      -  25   \n1601   1189  154250684  154250998  chrX    2157      F8  NM_000132      -  26   \n5975  15336  154487519  154490514  chrX  116442  RAB39B  NM_171998      -   1   \n5976  15336  154493358  154493776  chrX  116442  RAB39B  NM_171998      -   2   \n6032  18569    2654895    2655723  chrY    6736     SRY  NM_003140      -   1   \n\n      -  exon  \n1600  2     2  \n1601  1     1  \n5975  2     2  \n5976  1     1  \n6032  1     1  ",
      "text/html": "<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>row</th>\n      <th>start</th>\n      <th>end</th>\n      <th>chrom</th>\n      <th>GeneID</th>\n      <th>name2</th>\n      <th>name</th>\n      <th>strand</th>\n      <th>+</th>\n      <th>-</th>\n      <th>exon</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>1600</th>\n      <td>1189</td>\n      <td>154227753</td>\n      <td>154227875</td>\n      <td>chrX</td>\n      <td>2157</td>\n      <td>F8</td>\n      <td>NM_000132</td>\n      <td>-</td>\n      <td>25</td>\n      <td>2</td>\n      <td>2</td>\n    </tr>\n    <tr>\n      <th>1601</th>\n      <td>1189</td>\n      <td>154250684</td>\n      <td>154250998</td>\n      <td>chrX</td>\n      <td>2157</td>\n      <td>F8</td>\n      <td>NM_000132</td>\n      <td>-</td>\n      <td>26</td>\n      <td>1</td>\n      <td>1</td>\n    </tr>\n    <tr>\n      <th>5975</th>\n      <td>15336</td>\n      <td>154487519</td>\n      <td>154490514</td>\n      <td>chrX</td>\n      <td>116442</td>\n      <td>RAB39B</td>\n      <td>NM_171998</td>\n      <td>-</td>\n      <td>1</td>\n      <td>2</td>\n      <td>2</td>\n    </tr>\n    <tr>\n      <th>5976</th>\n      <td>15336</td>\n      <td>154493358</td>\n      <td>154493776</td>\n      <td>chrX</td>\n      <td>116442</td>\n      <td>RAB39B</td>\n      <td>NM_171998</td>\n      <td>-</td>\n      <td>2</td>\n      <td>1</td>\n      <td>1</td>\n    </tr>\n    <tr>\n      <th>6032</th>\n      <td>18569</td>\n      <td>2654895</td>\n      <td>2655723</td>\n      <td>chrY</td>\n      <td>6736</td>\n      <td>SRY</td>\n      <td>NM_003140</td>\n      <td>-</td>\n      <td>1</td>\n      <td>1</td>\n      <td>1</td>\n    </tr>\n  </tbody>\n</table>\n</div>"
     },
     "execution_count": 184,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exon['exon'] = pd.concat([exon.loc[exon['strand'] == '+', '+'], exon.loc[exon['strand'] == '-', '-']])\n",
    "exon.head()\n",
    "exon.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 185,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-05-11T06:00:59.267031800Z",
     "start_time": "2024-05-11T06:00:59.195198900Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": "        row      start        end chrom  GeneID   name2       name strand   +  \\\n1600   1189  154227753  154227875  chrX    2157      F8  NM_000132      -  25   \n1601   1189  154250684  154250998  chrX    2157      F8  NM_000132      -  26   \n5975  15336  154487519  154490514  chrX  116442  RAB39B  NM_171998      -   1   \n5976  15336  154493358  154493776  chrX  116442  RAB39B  NM_171998      -   2   \n6032  18569    2654895    2655723  chrY    6736     SRY  NM_003140      -   1   \n\n      -  exon  last_exon  \n1600  2     2      False  \n1601  1     1      False  \n5975  2     2       True  \n5976  1     1      False  \n6032  1     1       True  ",
      "text/html": "<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>row</th>\n      <th>start</th>\n      <th>end</th>\n      <th>chrom</th>\n      <th>GeneID</th>\n      <th>name2</th>\n      <th>name</th>\n      <th>strand</th>\n      <th>+</th>\n      <th>-</th>\n      <th>exon</th>\n      <th>last_exon</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>1600</th>\n      <td>1189</td>\n      <td>154227753</td>\n      <td>154227875</td>\n      <td>chrX</td>\n      <td>2157</td>\n      <td>F8</td>\n      <td>NM_000132</td>\n      <td>-</td>\n      <td>25</td>\n      <td>2</td>\n      <td>2</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>1601</th>\n      <td>1189</td>\n      <td>154250684</td>\n      <td>154250998</td>\n      <td>chrX</td>\n      <td>2157</td>\n      <td>F8</td>\n      <td>NM_000132</td>\n      <td>-</td>\n      <td>26</td>\n      <td>1</td>\n      <td>1</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>5975</th>\n      <td>15336</td>\n      <td>154487519</td>\n      <td>154490514</td>\n      <td>chrX</td>\n      <td>116442</td>\n      <td>RAB39B</td>\n      <td>NM_171998</td>\n      <td>-</td>\n      <td>1</td>\n      <td>2</td>\n      <td>2</td>\n      <td>True</td>\n    </tr>\n    <tr>\n      <th>5976</th>\n      <td>15336</td>\n      <td>154493358</td>\n      <td>154493776</td>\n      <td>chrX</td>\n      <td>116442</td>\n      <td>RAB39B</td>\n      <td>NM_171998</td>\n      <td>-</td>\n      <td>2</td>\n      <td>1</td>\n      <td>1</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>6032</th>\n      <td>18569</td>\n      <td>2654895</td>\n      <td>2655723</td>\n      <td>chrY</td>\n      <td>6736</td>\n      <td>SRY</td>\n      <td>NM_003140</td>\n      <td>-</td>\n      <td>1</td>\n      <td>1</td>\n      <td>1</td>\n      <td>True</td>\n    </tr>\n  </tbody>\n</table>\n</div>"
     },
     "execution_count": 185,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exon['last_exon'] = exon.groupby(['name2', 'name'])['exon'].transform('max') == exon['exon']\n",
    "exon.head()\n",
    "exon.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 186,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-05-11T06:00:59.309917300Z",
     "start_time": "2024-05-11T06:00:59.195198900Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": "    chrom    start      end  GeneID   name2          name  exon  last_exon\n128  chr1  6845513  6845635   23261  CAMTA1  NM_001349609     1      False\n129  chr1  6880240  6880310   23261  CAMTA1  NM_001349609     2      False\n130  chr1  6885151  6885270   23261  CAMTA1  NM_001349609     3      False\n131  chr1  7151363  7151431   23261  CAMTA1  NM_001349609     4      False\n132  chr1  7309550  7309686   23261  CAMTA1  NM_001349609     5      False",
      "text/html": "<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>chrom</th>\n      <th>start</th>\n      <th>end</th>\n      <th>GeneID</th>\n      <th>name2</th>\n      <th>name</th>\n      <th>exon</th>\n      <th>last_exon</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>128</th>\n      <td>chr1</td>\n      <td>6845513</td>\n      <td>6845635</td>\n      <td>23261</td>\n      <td>CAMTA1</td>\n      <td>NM_001349609</td>\n      <td>1</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>129</th>\n      <td>chr1</td>\n      <td>6880240</td>\n      <td>6880310</td>\n      <td>23261</td>\n      <td>CAMTA1</td>\n      <td>NM_001349609</td>\n      <td>2</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>130</th>\n      <td>chr1</td>\n      <td>6885151</td>\n      <td>6885270</td>\n      <td>23261</td>\n      <td>CAMTA1</td>\n      <td>NM_001349609</td>\n      <td>3</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>131</th>\n      <td>chr1</td>\n      <td>7151363</td>\n      <td>7151431</td>\n      <td>23261</td>\n      <td>CAMTA1</td>\n      <td>NM_001349609</td>\n      <td>4</td>\n      <td>False</td>\n    </tr>\n    <tr>\n      <th>132</th>\n      <td>chr1</td>\n      <td>7309550</td>\n      <td>7309686</td>\n      <td>23261</td>\n      <td>CAMTA1</td>\n      <td>NM_001349609</td>\n      <td>5</td>\n      <td>False</td>\n    </tr>\n  </tbody>\n</table>\n</div>"
     },
     "execution_count": 186,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exon = exon[\n",
    "    ['chrom', 'start', 'end', 'GeneID', 'name2', 'name', 'exon', 'last_exon']\n",
    "].sort_values(['chrom', 'start', 'end'])\n",
    "exon.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 187,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-05-11T06:00:59.309917300Z",
     "start_time": "2024-05-11T06:00:59.243095900Z"
    }
   },
   "outputs": [],
   "source": [
    "exon.rename(columns={\n",
    "    'chrom': '#chrom',\n",
    "    'GeneID': 'gene_id', 'name2': 'symbol', 'name': 'transcript'\n",
    "}).to_csv(hi_exon_file, index=False, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 188,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-05-11T06:01:01.293240600Z",
     "start_time": "2024-05-11T06:00:59.296953300Z"
    }
   },
   "outputs": [],
   "source": [
    "!bgzip -cf {hi_exon_file} > {hi_exon_file}.gz\n",
    "!tabix -fp bed {hi_exon_file}.gz\n",
    "!rm {hi_exon_file}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 189,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-05-11T06:01:01.331650200Z",
     "start_time": "2024-05-11T06:01:01.299222200Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": "     chrom     start       end  GeneID   name2          name  exon  last_exon\n150   chr1   7826518   7829766   23261  CAMTA1  NM_001349609    23       True\n4572  chr1  17345220  17345453    6390    SDHB     NM_003000     8       True\n2930  chr1  27105513  27108595    8289  ARID1A     NM_139135    20       True\n5919  chr1  27860755  27861427   27245   AHDC1  NM_001371928     9       True\n4776  chr1  43391025  43392912    6513  SLC2A1     NM_006516    10       True",
      "text/html": "<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>chrom</th>\n      <th>start</th>\n      <th>end</th>\n      <th>GeneID</th>\n      <th>name2</th>\n      <th>name</th>\n      <th>exon</th>\n      <th>last_exon</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>150</th>\n      <td>chr1</td>\n      <td>7826518</td>\n      <td>7829766</td>\n      <td>23261</td>\n      <td>CAMTA1</td>\n      <td>NM_001349609</td>\n      <td>23</td>\n      <td>True</td>\n    </tr>\n    <tr>\n      <th>4572</th>\n      <td>chr1</td>\n      <td>17345220</td>\n      <td>17345453</td>\n      <td>6390</td>\n      <td>SDHB</td>\n      <td>NM_003000</td>\n      <td>8</td>\n      <td>True</td>\n    </tr>\n    <tr>\n      <th>2930</th>\n      <td>chr1</td>\n      <td>27105513</td>\n      <td>27108595</td>\n      <td>8289</td>\n      <td>ARID1A</td>\n      <td>NM_139135</td>\n      <td>20</td>\n      <td>True</td>\n    </tr>\n    <tr>\n      <th>5919</th>\n      <td>chr1</td>\n      <td>27860755</td>\n      <td>27861427</td>\n      <td>27245</td>\n      <td>AHDC1</td>\n      <td>NM_001371928</td>\n      <td>9</td>\n      <td>True</td>\n    </tr>\n    <tr>\n      <th>4776</th>\n      <td>chr1</td>\n      <td>43391025</td>\n      <td>43392912</td>\n      <td>6513</td>\n      <td>SLC2A1</td>\n      <td>NM_006516</td>\n      <td>10</td>\n      <td>True</td>\n    </tr>\n  </tbody>\n</table>\n</div>"
     },
     "execution_count": 189,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "last_exon = exon[exon['last_exon']]\n",
    "last_exon.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 190,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-05-11T06:01:01.391518300Z",
     "start_time": "2024-05-11T06:01:01.326663500Z"
    }
   },
   "outputs": [],
   "source": [
    "last_exon_region = last_exon['chrom'] + ':' + last_exon['start'].astype(str) + '-' + last_exon['end'].astype(str)\n",
    "last_exon_region = last_exon_region.str.replace('chr', '')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 191,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-05-11T06:01:01.391518300Z",
     "start_time": "2024-05-11T06:01:01.366556400Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": "150       1:7826518-7829766\n4572    1:17345220-17345453\n2930    1:27105513-27108595\n5919    1:27860755-27861427\n4776    1:43391025-43392912\ndtype: object"
     },
     "execution_count": 191,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "last_exon_region.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 192,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-05-11T06:01:01.391518300Z",
     "start_time": "2024-05-11T06:01:01.366556400Z"
    }
   },
   "outputs": [],
   "source": [
    "need_fields = [\n",
    "    'variants/CHROM', 'variants/POS', 'variants/REF', 'variants/ALT',\n",
    "    'variants/AF_ESP', 'variants/AF_EXAC', 'variants/AF_TGP', 'variants/CLNSIG'\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 193,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-05-11T06:01:06.442969900Z",
     "start_time": "2024-05-11T06:01:01.367555300Z"
    }
   },
   "outputs": [],
   "source": [
    "!tabix -f {clinvar_ori_vcf_file}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 194,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-05-11T06:01:17.273153600Z",
     "start_time": "2024-05-11T06:01:06.446970600Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/staging/software/lib/python3.9/site-packages/autocnv-1-py3.9.egg/data/clinvar-pathogenic.sorted.vcf\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "360it [00:00, 2090.62it/s]\n"
     ]
    }
   ],
   "source": [
    "print(clinvar_file)\n",
    "with open(clinvar_file, 'w') as f:\n",
    "    headers = allel.read_vcf_headers(clinvar_ori_vcf_file)\n",
    "    f.write(''.join(headers.headers))\n",
    "    \n",
    "    def fetch_variants(region):\n",
    "        fields, samples, headers, it = allel.iter_vcf_chunks(\n",
    "            clinvar_ori_vcf_file, fields=need_fields, alt_number=1, region=region\n",
    "        )\n",
    "        for variants, *_ in it:\n",
    "            esp_filter = np.isnan(variants['variants/AF_ESP'])\n",
    "            esp_filter[~esp_filter] |= variants['variants/AF_ESP'][~esp_filter] < 0.01\n",
    "\n",
    "            exac_filter = np.isnan(variants['variants/AF_EXAC'])\n",
    "            exac_filter[~exac_filter] |= variants['variants/AF_EXAC'][~exac_filter] < 0.01\n",
    "\n",
    "            tgp_filter = np.isnan(variants['variants/AF_TGP'])\n",
    "            tgp_filter[~tgp_filter] |= variants['variants/AF_TGP'][~tgp_filter] < 0.01\n",
    "\n",
    "            pathogenic_filter = np.isin(\n",
    "                variants['variants/CLNSIG'], ['Likely_pathogenic', 'Pathogenic', 'Pathogenic/Likely_pathogenic']\n",
    "            )\n",
    "\n",
    "            af_filter = esp_filter & exac_filter & tgp_filter & pathogenic_filter\n",
    "\n",
    "            filtered_variants = {k: v[af_filter] for k, v in variants.items()}\n",
    "\n",
    "            filtered_variants['variants/CHROM'] = 'chr' + filtered_variants['variants/CHROM']\n",
    "\n",
    "            return allel.normalize_callset(filtered_variants)\n",
    "    \n",
    "    with Pool(processes=32) as pool:\n",
    "        variants = pool.map(fetch_variants, last_exon_region)\n",
    "    \n",
    "    for names, callset in tqdm(filter(lambda x: x is not None, variants)):\n",
    "        allel.write_vcf_data(f, names, callset, None, {field: np.nan for field in need_fields})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 195,
   "metadata": {
    "tags": [],
    "ExecuteTime": {
     "end_time": "2024-05-11T06:01:18.020546500Z",
     "start_time": "2024-05-11T06:01:17.276145600Z"
    }
   },
   "outputs": [],
   "source": [
    "!sed -i 's/AF_.\\+=nan;//' {clinvar_file}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 196,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-05-11T06:01:20.207505500Z",
     "start_time": "2024-05-11T06:01:18.117697300Z"
    }
   },
   "outputs": [],
   "source": [
    "!bgzip -cf {clinvar_file} > {clinvar_file}.gz\n",
    "!tabix -fp vcf {clinvar_file}.gz\n",
    "!rm {clinvar_file}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 197,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-05-11T06:01:20.207505500Z",
     "start_time": "2024-05-11T06:01:20.187887800Z"
    }
   },
   "outputs": [],
   "source": [
    "uhi_genes = set(\n",
    "    curation_gene.loc[curation_gene['Haploinsufficiency Score'] == '40', '#Gene Symbol']\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 198,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-05-11T06:01:20.238919200Z",
     "start_time": "2024-05-11T06:01:20.207505500Z"
    }
   },
   "outputs": [],
   "source": [
    "uhi_gene = refgene_info.loc[\n",
    "    refgene_info['name2'].isin(uhi_genes),\n",
    "    ['chrom', 'txStart', 'txEnd', 'GeneID', 'name2', 'name', 'strand']\n",
    "].sort_values(['chrom', 'txStart', 'txEnd'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 199,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-05-11T06:01:20.267807300Z",
     "start_time": "2024-05-11T06:01:20.238919200Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": "       chrom    txStart      txEnd  GeneID    name2          name strand\n4986    chr1   32117864   32169618    1307  COL16A1     NM_001856      -\n17425   chr1   40974390   40981722   64789     EXO5  NM_001346956      +\n8295    chr1   55505220   55530525  255738    PCSK9     NM_174936      +\n8352    chr1  171154438  171181824    2327     FMO2     NM_001460      +\n9908   chr10   74870212   74891581   25961   NUDT13     NM_015901      +",
      "text/html": "<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>chrom</th>\n      <th>txStart</th>\n      <th>txEnd</th>\n      <th>GeneID</th>\n      <th>name2</th>\n      <th>name</th>\n      <th>strand</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>4986</th>\n      <td>chr1</td>\n      <td>32117864</td>\n      <td>32169618</td>\n      <td>1307</td>\n      <td>COL16A1</td>\n      <td>NM_001856</td>\n      <td>-</td>\n    </tr>\n    <tr>\n      <th>17425</th>\n      <td>chr1</td>\n      <td>40974390</td>\n      <td>40981722</td>\n      <td>64789</td>\n      <td>EXO5</td>\n      <td>NM_001346956</td>\n      <td>+</td>\n    </tr>\n    <tr>\n      <th>8295</th>\n      <td>chr1</td>\n      <td>55505220</td>\n      <td>55530525</td>\n      <td>255738</td>\n      <td>PCSK9</td>\n      <td>NM_174936</td>\n      <td>+</td>\n    </tr>\n    <tr>\n      <th>8352</th>\n      <td>chr1</td>\n      <td>171154438</td>\n      <td>171181824</td>\n      <td>2327</td>\n      <td>FMO2</td>\n      <td>NM_001460</td>\n      <td>+</td>\n    </tr>\n    <tr>\n      <th>9908</th>\n      <td>chr10</td>\n      <td>74870212</td>\n      <td>74891581</td>\n      <td>25961</td>\n      <td>NUDT13</td>\n      <td>NM_015901</td>\n      <td>+</td>\n    </tr>\n  </tbody>\n</table>\n</div>"
     },
     "execution_count": 199,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "uhi_gene.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 200,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-05-11T06:01:20.268804300Z",
     "start_time": "2024-05-11T06:01:20.242873200Z"
    }
   },
   "outputs": [],
   "source": [
    "uhi_gene.rename(columns={\n",
    "    'chrom': '#chrom', 'txStart': 'start', 'txEnd': 'end',\n",
    "    'GeneID': 'gene_id', 'name2': 'symbol', 'name': 'transcript'\n",
    "}).to_csv(uhi_gene_file, index=False, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 201,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-05-11T06:01:22.324879600Z",
     "start_time": "2024-05-11T06:01:20.345264600Z"
    }
   },
   "outputs": [],
   "source": [
    "!bgzip -cf {uhi_gene_file} > {uhi_gene_file}.gz\n",
    "!tabix -fp bed {uhi_gene_file}.gz\n",
    "!rm {uhi_gene_file}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 249,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-05-11T06:05:08.089021800Z",
     "start_time": "2024-05-11T06:05:07.369055700Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": "     #ISCA ID                                   ISCA Region Name  \\\n0  ISCA-46747  22q11.2 recurrent region (distal type III, E/F...   \n1  ISCA-46743                       Xq25 region (includes STAG2)   \n2  ISCA-46742                      7p22.1 region (includes ACTB)   \n3  ISCA-46741                     9p24.3 Region (includes DMRT1)   \n4  ISCA-46739  Yq11.223 population region (DGV_Gold_Standard_...   \n\n         cytoBand          Genomic Location Haploinsufficiency Score  \\\n0  22q11.22q11.23   chr22:23119415-24994433                        3   \n1            Xq25  chrX:123034319-123236519                        3   \n2          7p22.1      chr7:5536848-5799722                        3   \n3          9p24.3            chr9:1-1982069                        2   \n4        Yq11.223    chrY:25853189-26146959                       40   \n\n                 Haploinsufficiency Description Haploinsufficiency PMID1  \\\n0  Sufficient evidence for dosage pathogenicity                      NaN   \n1  Sufficient evidence for dosage pathogenicity                      NaN   \n2  Sufficient evidence for dosage pathogenicity                 27633570   \n3        Some evidence for dosage pathogenicity                 20685758   \n4                   Dosage sensitivity unlikely                      NaN   \n\n  Haploinsufficiency PMID2 Haploinsufficiency PMID3 Haploinsufficiency PMID4  \\\n0                      NaN                      NaN                      NaN   \n1                      NaN                      NaN                      NaN   \n2                 29220674                 32562408                      NaN   \n3                 22573722                 23555275                 32741963   \n4                      NaN                      NaN                      NaN   \n\n  Haploinsufficiency PMID5 Haploinsufficiency PMID6 Triplosensitivity Score  \\\n0                      NaN                      NaN                       1   \n1                      NaN                      NaN                       3   \n2                      NaN                      NaN                       1   \n3                      NaN                      NaN       Not yet evaluated   \n4                      NaN                      NaN                       0   \n\n                  Triplosensitivity Description Triplosensitivity PMID1  \\\n0      Little evidence for dosage pathogenicity                19193630   \n1  Sufficient evidence for dosage pathogenicity                23637084   \n2      Little evidence for dosage pathogenicity                26297194   \n3                             Not yet evaluated                     NaN   \n4                         No evidence available                     NaN   \n\n  Triplosensitivity PMID2 Triplosensitivity PMID3 Triplosensitivity PMID4  \\\n0                22140377                19006218                26247050   \n1                24733578                26443594                25677961   \n2                21998864                22495914                25893121   \n3                     NaN                     NaN                     NaN   \n4                     NaN                     NaN                     NaN   \n\n  Triplosensitivity PMID5 Triplosensitivity PMID6 Date Last Evaluated  \\\n0                     NaN                     NaN          2023-01-25   \n1                31609727                     NaN          2022-11-30   \n2                25124455                27866048          2022-02-08   \n3                     NaN                     NaN          2023-11-10   \n4                     NaN                     NaN          2021-08-03   \n\n  Haploinsufficiency Disease ID Triplosensitivity Disease ID  \n0                           NaN                          NaN  \n1                           NaN                MONDO:0100038  \n2                 MONDO:0000508                MONDO:0100038  \n3                 MONDO:0002145                          NaN  \n4                           NaN                          NaN  ",
      "text/html": "<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>#ISCA ID</th>\n      <th>ISCA Region Name</th>\n      <th>cytoBand</th>\n      <th>Genomic Location</th>\n      <th>Haploinsufficiency Score</th>\n      <th>Haploinsufficiency Description</th>\n      <th>Haploinsufficiency PMID1</th>\n      <th>Haploinsufficiency PMID2</th>\n      <th>Haploinsufficiency PMID3</th>\n      <th>Haploinsufficiency PMID4</th>\n      <th>Haploinsufficiency PMID5</th>\n      <th>Haploinsufficiency PMID6</th>\n      <th>Triplosensitivity Score</th>\n      <th>Triplosensitivity Description</th>\n      <th>Triplosensitivity PMID1</th>\n      <th>Triplosensitivity PMID2</th>\n      <th>Triplosensitivity PMID3</th>\n      <th>Triplosensitivity PMID4</th>\n      <th>Triplosensitivity PMID5</th>\n      <th>Triplosensitivity PMID6</th>\n      <th>Date Last Evaluated</th>\n      <th>Haploinsufficiency Disease ID</th>\n      <th>Triplosensitivity Disease ID</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>0</th>\n      <td>ISCA-46747</td>\n      <td>22q11.2 recurrent region (distal type III, E/F...</td>\n      <td>22q11.22q11.23</td>\n      <td>chr22:23119415-24994433</td>\n      <td>3</td>\n      <td>Sufficient evidence for dosage pathogenicity</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>1</td>\n      <td>Little evidence for dosage pathogenicity</td>\n      <td>19193630</td>\n      <td>22140377</td>\n      <td>19006218</td>\n      <td>26247050</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>2023-01-25</td>\n      <td>NaN</td>\n      <td>NaN</td>\n    </tr>\n    <tr>\n      <th>1</th>\n      <td>ISCA-46743</td>\n      <td>Xq25 region (includes STAG2)</td>\n      <td>Xq25</td>\n      <td>chrX:123034319-123236519</td>\n      <td>3</td>\n      <td>Sufficient evidence for dosage pathogenicity</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>3</td>\n      <td>Sufficient evidence for dosage pathogenicity</td>\n      <td>23637084</td>\n      <td>24733578</td>\n      <td>26443594</td>\n      <td>25677961</td>\n      <td>31609727</td>\n      <td>NaN</td>\n      <td>2022-11-30</td>\n      <td>NaN</td>\n      <td>MONDO:0100038</td>\n    </tr>\n    <tr>\n      <th>2</th>\n      <td>ISCA-46742</td>\n      <td>7p22.1 region (includes ACTB)</td>\n      <td>7p22.1</td>\n      <td>chr7:5536848-5799722</td>\n      <td>3</td>\n      <td>Sufficient evidence for dosage pathogenicity</td>\n      <td>27633570</td>\n      <td>29220674</td>\n      <td>32562408</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>1</td>\n      <td>Little evidence for dosage pathogenicity</td>\n      <td>26297194</td>\n      <td>21998864</td>\n      <td>22495914</td>\n      <td>25893121</td>\n      <td>25124455</td>\n      <td>27866048</td>\n      <td>2022-02-08</td>\n      <td>MONDO:0000508</td>\n      <td>MONDO:0100038</td>\n    </tr>\n    <tr>\n      <th>3</th>\n      <td>ISCA-46741</td>\n      <td>9p24.3 Region (includes DMRT1)</td>\n      <td>9p24.3</td>\n      <td>chr9:1-1982069</td>\n      <td>2</td>\n      <td>Some evidence for dosage pathogenicity</td>\n      <td>20685758</td>\n      <td>22573722</td>\n      <td>23555275</td>\n      <td>32741963</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>Not yet evaluated</td>\n      <td>Not yet evaluated</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>2023-11-10</td>\n      <td>MONDO:0002145</td>\n      <td>NaN</td>\n    </tr>\n    <tr>\n      <th>4</th>\n      <td>ISCA-46739</td>\n      <td>Yq11.223 population region (DGV_Gold_Standard_...</td>\n      <td>Yq11.223</td>\n      <td>chrY:25853189-26146959</td>\n      <td>40</td>\n      <td>Dosage sensitivity unlikely</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>0</td>\n      <td>No evidence available</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>2021-08-03</td>\n      <td>NaN</td>\n      <td>NaN</td>\n    </tr>\n  </tbody>\n</table>\n</div>"
     },
     "execution_count": 249,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# ISCA-46748 invalid region moved to line 5 to skip\n",
    "!sed -i '/^ISCA-46748/d' {clingen_region_ori_file}\n",
    "region = pd.read_csv(clingen_region_ori_file, sep='\\t', skiprows=5, dtype=str)\n",
    "region.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 203,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-05-11T06:01:22.388881400Z",
     "start_time": "2024-05-11T06:01:22.366939800Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": "   chrom      start        end\n0  chr22   23119415   24994433\n1   chrX  123034319  123236519\n2   chr7    5536848    5799722\n3   chr9          1    1982069\n4   chrY   25853189   26146959",
      "text/html": "<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>chrom</th>\n      <th>start</th>\n      <th>end</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>0</th>\n      <td>chr22</td>\n      <td>23119415</td>\n      <td>24994433</td>\n    </tr>\n    <tr>\n      <th>1</th>\n      <td>chrX</td>\n      <td>123034319</td>\n      <td>123236519</td>\n    </tr>\n    <tr>\n      <th>2</th>\n      <td>chr7</td>\n      <td>5536848</td>\n      <td>5799722</td>\n    </tr>\n    <tr>\n      <th>3</th>\n      <td>chr9</td>\n      <td>1</td>\n      <td>1982069</td>\n    </tr>\n    <tr>\n      <th>4</th>\n      <td>chrY</td>\n      <td>25853189</td>\n      <td>26146959</td>\n    </tr>\n  </tbody>\n</table>\n</div>"
     },
     "execution_count": 203,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "position = region['Genomic Location'].str.extract(r'(?P<chrom>chr\\w+)\\s*:\\s*(?P<start>\\d+)\\s*-\\s*(?P<end>\\d+)')\n",
    "position.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 204,
   "outputs": [
    {
     "data": {
      "text/plain": "   chrom      start        end\n0  chr22   23119415   24994433\n1   chrX  123034319  123236519\n2   chr7    5536848    5799722\n3   chr9          1    1982069\n4   chrY   25853189   26146959",
      "text/html": "<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>chrom</th>\n      <th>start</th>\n      <th>end</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>0</th>\n      <td>chr22</td>\n      <td>23119415</td>\n      <td>24994433</td>\n    </tr>\n    <tr>\n      <th>1</th>\n      <td>chrX</td>\n      <td>123034319</td>\n      <td>123236519</td>\n    </tr>\n    <tr>\n      <th>2</th>\n      <td>chr7</td>\n      <td>5536848</td>\n      <td>5799722</td>\n    </tr>\n    <tr>\n      <th>3</th>\n      <td>chr9</td>\n      <td>1</td>\n      <td>1982069</td>\n    </tr>\n    <tr>\n      <th>4</th>\n      <td>chrY</td>\n      <td>25853189</td>\n      <td>26146959</td>\n    </tr>\n  </tbody>\n</table>\n</div>"
     },
     "execution_count": 204,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "position['start'] = position['start'].astype(int)\n",
    "position['end'] = position['end'].astype(int)\n",
    "position.head()"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2024-05-11T06:01:22.403841100Z",
     "start_time": "2024-05-11T06:01:22.383894700Z"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 205,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-05-11T06:01:22.572052300Z",
     "start_time": "2024-05-11T06:01:22.396860500Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": "     #ISCA ID                                   ISCA Region Name  \\\n0  ISCA-46747  22q11.2 recurrent region (distal type III, E/F...   \n1  ISCA-46743                       Xq25 region (includes STAG2)   \n2  ISCA-46742                      7p22.1 region (includes ACTB)   \n3  ISCA-46741                     9p24.3 Region (includes DMRT1)   \n4  ISCA-46739  Yq11.223 population region (DGV_Gold_Standard_...   \n\n         cytoBand          Genomic Location Haploinsufficiency Score  \\\n0  22q11.22q11.23   chr22:23119415-24994433                        3   \n1            Xq25  chrX:123034319-123236519                        3   \n2          7p22.1      chr7:5536848-5799722                        3   \n3          9p24.3            chr9:1-1982069                        2   \n4        Yq11.223    chrY:25853189-26146959                       40   \n\n                 Haploinsufficiency Description Haploinsufficiency PMID1  \\\n0  Sufficient evidence for dosage pathogenicity                      NaN   \n1  Sufficient evidence for dosage pathogenicity                      NaN   \n2  Sufficient evidence for dosage pathogenicity                 27633570   \n3        Some evidence for dosage pathogenicity                 20685758   \n4                   Dosage sensitivity unlikely                      NaN   \n\n  Haploinsufficiency PMID2 Haploinsufficiency PMID3 Haploinsufficiency PMID4  \\\n0                      NaN                      NaN                      NaN   \n1                      NaN                      NaN                      NaN   \n2                 29220674                 32562408                      NaN   \n3                 22573722                 23555275                 32741963   \n4                      NaN                      NaN                      NaN   \n\n  Haploinsufficiency PMID5 Haploinsufficiency PMID6 Triplosensitivity Score  \\\n0                      NaN                      NaN                       1   \n1                      NaN                      NaN                       3   \n2                      NaN                      NaN                       1   \n3                      NaN                      NaN       Not yet evaluated   \n4                      NaN                      NaN                       0   \n\n                  Triplosensitivity Description Triplosensitivity PMID1  \\\n0      Little evidence for dosage pathogenicity                19193630   \n1  Sufficient evidence for dosage pathogenicity                23637084   \n2      Little evidence for dosage pathogenicity                26297194   \n3                             Not yet evaluated                     NaN   \n4                         No evidence available                     NaN   \n\n  Triplosensitivity PMID2 Triplosensitivity PMID3 Triplosensitivity PMID4  \\\n0                22140377                19006218                26247050   \n1                24733578                26443594                25677961   \n2                21998864                22495914                25893121   \n3                     NaN                     NaN                     NaN   \n4                     NaN                     NaN                     NaN   \n\n  Triplosensitivity PMID5 Triplosensitivity PMID6 Date Last Evaluated  \\\n0                     NaN                     NaN          2023-01-25   \n1                31609727                     NaN          2022-11-30   \n2                25124455                27866048          2022-02-08   \n3                     NaN                     NaN          2023-11-10   \n4                     NaN                     NaN          2021-08-03   \n\n  Haploinsufficiency Disease ID Triplosensitivity Disease ID  chrom  \\\n0                           NaN                          NaN  chr22   \n1                           NaN                MONDO:0100038   chrX   \n2                 MONDO:0000508                MONDO:0100038   chr7   \n3                 MONDO:0002145                          NaN   chr9   \n4                           NaN                          NaN   chrY   \n\n       start        end  \n0   23119415   24994433  \n1  123034319  123236519  \n2    5536848    5799722  \n3          1    1982069  \n4   25853189   26146959  ",
      "text/html": "<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>#ISCA ID</th>\n      <th>ISCA Region Name</th>\n      <th>cytoBand</th>\n      <th>Genomic Location</th>\n      <th>Haploinsufficiency Score</th>\n      <th>Haploinsufficiency Description</th>\n      <th>Haploinsufficiency PMID1</th>\n      <th>Haploinsufficiency PMID2</th>\n      <th>Haploinsufficiency PMID3</th>\n      <th>Haploinsufficiency PMID4</th>\n      <th>Haploinsufficiency PMID5</th>\n      <th>Haploinsufficiency PMID6</th>\n      <th>Triplosensitivity Score</th>\n      <th>Triplosensitivity Description</th>\n      <th>Triplosensitivity PMID1</th>\n      <th>Triplosensitivity PMID2</th>\n      <th>Triplosensitivity PMID3</th>\n      <th>Triplosensitivity PMID4</th>\n      <th>Triplosensitivity PMID5</th>\n      <th>Triplosensitivity PMID6</th>\n      <th>Date Last Evaluated</th>\n      <th>Haploinsufficiency Disease ID</th>\n      <th>Triplosensitivity Disease ID</th>\n      <th>chrom</th>\n      <th>start</th>\n      <th>end</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>0</th>\n      <td>ISCA-46747</td>\n      <td>22q11.2 recurrent region (distal type III, E/F...</td>\n      <td>22q11.22q11.23</td>\n      <td>chr22:23119415-24994433</td>\n      <td>3</td>\n      <td>Sufficient evidence for dosage pathogenicity</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>1</td>\n      <td>Little evidence for dosage pathogenicity</td>\n      <td>19193630</td>\n      <td>22140377</td>\n      <td>19006218</td>\n      <td>26247050</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>2023-01-25</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>chr22</td>\n      <td>23119415</td>\n      <td>24994433</td>\n    </tr>\n    <tr>\n      <th>1</th>\n      <td>ISCA-46743</td>\n      <td>Xq25 region (includes STAG2)</td>\n      <td>Xq25</td>\n      <td>chrX:123034319-123236519</td>\n      <td>3</td>\n      <td>Sufficient evidence for dosage pathogenicity</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>3</td>\n      <td>Sufficient evidence for dosage pathogenicity</td>\n      <td>23637084</td>\n      <td>24733578</td>\n      <td>26443594</td>\n      <td>25677961</td>\n      <td>31609727</td>\n      <td>NaN</td>\n      <td>2022-11-30</td>\n      <td>NaN</td>\n      <td>MONDO:0100038</td>\n      <td>chrX</td>\n      <td>123034319</td>\n      <td>123236519</td>\n    </tr>\n    <tr>\n      <th>2</th>\n      <td>ISCA-46742</td>\n      <td>7p22.1 region (includes ACTB)</td>\n      <td>7p22.1</td>\n      <td>chr7:5536848-5799722</td>\n      <td>3</td>\n      <td>Sufficient evidence for dosage pathogenicity</td>\n      <td>27633570</td>\n      <td>29220674</td>\n      <td>32562408</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>1</td>\n      <td>Little evidence for dosage pathogenicity</td>\n      <td>26297194</td>\n      <td>21998864</td>\n      <td>22495914</td>\n      <td>25893121</td>\n      <td>25124455</td>\n      <td>27866048</td>\n      <td>2022-02-08</td>\n      <td>MONDO:0000508</td>\n      <td>MONDO:0100038</td>\n      <td>chr7</td>\n      <td>5536848</td>\n      <td>5799722</td>\n    </tr>\n    <tr>\n      <th>3</th>\n      <td>ISCA-46741</td>\n      <td>9p24.3 Region (includes DMRT1)</td>\n      <td>9p24.3</td>\n      <td>chr9:1-1982069</td>\n      <td>2</td>\n      <td>Some evidence for dosage pathogenicity</td>\n      <td>20685758</td>\n      <td>22573722</td>\n      <td>23555275</td>\n      <td>32741963</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>Not yet evaluated</td>\n      <td>Not yet evaluated</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>2023-11-10</td>\n      <td>MONDO:0002145</td>\n      <td>NaN</td>\n      <td>chr9</td>\n      <td>1</td>\n      <td>1982069</td>\n    </tr>\n    <tr>\n      <th>4</th>\n      <td>ISCA-46739</td>\n      <td>Yq11.223 population region (DGV_Gold_Standard_...</td>\n      <td>Yq11.223</td>\n      <td>chrY:25853189-26146959</td>\n      <td>40</td>\n      <td>Dosage sensitivity unlikely</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>0</td>\n      <td>No evidence available</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>2021-08-03</td>\n      <td>NaN</td>\n      <td>NaN</td>\n      <td>chrY</td>\n      <td>25853189</td>\n      <td>26146959</td>\n    </tr>\n  </tbody>\n</table>\n</div>"
     },
     "execution_count": 205,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "region_pos = region.merge(position, how='left', left_index=True, right_index=True)\n",
    "region_pos.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 206,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-05-11T06:01:22.643888600Z",
     "start_time": "2024-05-11T06:01:22.436753400Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": "     chrom      start        end    #ISCA ID  \\\n484   chr1     834083    6289973  ISCA-37434   \n490   chr1  145386507  145748064  ISCA-37428   \n494   chr1  146577486  147394506  ISCA-37421   \n459   chr1  243287730  245318287  ISCA-37493   \n455  chr10   49389703   51053583  ISCA-37497   \n\n                                      ISCA Region Name  \n484              1p36 terminal region (includes GABRD)  \n490  1q21.1 recurrent (TAR syndrome) region (proxim...  \n494  1q21.1 recurrent region (BP3-BP4, distal) (inc...  \n459            1q43q44 terminal region (includes AKT3)  \n455             10q11.2 recurrent region (LCR C-LCR D)  ",
      "text/html": "<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>chrom</th>\n      <th>start</th>\n      <th>end</th>\n      <th>#ISCA ID</th>\n      <th>ISCA Region Name</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>484</th>\n      <td>chr1</td>\n      <td>834083</td>\n      <td>6289973</td>\n      <td>ISCA-37434</td>\n      <td>1p36 terminal region (includes GABRD)</td>\n    </tr>\n    <tr>\n      <th>490</th>\n      <td>chr1</td>\n      <td>145386507</td>\n      <td>145748064</td>\n      <td>ISCA-37428</td>\n      <td>1q21.1 recurrent (TAR syndrome) region (proxim...</td>\n    </tr>\n    <tr>\n      <th>494</th>\n      <td>chr1</td>\n      <td>146577486</td>\n      <td>147394506</td>\n      <td>ISCA-37421</td>\n      <td>1q21.1 recurrent region (BP3-BP4, distal) (inc...</td>\n    </tr>\n    <tr>\n      <th>459</th>\n      <td>chr1</td>\n      <td>243287730</td>\n      <td>245318287</td>\n      <td>ISCA-37493</td>\n      <td>1q43q44 terminal region (includes AKT3)</td>\n    </tr>\n    <tr>\n      <th>455</th>\n      <td>chr10</td>\n      <td>49389703</td>\n      <td>51053583</td>\n      <td>ISCA-37497</td>\n      <td>10q11.2 recurrent region (LCR C-LCR D)</td>\n    </tr>\n  </tbody>\n</table>\n</div>"
     },
     "execution_count": 206,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "func_region = region_pos.loc[\n",
    "    (region_pos['Haploinsufficiency Score'].isin(['1', '2','3']))\n",
    "     | (region_pos['Triplosensitivity Score'].isin(['1', '2','3'])),\n",
    "     ['chrom', 'start', 'end', '#ISCA ID', 'ISCA Region Name']\n",
    "].sort_values(['chrom', 'start', 'end'])\n",
    "func_region.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 207,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-05-11T06:01:22.644886800Z",
     "start_time": "2024-05-11T06:01:22.476646800Z"
    }
   },
   "outputs": [],
   "source": [
    "func_region.rename(columns={\n",
    "    'chrom': '#chrom',\n",
    "    '#ISCA ID': 'isca_id', 'ISCA Region Name': 'name'\n",
    "}).to_csv(func_region_file, sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 208,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-05-11T06:01:24.494864100Z",
     "start_time": "2024-05-11T06:01:22.476646800Z"
    }
   },
   "outputs": [],
   "source": [
    "!bgzip -cf {func_region_file} > {func_region_file}.gz\n",
    "!tabix -fp bed {func_region_file}.gz\n",
    "!rm {func_region_file}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 209,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-05-11T06:01:24.556927800Z",
     "start_time": "2024-05-11T06:01:24.524035200Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": "     chrom      start        end    #ISCA ID  \\\n484   chr1     834083    6289973  ISCA-37434   \n494   chr1  146577486  147394506  ISCA-37421   \n459   chr1  243287730  245318287  ISCA-37493   \n492  chr10   81682843   88739388  ISCA-37424   \n504  chr11   31803509   32510988  ISCA-37401   \n\n                                      ISCA Region Name  \n484              1p36 terminal region (includes GABRD)  \n494  1q21.1 recurrent region (BP3-BP4, distal) (inc...  \n459            1q43q44 terminal region (includes AKT3)  \n492    10q22.3q23.2 recurrent region (includes BMPR1A)  \n504                       11p13 (WAGR syndrome) region  ",
      "text/html": "<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>chrom</th>\n      <th>start</th>\n      <th>end</th>\n      <th>#ISCA ID</th>\n      <th>ISCA Region Name</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>484</th>\n      <td>chr1</td>\n      <td>834083</td>\n      <td>6289973</td>\n      <td>ISCA-37434</td>\n      <td>1p36 terminal region (includes GABRD)</td>\n    </tr>\n    <tr>\n      <th>494</th>\n      <td>chr1</td>\n      <td>146577486</td>\n      <td>147394506</td>\n      <td>ISCA-37421</td>\n      <td>1q21.1 recurrent region (BP3-BP4, distal) (inc...</td>\n    </tr>\n    <tr>\n      <th>459</th>\n      <td>chr1</td>\n      <td>243287730</td>\n      <td>245318287</td>\n      <td>ISCA-37493</td>\n      <td>1q43q44 terminal region (includes AKT3)</td>\n    </tr>\n    <tr>\n      <th>492</th>\n      <td>chr10</td>\n      <td>81682843</td>\n      <td>88739388</td>\n      <td>ISCA-37424</td>\n      <td>10q22.3q23.2 recurrent region (includes BMPR1A)</td>\n    </tr>\n    <tr>\n      <th>504</th>\n      <td>chr11</td>\n      <td>31803509</td>\n      <td>32510988</td>\n      <td>ISCA-37401</td>\n      <td>11p13 (WAGR syndrome) region</td>\n    </tr>\n  </tbody>\n</table>\n</div>"
     },
     "execution_count": 209,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hi_region = region_pos.loc[\n",
    "    region_pos['Haploinsufficiency Score'] == '3',\n",
    "    ['chrom', 'start', 'end', '#ISCA ID', 'ISCA Region Name']\n",
    "].sort_values(['chrom', 'start', 'end'])\n",
    "hi_region.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 210,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-05-11T06:01:24.717693500Z",
     "start_time": "2024-05-11T06:01:24.552936Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": "     chrom      start        end    #ISCA ID  \\\n484   chr1     834083    6289973  ISCA-37434   \n494   chr1  146577486  147394506  ISCA-37421   \n459   chr1  243287730  245318287  ISCA-37493   \n492  chr10   81682843   88739388  ISCA-37424   \n504  chr11   31803509   32510988  ISCA-37401   \n\n                                      ISCA Region Name  \\\n484              1p36 terminal region (includes GABRD)   \n494  1q21.1 recurrent region (BP3-BP4, distal) (inc...   \n459            1q43q44 terminal region (includes AKT3)   \n492    10q22.3q23.2 recurrent region (includes BMPR1A)   \n504                       11p13 (WAGR syndrome) region   \n\n                                            omim_genes  \n484  SAMD11,NOC2L,KLHL17,PERM1,HES4,ISG15,AGRN,MIR2...  \n494              PRKAB2,FMO5,CHD1L,BCL9,ACP6,GJA5,GJA8  \n459  CEP170,SDCCAG8,AKT3,ZBTB18,ADSS2,CATSPERE,DESI...  \n492  SFTPD,PLAC9,ANXA11,MAT1A,DYDC1,PRXL2A,NRG3,GHI...  \n504                          ELP4,PAX6,RCN1,WT1,WT1-AS  ",
      "text/html": "<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>chrom</th>\n      <th>start</th>\n      <th>end</th>\n      <th>#ISCA ID</th>\n      <th>ISCA Region Name</th>\n      <th>omim_genes</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>484</th>\n      <td>chr1</td>\n      <td>834083</td>\n      <td>6289973</td>\n      <td>ISCA-37434</td>\n      <td>1p36 terminal region (includes GABRD)</td>\n      <td>SAMD11,NOC2L,KLHL17,PERM1,HES4,ISG15,AGRN,MIR2...</td>\n    </tr>\n    <tr>\n      <th>494</th>\n      <td>chr1</td>\n      <td>146577486</td>\n      <td>147394506</td>\n      <td>ISCA-37421</td>\n      <td>1q21.1 recurrent region (BP3-BP4, distal) (inc...</td>\n      <td>PRKAB2,FMO5,CHD1L,BCL9,ACP6,GJA5,GJA8</td>\n    </tr>\n    <tr>\n      <th>459</th>\n      <td>chr1</td>\n      <td>243287730</td>\n      <td>245318287</td>\n      <td>ISCA-37493</td>\n      <td>1q43q44 terminal region (includes AKT3)</td>\n      <td>CEP170,SDCCAG8,AKT3,ZBTB18,ADSS2,CATSPERE,DESI...</td>\n    </tr>\n    <tr>\n      <th>492</th>\n      <td>chr10</td>\n      <td>81682843</td>\n      <td>88739388</td>\n      <td>ISCA-37424</td>\n      <td>10q22.3q23.2 recurrent region (includes BMPR1A)</td>\n      <td>SFTPD,PLAC9,ANXA11,MAT1A,DYDC1,PRXL2A,NRG3,GHI...</td>\n    </tr>\n    <tr>\n      <th>504</th>\n      <td>chr11</td>\n      <td>31803509</td>\n      <td>32510988</td>\n      <td>ISCA-37401</td>\n      <td>11p13 (WAGR syndrome) region</td>\n      <td>ELP4,PAX6,RCN1,WT1,WT1-AS</td>\n    </tr>\n  </tbody>\n</table>\n</div>"
     },
     "execution_count": 210,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hi_region['omim_genes'] = hi_region.apply(\n",
    "    lambda row: ','.join(omim_gene.loc[\n",
    "        (omim_gene['chrom'] == row['chrom'])\n",
    "        & (omim_gene['txEnd'] >= row['start'])\n",
    "        & (omim_gene['txStart'] <= row['end']),\n",
    "        'name2'\n",
    "    ]),\n",
    "    axis=1\n",
    ")\n",
    "hi_region.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 211,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-05-11T06:01:24.718690800Z",
     "start_time": "2024-05-11T06:01:24.666836Z"
    }
   },
   "outputs": [],
   "source": [
    "hi_region.rename(columns={\n",
    "    'chrom': '#chrom',\n",
    "    '#ISCA ID': 'isca_id', 'ISCA Region Name': 'name'\n",
    "}).to_csv(hi_region_file, sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 212,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-05-11T06:01:26.716394Z",
     "start_time": "2024-05-11T06:01:24.757586200Z"
    }
   },
   "outputs": [],
   "source": [
    "!bgzip -cf {hi_region_file} > {hi_region_file}.gz\n",
    "!tabix -fp bed {hi_region_file}.gz\n",
    "!rm {hi_region_file}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 213,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-05-11T06:01:26.782896900Z",
     "start_time": "2024-05-11T06:01:26.728954700Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": "    chrom     start       end    #ISCA ID  \\\n433  chr1    363999    401750  ISCA-46306   \n432  chr1   1640999   1647000  ISCA-46307   \n180  chr1  12886641  12942069  ISCA-46561   \n430  chr1  12896424  12930275  ISCA-46309   \n179  chr1  12899395  12924283  ISCA-46562   \n\n                                      ISCA Region Name  \n433  1p36.33 population region (gnomAD-SV_v2.1_DEL_...  \n432  1p36.33 population region (gnomAD-SV_v2.1_DEL_...  \n180  1p36.21 population region (DGV_Gold_Standard_J...  \n430  1p36.21 population region (gnomAD-SV_v2.1_DEL_...  \n179  1p36.21 population region (DGV_Gold_Standard_J...  ",
      "text/html": "<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>chrom</th>\n      <th>start</th>\n      <th>end</th>\n      <th>#ISCA ID</th>\n      <th>ISCA Region Name</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>433</th>\n      <td>chr1</td>\n      <td>363999</td>\n      <td>401750</td>\n      <td>ISCA-46306</td>\n      <td>1p36.33 population region (gnomAD-SV_v2.1_DEL_...</td>\n    </tr>\n    <tr>\n      <th>432</th>\n      <td>chr1</td>\n      <td>1640999</td>\n      <td>1647000</td>\n      <td>ISCA-46307</td>\n      <td>1p36.33 population region (gnomAD-SV_v2.1_DEL_...</td>\n    </tr>\n    <tr>\n      <th>180</th>\n      <td>chr1</td>\n      <td>12886641</td>\n      <td>12942069</td>\n      <td>ISCA-46561</td>\n      <td>1p36.21 population region (DGV_Gold_Standard_J...</td>\n    </tr>\n    <tr>\n      <th>430</th>\n      <td>chr1</td>\n      <td>12896424</td>\n      <td>12930275</td>\n      <td>ISCA-46309</td>\n      <td>1p36.21 population region (gnomAD-SV_v2.1_DEL_...</td>\n    </tr>\n    <tr>\n      <th>179</th>\n      <td>chr1</td>\n      <td>12899395</td>\n      <td>12924283</td>\n      <td>ISCA-46562</td>\n      <td>1p36.21 population region (DGV_Gold_Standard_J...</td>\n    </tr>\n  </tbody>\n</table>\n</div>"
     },
     "execution_count": 213,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "uhi_region = region_pos.loc[\n",
    "    region_pos['Haploinsufficiency Score'] == '40',\n",
    "    ['chrom', 'start', 'end', '#ISCA ID', 'ISCA Region Name']\n",
    "].sort_values(['chrom', 'start', 'end'])\n",
    "uhi_region.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 214,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-05-11T06:01:27.413655300Z",
     "start_time": "2024-05-11T06:01:26.754971200Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": "    chrom     start       end    #ISCA ID  \\\n433  chr1    363999    401750  ISCA-46306   \n432  chr1   1640999   1647000  ISCA-46307   \n180  chr1  12886641  12942069  ISCA-46561   \n430  chr1  12896424  12930275  ISCA-46309   \n179  chr1  12899395  12924283  ISCA-46562   \n\n                                      ISCA Region Name  \\\n433  1p36.33 population region (gnomAD-SV_v2.1_DEL_...   \n432  1p36.33 population region (gnomAD-SV_v2.1_DEL_...   \n180  1p36.21 population region (DGV_Gold_Standard_J...   \n430  1p36.21 population region (gnomAD-SV_v2.1_DEL_...   \n179  1p36.21 population region (DGV_Gold_Standard_J...   \n\n                                 genes  \n433                             OR4F29  \n432                      CDK11B,CDK11A  \n180  PRAMEF11,HNRNPCL1,PRAMEF2,PRAMEF4  \n430                   HNRNPCL1,PRAMEF2  \n179                   HNRNPCL1,PRAMEF2  ",
      "text/html": "<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>chrom</th>\n      <th>start</th>\n      <th>end</th>\n      <th>#ISCA ID</th>\n      <th>ISCA Region Name</th>\n      <th>genes</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>433</th>\n      <td>chr1</td>\n      <td>363999</td>\n      <td>401750</td>\n      <td>ISCA-46306</td>\n      <td>1p36.33 population region (gnomAD-SV_v2.1_DEL_...</td>\n      <td>OR4F29</td>\n    </tr>\n    <tr>\n      <th>432</th>\n      <td>chr1</td>\n      <td>1640999</td>\n      <td>1647000</td>\n      <td>ISCA-46307</td>\n      <td>1p36.33 population region (gnomAD-SV_v2.1_DEL_...</td>\n      <td>CDK11B,CDK11A</td>\n    </tr>\n    <tr>\n      <th>180</th>\n      <td>chr1</td>\n      <td>12886641</td>\n      <td>12942069</td>\n      <td>ISCA-46561</td>\n      <td>1p36.21 population region (DGV_Gold_Standard_J...</td>\n      <td>PRAMEF11,HNRNPCL1,PRAMEF2,PRAMEF4</td>\n    </tr>\n    <tr>\n      <th>430</th>\n      <td>chr1</td>\n      <td>12896424</td>\n      <td>12930275</td>\n      <td>ISCA-46309</td>\n      <td>1p36.21 population region (gnomAD-SV_v2.1_DEL_...</td>\n      <td>HNRNPCL1,PRAMEF2</td>\n    </tr>\n    <tr>\n      <th>179</th>\n      <td>chr1</td>\n      <td>12899395</td>\n      <td>12924283</td>\n      <td>ISCA-46562</td>\n      <td>1p36.21 population region (DGV_Gold_Standard_J...</td>\n      <td>HNRNPCL1,PRAMEF2</td>\n    </tr>\n  </tbody>\n</table>\n</div>"
     },
     "execution_count": 214,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "uhi_region['genes'] = uhi_region.apply(\n",
    "    lambda row: ','.join(gene.loc[\n",
    "        (gene['chrom'] == row['chrom'])\n",
    "        & (gene['txEnd'] >= row['start'])\n",
    "        & (gene['txStart'] <= row['end']),\n",
    "        'name2'\n",
    "    ]),\n",
    "    axis=1\n",
    ")\n",
    "uhi_region.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 215,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-05-11T06:01:27.420638800Z",
     "start_time": "2024-05-11T06:01:27.415637500Z"
    }
   },
   "outputs": [],
   "source": [
    "uhi_region.rename(columns={\n",
    "    'chrom': '#chrom',\n",
    "    '#ISCA ID': 'isca_id', 'ISCA Region Name': 'name'\n",
    "}).to_csv(uhi_region_file, sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 216,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-05-11T06:01:29.462569500Z",
     "start_time": "2024-05-11T06:01:27.496689600Z"
    }
   },
   "outputs": [],
   "source": [
    "!bgzip -cf {uhi_region_file} > {uhi_region_file}.gz\n",
    "!tabix -fp bed {uhi_region_file}.gz\n",
    "!rm {uhi_region_file}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 217,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-05-11T06:01:29.635512400Z",
     "start_time": "2024-05-11T06:01:29.468596200Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": "    symbol     hi_score  hi_index  chrom    txStart      txEnd  GeneID  \\\n0   ANXA2R  0.000047349    100.00   chr5   43039472   43043349  389289   \n1  SCGB1D1  0.000054551     99.99  chr11   61957687   61961011   10648   \n2     IL31  0.000057228     99.99  chr12  122656575  122658768  386653   \n3   BPIFA2  0.000061264     99.98  chr20   31749575   31769218  140683   \n4    MUCL1  0.000067779     99.98  chr12   55248299   55252171  118430   \n\n     name2          name strand  \n0   ANXA2R  NM_001382352      -  \n1  SCGB1D1     NM_006552      +  \n2     IL31  NM_001014336      -  \n3   BPIFA2  NM_001319164      +  \n4    MUCL1     NM_058173      +  ",
      "text/html": "<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>symbol</th>\n      <th>hi_score</th>\n      <th>hi_index</th>\n      <th>chrom</th>\n      <th>txStart</th>\n      <th>txEnd</th>\n      <th>GeneID</th>\n      <th>name2</th>\n      <th>name</th>\n      <th>strand</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>0</th>\n      <td>ANXA2R</td>\n      <td>0.000047349</td>\n      <td>100.00</td>\n      <td>chr5</td>\n      <td>43039472</td>\n      <td>43043349</td>\n      <td>389289</td>\n      <td>ANXA2R</td>\n      <td>NM_001382352</td>\n      <td>-</td>\n    </tr>\n    <tr>\n      <th>1</th>\n      <td>SCGB1D1</td>\n      <td>0.000054551</td>\n      <td>99.99</td>\n      <td>chr11</td>\n      <td>61957687</td>\n      <td>61961011</td>\n      <td>10648</td>\n      <td>SCGB1D1</td>\n      <td>NM_006552</td>\n      <td>+</td>\n    </tr>\n    <tr>\n      <th>2</th>\n      <td>IL31</td>\n      <td>0.000057228</td>\n      <td>99.99</td>\n      <td>chr12</td>\n      <td>122656575</td>\n      <td>122658768</td>\n      <td>386653</td>\n      <td>IL31</td>\n      <td>NM_001014336</td>\n      <td>-</td>\n    </tr>\n    <tr>\n      <th>3</th>\n      <td>BPIFA2</td>\n      <td>0.000061264</td>\n      <td>99.98</td>\n      <td>chr20</td>\n      <td>31749575</td>\n      <td>31769218</td>\n      <td>140683</td>\n      <td>BPIFA2</td>\n      <td>NM_001319164</td>\n      <td>+</td>\n    </tr>\n    <tr>\n      <th>4</th>\n      <td>MUCL1</td>\n      <td>0.000067779</td>\n      <td>99.98</td>\n      <td>chr12</td>\n      <td>55248299</td>\n      <td>55252171</td>\n      <td>118430</td>\n      <td>MUCL1</td>\n      <td>NM_058173</td>\n      <td>+</td>\n    </tr>\n  </tbody>\n</table>\n</div>"
     },
     "execution_count": 217,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "decipher = pd.read_csv(hi_pred_ori_file, sep='\\t',skiprows=1, header=None, usecols=[3,])\n",
    "decipher = decipher[3].str.split('|', expand=True).rename(columns={0: 'symbol', 1: 'hi_score', 2: 'hi_index'})\n",
    "decipher['hi_index'] = decipher['hi_index'].str.replace('%', '').astype(float)\n",
    "decipher = decipher.merge(gene, left_on='symbol', right_on='name2')\n",
    "decipher.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 218,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-05-11T06:01:30.115886400Z",
     "start_time": "2024-05-11T06:01:29.636512900Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": "            transcript  obs_mis  exp_mis   oe_mis    mu_mis  possible_mis  \\\ngene                                                                        \nMED13  ENST00000397786      871  1117.80  0.77921  0.000056         14195   \nNIPBL  ENST00000282516      846  1441.50  0.58688  0.000074         18540   \nSMC3   ENST00000361804      178   630.07  0.28251  0.000032          8109   \nCNOT1  ENST00000317147      561  1295.90  0.43290  0.000069         15670   \nRLF    ENST00000372771      669   972.87  0.68766  0.000047         12682   \n\n       obs_mis_pphen  exp_mis_pphen  oe_mis_pphen  possible_mis_pphen  \\\ngene                                                                    \nMED13          314.0         529.75       0.59273              6708.0   \nNIPBL          158.0         543.10       0.29092              7135.0   \nSMC3            21.0         182.52       0.11506              2197.0   \nCNOT1           51.0         290.68       0.17545              3560.0   \nRLF            107.0         321.14       0.33319              4151.0   \n\n       obs_syn  exp_syn  oe_syn    mu_syn  possible_syn  obs_lof    mu_lof  \\\ngene                                                                         \nMED13      422   387.53  1.0890  0.000019          4248      0.0  0.000005   \nNIPBL      496   495.01  1.0020  0.000025          5211      1.0  0.000009   \nSMC3       215   203.25  1.0578  0.000010          2091      0.0  0.000005   \nCNOT1      470   456.03  1.0306  0.000024          4564      1.0  0.000007   \nRLF        358   352.62  1.0153  0.000017          3482      0.0  0.000004   \n\n       possible_lof  exp_lof  pLI         pNull          pRec    oe_lof  \\\ngene                                                                      \nMED13        1257.0   98.429  1.0  8.943600e-40  1.838300e-16  0.000000   \nNIPBL        1781.0  150.320  1.0  2.977300e-59  3.572400e-24  0.006653   \nSMC3          937.0   79.490  1.0  2.785300e-32  2.191400e-13  0.000000   \nCNOT1        1440.0  125.030  1.0  2.992400e-49  4.562900e-20  0.007998   \nRLF          1024.0   73.222  1.0  8.405500e-30  2.284200e-12  0.000000   \n\n       oe_syn_lower  oe_syn_upper  oe_mis_lower  oe_mis_upper  oe_lof_lower  \\\ngene                                                                          \nMED13         1.005         1.180         0.736         0.824         0.000   \nNIPBL         0.930         1.079         0.554         0.621         0.001   \nSMC3          0.946         1.184         0.249         0.320         0.000   \nCNOT1         0.955         1.112         0.403         0.464         0.002   \nRLF           0.930         1.108         0.645         0.733         0.000   \n\n       oe_lof_upper constraint_flag     syn_z   mis_z    lof_z  \\\ngene                                                             \nMED13         0.030             NaN -1.376500  2.6232   9.1935   \nNIPBL         0.032             NaN -0.035119  5.5737  11.2860   \nSMC3          0.037             NaN -0.647760  6.3999   8.2618   \nCNOT1         0.038             NaN -0.514100  7.2546  10.2790   \nRLF           0.040             NaN -0.225180  3.4620   7.9294   \n\n       oe_lof_upper_rank  oe_lof_upper_bin  oe_lof_upper_bin_6  n_sites  \\\ngene                                                                      \nMED13                0.0               0.0                 0.0      2.0   \nNIPBL                1.0               0.0                 0.0      2.0   \nSMC3                 2.0               0.0                 0.0      8.0   \nCNOT1                3.0               0.0                 0.0      5.0   \nRLF                  4.0               0.0                 0.0      1.0   \n\n       classic_caf    max_af   no_lofs  obs_het_lof  obs_hom_lof   defined  \\\ngene                                                                         \nMED13     0.000012  0.000008  124782.0          3.0          0.0  124785.0   \nNIPBL     0.000012  0.000008  125693.0          3.0          0.0  125696.0   \nSMC3      0.000032  0.000004  125731.0          8.0          0.0  125739.0   \nCNOT1     0.000020  0.000004  125740.0          4.0          0.0  125744.0   \nRLF       0.000004  0.000004  125122.0          1.0          0.0  125123.0   \n\n              p  exp_hom_lof  classic_caf_afr  classic_caf_amr  \\\ngene                                                             \nMED13  0.000012     0.000018         0.000000         0.000000   \nNIPBL  0.000012     0.000018         0.000000         0.000000   \nSMC3   0.000032     0.000127         0.000000         0.000000   \nCNOT1  0.000016     0.000032         0.000000         0.000029   \nRLF    0.000004     0.000002         0.000062         0.000000   \n\n       classic_caf_asj  classic_caf_eas  classic_caf_fin  classic_caf_nfe  \\\ngene                                                                        \nMED13         0.000000         0.000000         0.000093         0.000009   \nNIPBL         0.000099         0.000000         0.000000         0.000000   \nSMC3          0.000099         0.000054         0.000000         0.000044   \nCNOT1         0.000000         0.000055         0.000000         0.000026   \nRLF           0.000000         0.000000         0.000000         0.000000   \n\n       classic_caf_oth  classic_caf_sas     p_afr     p_amr     p_asj  \\\ngene                                                                    \nMED13              0.0         0.000000  0.000000  0.000000  0.000000   \nNIPBL              0.0         0.000065  0.000000  0.000000  0.000099   \nSMC3               0.0         0.000033  0.000000  0.000000  0.000099   \nCNOT1              0.0         0.000000  0.000000  0.000029  0.000000   \nRLF                0.0         0.000000  0.000062  0.000000  0.000000   \n\n          p_eas     p_fin     p_nfe  p_oth     p_sas transcript_type  \\\ngene                                                                   \nMED13  0.000000  0.000093  0.000009    0.0  0.000000  protein_coding   \nNIPBL  0.000000  0.000000  0.000000    0.0  0.000065  protein_coding   \nSMC3   0.000054  0.000000  0.000044    0.0  0.000033  protein_coding   \nCNOT1  0.000054  0.000000  0.000018    0.0  0.000000  protein_coding   \nRLF    0.000000  0.000000  0.000000    0.0  0.000000  protein_coding   \n\n               gene_id  transcript_level  cds_length  num_coding_exons  \\\ngene                                                                     \nMED13  ENSG00000108510                 2        6522                30   \nNIPBL  ENSG00000164190                 2        8412                46   \nSMC3   ENSG00000108055                 2        3651                29   \nCNOT1  ENSG00000125107                 2        7128                48   \nRLF    ENSG00000117000                 2        5742                 8   \n\n            gene_type  gene_length  exac_pLI  exac_obs_lof  exac_exp_lof  \\\ngene                                                                       \nMED13  protein_coding       122678       1.0           0.0        64.393   \nNIPBL  protein_coding       189655       1.0           1.0       110.570   \nSMC3   protein_coding        36946       1.0           0.0        58.523   \nCNOT1  protein_coding       109936       1.0           3.0        90.130   \nRLF    protein_coding        79549       1.0           0.0        43.607   \n\n       exac_oe_lof  brain_expression chromosome  start_position  end_position  \ngene                                                                           \nMED13     0.000000               NaN         17        60019966      60142643  \nNIPBL     0.009044               NaN          5        36876861      37066515  \nSMC3      0.000000               NaN         10       112327449     112364394  \nCNOT1     0.033285               NaN         16        58553855      58663790  \nRLF       0.000000               NaN          1        40627045      40706593  ",
      "text/html": "<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>transcript</th>\n      <th>obs_mis</th>\n      <th>exp_mis</th>\n      <th>oe_mis</th>\n      <th>mu_mis</th>\n      <th>possible_mis</th>\n      <th>obs_mis_pphen</th>\n      <th>exp_mis_pphen</th>\n      <th>oe_mis_pphen</th>\n      <th>possible_mis_pphen</th>\n      <th>obs_syn</th>\n      <th>exp_syn</th>\n      <th>oe_syn</th>\n      <th>mu_syn</th>\n      <th>possible_syn</th>\n      <th>obs_lof</th>\n      <th>mu_lof</th>\n      <th>possible_lof</th>\n      <th>exp_lof</th>\n      <th>pLI</th>\n      <th>pNull</th>\n      <th>pRec</th>\n      <th>oe_lof</th>\n      <th>oe_syn_lower</th>\n      <th>oe_syn_upper</th>\n      <th>oe_mis_lower</th>\n      <th>oe_mis_upper</th>\n      <th>oe_lof_lower</th>\n      <th>oe_lof_upper</th>\n      <th>constraint_flag</th>\n      <th>syn_z</th>\n      <th>mis_z</th>\n      <th>lof_z</th>\n      <th>oe_lof_upper_rank</th>\n      <th>oe_lof_upper_bin</th>\n      <th>oe_lof_upper_bin_6</th>\n      <th>n_sites</th>\n      <th>classic_caf</th>\n      <th>max_af</th>\n      <th>no_lofs</th>\n      <th>obs_het_lof</th>\n      <th>obs_hom_lof</th>\n      <th>defined</th>\n      <th>p</th>\n      <th>exp_hom_lof</th>\n      <th>classic_caf_afr</th>\n      <th>classic_caf_amr</th>\n      <th>classic_caf_asj</th>\n      <th>classic_caf_eas</th>\n      <th>classic_caf_fin</th>\n      <th>classic_caf_nfe</th>\n      <th>classic_caf_oth</th>\n      <th>classic_caf_sas</th>\n      <th>p_afr</th>\n      <th>p_amr</th>\n      <th>p_asj</th>\n      <th>p_eas</th>\n      <th>p_fin</th>\n      <th>p_nfe</th>\n      <th>p_oth</th>\n      <th>p_sas</th>\n      <th>transcript_type</th>\n      <th>gene_id</th>\n      <th>transcript_level</th>\n      <th>cds_length</th>\n      <th>num_coding_exons</th>\n      <th>gene_type</th>\n      <th>gene_length</th>\n      <th>exac_pLI</th>\n      <th>exac_obs_lof</th>\n      <th>exac_exp_lof</th>\n      <th>exac_oe_lof</th>\n      <th>brain_expression</th>\n      <th>chromosome</th>\n      <th>start_position</th>\n      <th>end_position</th>\n    </tr>\n    <tr>\n      <th>gene</th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n      <th></th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>MED13</th>\n      <td>ENST00000397786</td>\n      <td>871</td>\n      <td>1117.80</td>\n      <td>0.77921</td>\n      <td>0.000056</td>\n      <td>14195</td>\n      <td>314.0</td>\n      <td>529.75</td>\n      <td>0.59273</td>\n      <td>6708.0</td>\n      <td>422</td>\n      <td>387.53</td>\n      <td>1.0890</td>\n      <td>0.000019</td>\n      <td>4248</td>\n      <td>0.0</td>\n      <td>0.000005</td>\n      <td>1257.0</td>\n      <td>98.429</td>\n      <td>1.0</td>\n      <td>8.943600e-40</td>\n      <td>1.838300e-16</td>\n      <td>0.000000</td>\n      <td>1.005</td>\n      <td>1.180</td>\n      <td>0.736</td>\n      <td>0.824</td>\n      <td>0.000</td>\n      <td>0.030</td>\n      <td>NaN</td>\n      <td>-1.376500</td>\n      <td>2.6232</td>\n      <td>9.1935</td>\n      <td>0.0</td>\n      <td>0.0</td>\n      <td>0.0</td>\n      <td>2.0</td>\n      <td>0.000012</td>\n      <td>0.000008</td>\n      <td>124782.0</td>\n      <td>3.0</td>\n      <td>0.0</td>\n      <td>124785.0</td>\n      <td>0.000012</td>\n      <td>0.000018</td>\n      <td>0.000000</td>\n      <td>0.000000</td>\n      <td>0.000000</td>\n      <td>0.000000</td>\n      <td>0.000093</td>\n      <td>0.000009</td>\n      <td>0.0</td>\n      <td>0.000000</td>\n      <td>0.000000</td>\n      <td>0.000000</td>\n      <td>0.000000</td>\n      <td>0.000000</td>\n      <td>0.000093</td>\n      <td>0.000009</td>\n      <td>0.0</td>\n      <td>0.000000</td>\n      <td>protein_coding</td>\n      <td>ENSG00000108510</td>\n      <td>2</td>\n      <td>6522</td>\n      <td>30</td>\n      <td>protein_coding</td>\n      <td>122678</td>\n      <td>1.0</td>\n      <td>0.0</td>\n      <td>64.393</td>\n      <td>0.000000</td>\n      <td>NaN</td>\n      <td>17</td>\n      <td>60019966</td>\n      <td>60142643</td>\n    </tr>\n    <tr>\n      <th>NIPBL</th>\n      <td>ENST00000282516</td>\n      <td>846</td>\n      <td>1441.50</td>\n      <td>0.58688</td>\n      <td>0.000074</td>\n      <td>18540</td>\n      <td>158.0</td>\n      <td>543.10</td>\n      <td>0.29092</td>\n      <td>7135.0</td>\n      <td>496</td>\n      <td>495.01</td>\n      <td>1.0020</td>\n      <td>0.000025</td>\n      <td>5211</td>\n      <td>1.0</td>\n      <td>0.000009</td>\n      <td>1781.0</td>\n      <td>150.320</td>\n      <td>1.0</td>\n      <td>2.977300e-59</td>\n      <td>3.572400e-24</td>\n      <td>0.006653</td>\n      <td>0.930</td>\n      <td>1.079</td>\n      <td>0.554</td>\n      <td>0.621</td>\n      <td>0.001</td>\n      <td>0.032</td>\n      <td>NaN</td>\n      <td>-0.035119</td>\n      <td>5.5737</td>\n      <td>11.2860</td>\n      <td>1.0</td>\n      <td>0.0</td>\n      <td>0.0</td>\n      <td>2.0</td>\n      <td>0.000012</td>\n      <td>0.000008</td>\n      <td>125693.0</td>\n      <td>3.0</td>\n      <td>0.0</td>\n      <td>125696.0</td>\n      <td>0.000012</td>\n      <td>0.000018</td>\n      <td>0.000000</td>\n      <td>0.000000</td>\n      <td>0.000099</td>\n      <td>0.000000</td>\n      <td>0.000000</td>\n      <td>0.000000</td>\n      <td>0.0</td>\n      <td>0.000065</td>\n      <td>0.000000</td>\n      <td>0.000000</td>\n      <td>0.000099</td>\n      <td>0.000000</td>\n      <td>0.000000</td>\n      <td>0.000000</td>\n      <td>0.0</td>\n      <td>0.000065</td>\n      <td>protein_coding</td>\n      <td>ENSG00000164190</td>\n      <td>2</td>\n      <td>8412</td>\n      <td>46</td>\n      <td>protein_coding</td>\n      <td>189655</td>\n      <td>1.0</td>\n      <td>1.0</td>\n      <td>110.570</td>\n      <td>0.009044</td>\n      <td>NaN</td>\n      <td>5</td>\n      <td>36876861</td>\n      <td>37066515</td>\n    </tr>\n    <tr>\n      <th>SMC3</th>\n      <td>ENST00000361804</td>\n      <td>178</td>\n      <td>630.07</td>\n      <td>0.28251</td>\n      <td>0.000032</td>\n      <td>8109</td>\n      <td>21.0</td>\n      <td>182.52</td>\n      <td>0.11506</td>\n      <td>2197.0</td>\n      <td>215</td>\n      <td>203.25</td>\n      <td>1.0578</td>\n      <td>0.000010</td>\n      <td>2091</td>\n      <td>0.0</td>\n      <td>0.000005</td>\n      <td>937.0</td>\n      <td>79.490</td>\n      <td>1.0</td>\n      <td>2.785300e-32</td>\n      <td>2.191400e-13</td>\n      <td>0.000000</td>\n      <td>0.946</td>\n      <td>1.184</td>\n      <td>0.249</td>\n      <td>0.320</td>\n      <td>0.000</td>\n      <td>0.037</td>\n      <td>NaN</td>\n      <td>-0.647760</td>\n      <td>6.3999</td>\n      <td>8.2618</td>\n      <td>2.0</td>\n      <td>0.0</td>\n      <td>0.0</td>\n      <td>8.0</td>\n      <td>0.000032</td>\n      <td>0.000004</td>\n      <td>125731.0</td>\n      <td>8.0</td>\n      <td>0.0</td>\n      <td>125739.0</td>\n      <td>0.000032</td>\n      <td>0.000127</td>\n      <td>0.000000</td>\n      <td>0.000000</td>\n      <td>0.000099</td>\n      <td>0.000054</td>\n      <td>0.000000</td>\n      <td>0.000044</td>\n      <td>0.0</td>\n      <td>0.000033</td>\n      <td>0.000000</td>\n      <td>0.000000</td>\n      <td>0.000099</td>\n      <td>0.000054</td>\n      <td>0.000000</td>\n      <td>0.000044</td>\n      <td>0.0</td>\n      <td>0.000033</td>\n      <td>protein_coding</td>\n      <td>ENSG00000108055</td>\n      <td>2</td>\n      <td>3651</td>\n      <td>29</td>\n      <td>protein_coding</td>\n      <td>36946</td>\n      <td>1.0</td>\n      <td>0.0</td>\n      <td>58.523</td>\n      <td>0.000000</td>\n      <td>NaN</td>\n      <td>10</td>\n      <td>112327449</td>\n      <td>112364394</td>\n    </tr>\n    <tr>\n      <th>CNOT1</th>\n      <td>ENST00000317147</td>\n      <td>561</td>\n      <td>1295.90</td>\n      <td>0.43290</td>\n      <td>0.000069</td>\n      <td>15670</td>\n      <td>51.0</td>\n      <td>290.68</td>\n      <td>0.17545</td>\n      <td>3560.0</td>\n      <td>470</td>\n      <td>456.03</td>\n      <td>1.0306</td>\n      <td>0.000024</td>\n      <td>4564</td>\n      <td>1.0</td>\n      <td>0.000007</td>\n      <td>1440.0</td>\n      <td>125.030</td>\n      <td>1.0</td>\n      <td>2.992400e-49</td>\n      <td>4.562900e-20</td>\n      <td>0.007998</td>\n      <td>0.955</td>\n      <td>1.112</td>\n      <td>0.403</td>\n      <td>0.464</td>\n      <td>0.002</td>\n      <td>0.038</td>\n      <td>NaN</td>\n      <td>-0.514100</td>\n      <td>7.2546</td>\n      <td>10.2790</td>\n      <td>3.0</td>\n      <td>0.0</td>\n      <td>0.0</td>\n      <td>5.0</td>\n      <td>0.000020</td>\n      <td>0.000004</td>\n      <td>125740.0</td>\n      <td>4.0</td>\n      <td>0.0</td>\n      <td>125744.0</td>\n      <td>0.000016</td>\n      <td>0.000032</td>\n      <td>0.000000</td>\n      <td>0.000029</td>\n      <td>0.000000</td>\n      <td>0.000055</td>\n      <td>0.000000</td>\n      <td>0.000026</td>\n      <td>0.0</td>\n      <td>0.000000</td>\n      <td>0.000000</td>\n      <td>0.000029</td>\n      <td>0.000000</td>\n      <td>0.000054</td>\n      <td>0.000000</td>\n      <td>0.000018</td>\n      <td>0.0</td>\n      <td>0.000000</td>\n      <td>protein_coding</td>\n      <td>ENSG00000125107</td>\n      <td>2</td>\n      <td>7128</td>\n      <td>48</td>\n      <td>protein_coding</td>\n      <td>109936</td>\n      <td>1.0</td>\n      <td>3.0</td>\n      <td>90.130</td>\n      <td>0.033285</td>\n      <td>NaN</td>\n      <td>16</td>\n      <td>58553855</td>\n      <td>58663790</td>\n    </tr>\n    <tr>\n      <th>RLF</th>\n      <td>ENST00000372771</td>\n      <td>669</td>\n      <td>972.87</td>\n      <td>0.68766</td>\n      <td>0.000047</td>\n      <td>12682</td>\n      <td>107.0</td>\n      <td>321.14</td>\n      <td>0.33319</td>\n      <td>4151.0</td>\n      <td>358</td>\n      <td>352.62</td>\n      <td>1.0153</td>\n      <td>0.000017</td>\n      <td>3482</td>\n      <td>0.0</td>\n      <td>0.000004</td>\n      <td>1024.0</td>\n      <td>73.222</td>\n      <td>1.0</td>\n      <td>8.405500e-30</td>\n      <td>2.284200e-12</td>\n      <td>0.000000</td>\n      <td>0.930</td>\n      <td>1.108</td>\n      <td>0.645</td>\n      <td>0.733</td>\n      <td>0.000</td>\n      <td>0.040</td>\n      <td>NaN</td>\n      <td>-0.225180</td>\n      <td>3.4620</td>\n      <td>7.9294</td>\n      <td>4.0</td>\n      <td>0.0</td>\n      <td>0.0</td>\n      <td>1.0</td>\n      <td>0.000004</td>\n      <td>0.000004</td>\n      <td>125122.0</td>\n      <td>1.0</td>\n      <td>0.0</td>\n      <td>125123.0</td>\n      <td>0.000004</td>\n      <td>0.000002</td>\n      <td>0.000062</td>\n      <td>0.000000</td>\n      <td>0.000000</td>\n      <td>0.000000</td>\n      <td>0.000000</td>\n      <td>0.000000</td>\n      <td>0.0</td>\n      <td>0.000000</td>\n      <td>0.000062</td>\n      <td>0.000000</td>\n      <td>0.000000</td>\n      <td>0.000000</td>\n      <td>0.000000</td>\n      <td>0.000000</td>\n      <td>0.0</td>\n      <td>0.000000</td>\n      <td>protein_coding</td>\n      <td>ENSG00000117000</td>\n      <td>2</td>\n      <td>5742</td>\n      <td>8</td>\n      <td>protein_coding</td>\n      <td>79549</td>\n      <td>1.0</td>\n      <td>0.0</td>\n      <td>43.607</td>\n      <td>0.000000</td>\n      <td>NaN</td>\n      <td>1</td>\n      <td>40627045</td>\n      <td>40706593</td>\n    </tr>\n  </tbody>\n</table>\n</div>"
     },
     "execution_count": 218,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gnomad = pd.read_csv(gnomad_lof_ori_file, sep='\\t', index_col=0, compression='gzip')\n",
    "gnomad.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 219,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-05-11T06:01:30.132803800Z",
     "start_time": "2024-05-11T06:01:30.115886400Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": "    symbol     hi_score  hi_index  chrom    txStart      txEnd  GeneID  \\\n0   ANXA2R  0.000047349    100.00   chr5   43039472   43043349  389289   \n1  SCGB1D1  0.000054551     99.99  chr11   61957687   61961011   10648   \n2     IL31  0.000057228     99.99  chr12  122656575  122658768  386653   \n3   BPIFA2  0.000061264     99.98  chr20   31749575   31769218  140683   \n4    MUCL1  0.000067779     99.98  chr12   55248299   55252171  118430   \n\n     name2          name strand           pLI  oe_lof_upper  \n0   ANXA2R  NM_001382352      -           NaN           NaN  \n1  SCGB1D1     NM_006552      +  1.294000e-07         1.968  \n2     IL31  NM_001014336      -  1.144500e-02         1.715  \n3   BPIFA2  NM_001319164      +  8.012400e-07         1.495  \n4    MUCL1     NM_058173      +  3.092300e-05         1.908  ",
      "text/html": "<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>symbol</th>\n      <th>hi_score</th>\n      <th>hi_index</th>\n      <th>chrom</th>\n      <th>txStart</th>\n      <th>txEnd</th>\n      <th>GeneID</th>\n      <th>name2</th>\n      <th>name</th>\n      <th>strand</th>\n      <th>pLI</th>\n      <th>oe_lof_upper</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>0</th>\n      <td>ANXA2R</td>\n      <td>0.000047349</td>\n      <td>100.00</td>\n      <td>chr5</td>\n      <td>43039472</td>\n      <td>43043349</td>\n      <td>389289</td>\n      <td>ANXA2R</td>\n      <td>NM_001382352</td>\n      <td>-</td>\n      <td>NaN</td>\n      <td>NaN</td>\n    </tr>\n    <tr>\n      <th>1</th>\n      <td>SCGB1D1</td>\n      <td>0.000054551</td>\n      <td>99.99</td>\n      <td>chr11</td>\n      <td>61957687</td>\n      <td>61961011</td>\n      <td>10648</td>\n      <td>SCGB1D1</td>\n      <td>NM_006552</td>\n      <td>+</td>\n      <td>1.294000e-07</td>\n      <td>1.968</td>\n    </tr>\n    <tr>\n      <th>2</th>\n      <td>IL31</td>\n      <td>0.000057228</td>\n      <td>99.99</td>\n      <td>chr12</td>\n      <td>122656575</td>\n      <td>122658768</td>\n      <td>386653</td>\n      <td>IL31</td>\n      <td>NM_001014336</td>\n      <td>-</td>\n      <td>1.144500e-02</td>\n      <td>1.715</td>\n    </tr>\n    <tr>\n      <th>3</th>\n      <td>BPIFA2</td>\n      <td>0.000061264</td>\n      <td>99.98</td>\n      <td>chr20</td>\n      <td>31749575</td>\n      <td>31769218</td>\n      <td>140683</td>\n      <td>BPIFA2</td>\n      <td>NM_001319164</td>\n      <td>+</td>\n      <td>8.012400e-07</td>\n      <td>1.495</td>\n    </tr>\n    <tr>\n      <th>4</th>\n      <td>MUCL1</td>\n      <td>0.000067779</td>\n      <td>99.98</td>\n      <td>chr12</td>\n      <td>55248299</td>\n      <td>55252171</td>\n      <td>118430</td>\n      <td>MUCL1</td>\n      <td>NM_058173</td>\n      <td>+</td>\n      <td>3.092300e-05</td>\n      <td>1.908</td>\n    </tr>\n  </tbody>\n</table>\n</div>"
     },
     "execution_count": 219,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "decipher = decipher.join(gnomad['pLI'], on='name2')\n",
    "decipher = decipher.join(gnomad['oe_lof_upper'], on='name2')\n",
    "decipher.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 220,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-05-11T06:01:30.258668Z",
     "start_time": "2024-05-11T06:01:30.133802400Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": "      chrom   txStart     txEnd  GeneID   name2          name      pLI  \\\n17369  chr1   6845513   7829766   23261  CAMTA1  NM_001349609  1.00000   \n17427  chr1   8412463   8877699     473    RERE  NM_001042681  1.00000   \n17498  chr1   9751524   9788980    5293  PIK3CD  NM_001350235  0.99999   \n16243  chr1  10270763  10441661   23095   KIF1B     NM_015074  1.00000   \n17675  chr1  11166591  11322608    2475    MTOR     NM_004958  1.00000   \n\n          hi_score  \n17369  0.951338773  \n17427  0.962236687  \n17498  0.977549203  \n16243   0.70541308  \n17675  0.997612705  ",
      "text/html": "<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>chrom</th>\n      <th>txStart</th>\n      <th>txEnd</th>\n      <th>GeneID</th>\n      <th>name2</th>\n      <th>name</th>\n      <th>pLI</th>\n      <th>hi_score</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>17369</th>\n      <td>chr1</td>\n      <td>6845513</td>\n      <td>7829766</td>\n      <td>23261</td>\n      <td>CAMTA1</td>\n      <td>NM_001349609</td>\n      <td>1.00000</td>\n      <td>0.951338773</td>\n    </tr>\n    <tr>\n      <th>17427</th>\n      <td>chr1</td>\n      <td>8412463</td>\n      <td>8877699</td>\n      <td>473</td>\n      <td>RERE</td>\n      <td>NM_001042681</td>\n      <td>1.00000</td>\n      <td>0.962236687</td>\n    </tr>\n    <tr>\n      <th>17498</th>\n      <td>chr1</td>\n      <td>9751524</td>\n      <td>9788980</td>\n      <td>5293</td>\n      <td>PIK3CD</td>\n      <td>NM_001350235</td>\n      <td>0.99999</td>\n      <td>0.977549203</td>\n    </tr>\n    <tr>\n      <th>16243</th>\n      <td>chr1</td>\n      <td>10270763</td>\n      <td>10441661</td>\n      <td>23095</td>\n      <td>KIF1B</td>\n      <td>NM_015074</td>\n      <td>1.00000</td>\n      <td>0.70541308</td>\n    </tr>\n    <tr>\n      <th>17675</th>\n      <td>chr1</td>\n      <td>11166591</td>\n      <td>11322608</td>\n      <td>2475</td>\n      <td>MTOR</td>\n      <td>NM_004958</td>\n      <td>1.00000</td>\n      <td>0.997612705</td>\n    </tr>\n  </tbody>\n</table>\n</div>"
     },
     "execution_count": 220,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "decipher = decipher.loc[\n",
    "    (decipher['pLI'] >= 0.9) & (decipher['hi_index'] < 10) & (decipher['oe_lof_upper'] < 0.35),\n",
    "    ['chrom', 'txStart', 'txEnd', 'GeneID', 'name2', 'name', 'pLI', 'hi_score']\n",
    "].sort_values(['chrom', 'txStart', 'txEnd'])\n",
    "decipher.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 221,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-05-11T06:01:30.258668Z",
     "start_time": "2024-05-11T06:01:30.187227300Z"
    }
   },
   "outputs": [],
   "source": [
    "decipher.rename(columns={\n",
    "    'chrom': '#chrom', 'txStart': 'start', 'txEnd': 'end',\n",
    "    'GeneID': 'gene_id', 'name2': 'symbol', 'name': 'transcript'\n",
    "}).to_csv(decipher_gene_file, sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 222,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-05-11T06:01:32.253827500Z",
     "start_time": "2024-05-11T06:01:30.235599800Z"
    }
   },
   "outputs": [],
   "source": [
    "!bgzip -cf {decipher_gene_file} > {decipher_gene_file}.gz\n",
    "!tabix -fp bed {decipher_gene_file}.gz\n",
    "!rm {decipher_gene_file}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 223,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-05-11T06:01:32.272798300Z",
     "start_time": "2024-05-11T06:01:32.257812900Z"
    }
   },
   "outputs": [],
   "source": [
    "ts_genes = set(\n",
    "    curation_gene.loc[\n",
    "        curation_gene['Triplosensitivity Score'] == '3', '#Gene Symbol'\n",
    "    ]\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 224,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-05-11T06:01:32.299726300Z",
     "start_time": "2024-05-11T06:01:32.293742200Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": "      chrom    txStart      txEnd  GeneID  name2          name strand\n4327   chr5  126112827  126172712    4001  LMNB1     NM_005573      +\n10786  chrX  103031780  103047547    5354   PLP1  NM_001305004      +",
      "text/html": "<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>chrom</th>\n      <th>txStart</th>\n      <th>txEnd</th>\n      <th>GeneID</th>\n      <th>name2</th>\n      <th>name</th>\n      <th>strand</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>4327</th>\n      <td>chr5</td>\n      <td>126112827</td>\n      <td>126172712</td>\n      <td>4001</td>\n      <td>LMNB1</td>\n      <td>NM_005573</td>\n      <td>+</td>\n    </tr>\n    <tr>\n      <th>10786</th>\n      <td>chrX</td>\n      <td>103031780</td>\n      <td>103047547</td>\n      <td>5354</td>\n      <td>PLP1</td>\n      <td>NM_001305004</td>\n      <td>+</td>\n    </tr>\n  </tbody>\n</table>\n</div>"
     },
     "execution_count": 224,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ts_gene = refgene_info.loc[\n",
    "    refgene_info['name2'].isin(ts_genes),\n",
    "    ['chrom', 'txStart', 'txEnd', 'GeneID', 'name2', 'name', 'strand']\n",
    "].sort_values(['chrom', 'txStart', 'txEnd'])\n",
    "ts_gene.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 225,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-05-11T06:01:32.318675600Z",
     "start_time": "2024-05-11T06:01:32.304713Z"
    }
   },
   "outputs": [],
   "source": [
    "ts_gene.rename(columns={\n",
    "    'chrom': '#chrom', 'txStart': 'start', 'txEnd': 'end',\n",
    "    'GeneID': 'gene_id', 'name2': 'symbol', 'name': 'transcript'\n",
    "}).to_csv(ts_gene_file, index=False, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 226,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-05-11T06:01:34.418172400Z",
     "start_time": "2024-05-11T06:01:32.412176200Z"
    }
   },
   "outputs": [],
   "source": [
    "!bgzip -cf {ts_gene_file} > {ts_gene_file}.gz\n",
    "!tabix -fp bed {ts_gene_file}.gz\n",
    "!rm {ts_gene_file}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 227,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-05-11T06:01:34.420163200Z",
     "start_time": "2024-05-11T06:01:34.409193100Z"
    }
   },
   "outputs": [],
   "source": [
    "uts_genes = set(\n",
    "    curation_gene.loc[curation_gene['Triplosensitivity Score'] == '40', '#Gene Symbol']\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 228,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-05-11T06:01:34.497166Z",
     "start_time": "2024-05-11T06:01:34.425150600Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": "       chrom    txStart      txEnd  GeneID  name2       name strand\n7271   chr13  103498381  103528345    2073  ERCC5  NM_000123      +\n9351   chr19   45853094   45873831    2068  ERCC2  NM_000400      -\n18393   chrX    6451658    6453159   51481  VCX3A  NM_016379      -",
      "text/html": "<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>chrom</th>\n      <th>txStart</th>\n      <th>txEnd</th>\n      <th>GeneID</th>\n      <th>name2</th>\n      <th>name</th>\n      <th>strand</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>7271</th>\n      <td>chr13</td>\n      <td>103498381</td>\n      <td>103528345</td>\n      <td>2073</td>\n      <td>ERCC5</td>\n      <td>NM_000123</td>\n      <td>+</td>\n    </tr>\n    <tr>\n      <th>9351</th>\n      <td>chr19</td>\n      <td>45853094</td>\n      <td>45873831</td>\n      <td>2068</td>\n      <td>ERCC2</td>\n      <td>NM_000400</td>\n      <td>-</td>\n    </tr>\n    <tr>\n      <th>18393</th>\n      <td>chrX</td>\n      <td>6451658</td>\n      <td>6453159</td>\n      <td>51481</td>\n      <td>VCX3A</td>\n      <td>NM_016379</td>\n      <td>-</td>\n    </tr>\n  </tbody>\n</table>\n</div>"
     },
     "execution_count": 228,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "uts_gene = refgene_info.loc[\n",
    "    refgene_info['name2'].isin(uts_genes),\n",
    "    ['chrom', 'txStart', 'txEnd', 'GeneID', 'name2', 'name', 'strand']\n",
    "].sort_values(['chrom', 'txStart', 'txEnd'])\n",
    "uts_gene.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 229,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-05-11T06:01:34.530086900Z",
     "start_time": "2024-05-11T06:01:34.465261100Z"
    }
   },
   "outputs": [],
   "source": [
    "uts_gene.rename(columns={\n",
    "    'chrom': '#chrom', 'txStart': 'start', 'txEnd': 'end',\n",
    "    'GeneID': 'gene_id', 'name2': 'symbol', 'name': 'transcript'\n",
    "}).to_csv(uts_gene_file, index=False, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 230,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-05-11T06:01:36.510137600Z",
     "start_time": "2024-05-11T06:01:34.474202200Z"
    }
   },
   "outputs": [],
   "source": [
    "!bgzip -cf {uts_gene_file} > {uts_gene_file}.gz\n",
    "!tabix -fp bed {uts_gene_file}.gz\n",
    "!rm {uts_gene_file}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 231,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-05-11T06:01:36.547033900Z",
     "start_time": "2024-05-11T06:01:36.511130900Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": "     chrom      start        end    #ISCA ID  \\\n494   chr1  146577486  147394506  ISCA-37421   \n503  chr15   22832519   28379874  ISCA-37404   \n465  chr15   23747996   28379874  ISCA-37478   \n505  chr16   29649997   30199852  ISCA-37400   \n488  chr17    1247833    2588909  ISCA-37430   \n\n                                      ISCA Region Name  \n494  1q21.1 recurrent region (BP3-BP4, distal) (inc...  \n503  15q11.2q13 recurrent (PWS/AS) region (Class 1,...  \n465  15q11.2q13 recurrent (PWS/AS) region (Class 2,...  \n505  16p11.2 recurrent region (proximal, BP4-BP5) (...  \n488  17p13.3 (Miller-Dieker syndrome) region (inclu...  ",
      "text/html": "<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>chrom</th>\n      <th>start</th>\n      <th>end</th>\n      <th>#ISCA ID</th>\n      <th>ISCA Region Name</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>494</th>\n      <td>chr1</td>\n      <td>146577486</td>\n      <td>147394506</td>\n      <td>ISCA-37421</td>\n      <td>1q21.1 recurrent region (BP3-BP4, distal) (inc...</td>\n    </tr>\n    <tr>\n      <th>503</th>\n      <td>chr15</td>\n      <td>22832519</td>\n      <td>28379874</td>\n      <td>ISCA-37404</td>\n      <td>15q11.2q13 recurrent (PWS/AS) region (Class 1,...</td>\n    </tr>\n    <tr>\n      <th>465</th>\n      <td>chr15</td>\n      <td>23747996</td>\n      <td>28379874</td>\n      <td>ISCA-37478</td>\n      <td>15q11.2q13 recurrent (PWS/AS) region (Class 2,...</td>\n    </tr>\n    <tr>\n      <th>505</th>\n      <td>chr16</td>\n      <td>29649997</td>\n      <td>30199852</td>\n      <td>ISCA-37400</td>\n      <td>16p11.2 recurrent region (proximal, BP4-BP5) (...</td>\n    </tr>\n    <tr>\n      <th>488</th>\n      <td>chr17</td>\n      <td>1247833</td>\n      <td>2588909</td>\n      <td>ISCA-37430</td>\n      <td>17p13.3 (Miller-Dieker syndrome) region (inclu...</td>\n    </tr>\n  </tbody>\n</table>\n</div>"
     },
     "execution_count": 231,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ts_region = region_pos.loc[\n",
    "    region_pos['Triplosensitivity Score'] == '3',\n",
    "    ['chrom', 'start', 'end', '#ISCA ID', 'ISCA Region Name']\n",
    "].sort_values(['chrom', 'start', 'end'])\n",
    "ts_region.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 232,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-05-11T06:01:36.764523600Z",
     "start_time": "2024-05-11T06:01:36.541022400Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": "     chrom      start        end    #ISCA ID  \\\n494   chr1  146577486  147394506  ISCA-37421   \n503  chr15   22832519   28379874  ISCA-37404   \n465  chr15   23747996   28379874  ISCA-37478   \n505  chr16   29649997   30199852  ISCA-37400   \n488  chr17    1247833    2588909  ISCA-37430   \n\n                                      ISCA Region Name  \\\n494  1q21.1 recurrent region (BP3-BP4, distal) (inc...   \n503  15q11.2q13 recurrent (PWS/AS) region (Class 1,...   \n465  15q11.2q13 recurrent (PWS/AS) region (Class 2,...   \n505  16p11.2 recurrent region (proximal, BP4-BP5) (...   \n488  17p13.3 (Miller-Dieker syndrome) region (inclu...   \n\n                                            omim_genes  \n494              PRKAB2,FMO5,CHD1L,BCL9,ACP6,GJA5,GJA8  \n503  TUBGCP5,CYFIP1,NIPA2,NIPA1,MKRN3,MAGEL2,NDN,PW...  \n465  MKRN3,MAGEL2,NDN,PWRN2,PWRN1,NPAP1,SNHG14,SNRP...  \n505  SPN,QPRT,ZG16,KIF22,MAZ,PRRT2,PAGR1,MVP,CDIPT,...  \n488  YWHAE,CRK,MYO1C,INPP5K,PITPNA,SLC43A2,SCARF1,R...  ",
      "text/html": "<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>chrom</th>\n      <th>start</th>\n      <th>end</th>\n      <th>#ISCA ID</th>\n      <th>ISCA Region Name</th>\n      <th>omim_genes</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>494</th>\n      <td>chr1</td>\n      <td>146577486</td>\n      <td>147394506</td>\n      <td>ISCA-37421</td>\n      <td>1q21.1 recurrent region (BP3-BP4, distal) (inc...</td>\n      <td>PRKAB2,FMO5,CHD1L,BCL9,ACP6,GJA5,GJA8</td>\n    </tr>\n    <tr>\n      <th>503</th>\n      <td>chr15</td>\n      <td>22832519</td>\n      <td>28379874</td>\n      <td>ISCA-37404</td>\n      <td>15q11.2q13 recurrent (PWS/AS) region (Class 1,...</td>\n      <td>TUBGCP5,CYFIP1,NIPA2,NIPA1,MKRN3,MAGEL2,NDN,PW...</td>\n    </tr>\n    <tr>\n      <th>465</th>\n      <td>chr15</td>\n      <td>23747996</td>\n      <td>28379874</td>\n      <td>ISCA-37478</td>\n      <td>15q11.2q13 recurrent (PWS/AS) region (Class 2,...</td>\n      <td>MKRN3,MAGEL2,NDN,PWRN2,PWRN1,NPAP1,SNHG14,SNRP...</td>\n    </tr>\n    <tr>\n      <th>505</th>\n      <td>chr16</td>\n      <td>29649997</td>\n      <td>30199852</td>\n      <td>ISCA-37400</td>\n      <td>16p11.2 recurrent region (proximal, BP4-BP5) (...</td>\n      <td>SPN,QPRT,ZG16,KIF22,MAZ,PRRT2,PAGR1,MVP,CDIPT,...</td>\n    </tr>\n    <tr>\n      <th>488</th>\n      <td>chr17</td>\n      <td>1247833</td>\n      <td>2588909</td>\n      <td>ISCA-37430</td>\n      <td>17p13.3 (Miller-Dieker syndrome) region (inclu...</td>\n      <td>YWHAE,CRK,MYO1C,INPP5K,PITPNA,SLC43A2,SCARF1,R...</td>\n    </tr>\n  </tbody>\n</table>\n</div>"
     },
     "execution_count": 232,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ts_region['omim_genes'] = ts_region.apply(\n",
    "    lambda row: ','.join(omim_gene.loc[\n",
    "        (omim_gene['chrom'] == row['chrom'])\n",
    "        & (omim_gene['txEnd'] >= row['start'])\n",
    "        & (omim_gene['txStart'] <= row['end']),\n",
    "        'name2'\n",
    "    ]),\n",
    "    axis=1\n",
    ")\n",
    "ts_region.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 233,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-05-11T06:01:36.780479800Z",
     "start_time": "2024-05-11T06:01:36.631871500Z"
    }
   },
   "outputs": [],
   "source": [
    "ts_region.rename(columns={\n",
    "    'chrom': '#chrom',\n",
    "    '#ISCA ID': 'isca_id', 'ISCA Region Name': 'name'\n",
    "}).to_csv(ts_region_file, sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 234,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-05-11T06:01:38.649881200Z",
     "start_time": "2024-05-11T06:01:36.697696500Z"
    }
   },
   "outputs": [],
   "source": [
    "!bgzip -cf {ts_region_file} > {ts_region_file}.gz\n",
    "!tabix -fp bed {ts_region_file}.gz\n",
    "!rm {ts_region_file}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 235,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-05-11T06:01:38.708126500Z",
     "start_time": "2024-05-11T06:01:38.662853100Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": "    chrom      start        end    #ISCA ID  \\\n185  chr1     713039     756009  ISCA-46556   \n186  chr1     713039     756009  ISCA-46554   \n431  chr1   12894999   12940000  ISCA-46308   \n184  chr1   12897609   12922238  ISCA-46557   \n418  chr1  104190999  104303200  ISCA-46321   \n\n                                      ISCA Region Name  \n185  1p36.33 population region (DGV_Gold_Standard_J...  \n186  1p36.33 population region ((DGV_Gold_Standard_...  \n431  1p36.21 population region (gnomAD-SV_v2.1_DUP_...  \n184  1p36.21 population region (DGV_Gold_Standard_J...  \n418  1p21.1 population region (gnomAD-SV_v2.1_DUP_1...  ",
      "text/html": "<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>chrom</th>\n      <th>start</th>\n      <th>end</th>\n      <th>#ISCA ID</th>\n      <th>ISCA Region Name</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>185</th>\n      <td>chr1</td>\n      <td>713039</td>\n      <td>756009</td>\n      <td>ISCA-46556</td>\n      <td>1p36.33 population region (DGV_Gold_Standard_J...</td>\n    </tr>\n    <tr>\n      <th>186</th>\n      <td>chr1</td>\n      <td>713039</td>\n      <td>756009</td>\n      <td>ISCA-46554</td>\n      <td>1p36.33 population region ((DGV_Gold_Standard_...</td>\n    </tr>\n    <tr>\n      <th>431</th>\n      <td>chr1</td>\n      <td>12894999</td>\n      <td>12940000</td>\n      <td>ISCA-46308</td>\n      <td>1p36.21 population region (gnomAD-SV_v2.1_DUP_...</td>\n    </tr>\n    <tr>\n      <th>184</th>\n      <td>chr1</td>\n      <td>12897609</td>\n      <td>12922238</td>\n      <td>ISCA-46557</td>\n      <td>1p36.21 population region (DGV_Gold_Standard_J...</td>\n    </tr>\n    <tr>\n      <th>418</th>\n      <td>chr1</td>\n      <td>104190999</td>\n      <td>104303200</td>\n      <td>ISCA-46321</td>\n      <td>1p21.1 population region (gnomAD-SV_v2.1_DUP_1...</td>\n    </tr>\n  </tbody>\n</table>\n</div>"
     },
     "execution_count": 235,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "uts_region = region_pos.loc[\n",
    "    region_pos['Triplosensitivity Score'] == '40',\n",
    "    ['chrom', 'start', 'end', '#ISCA ID', 'ISCA Region Name']\n",
    "].sort_values(['chrom', 'start', 'end'])\n",
    "uts_region.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 236,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-05-11T06:01:38.893166100Z",
     "start_time": "2024-05-11T06:01:38.684022400Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": "    chrom      start        end    #ISCA ID  \\\n185  chr1     713039     756009  ISCA-46556   \n186  chr1     713039     756009  ISCA-46554   \n431  chr1   12894999   12940000  ISCA-46308   \n184  chr1   12897609   12922238  ISCA-46557   \n418  chr1  104190999  104303200  ISCA-46321   \n\n                                      ISCA Region Name  \\\n185  1p36.33 population region (DGV_Gold_Standard_J...   \n186  1p36.33 population region ((DGV_Gold_Standard_...   \n431  1p36.21 population region (gnomAD-SV_v2.1_DUP_...   \n184  1p36.21 population region (DGV_Gold_Standard_J...   \n418  1p21.1 population region (gnomAD-SV_v2.1_DUP_1...   \n\n                        genes  \n185                            \n186                            \n431  HNRNPCL1,PRAMEF2,PRAMEF4  \n184          HNRNPCL1,PRAMEF2  \n418         AMY1C,AMY1A,AMY1B  ",
      "text/html": "<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>chrom</th>\n      <th>start</th>\n      <th>end</th>\n      <th>#ISCA ID</th>\n      <th>ISCA Region Name</th>\n      <th>genes</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>185</th>\n      <td>chr1</td>\n      <td>713039</td>\n      <td>756009</td>\n      <td>ISCA-46556</td>\n      <td>1p36.33 population region (DGV_Gold_Standard_J...</td>\n      <td></td>\n    </tr>\n    <tr>\n      <th>186</th>\n      <td>chr1</td>\n      <td>713039</td>\n      <td>756009</td>\n      <td>ISCA-46554</td>\n      <td>1p36.33 population region ((DGV_Gold_Standard_...</td>\n      <td></td>\n    </tr>\n    <tr>\n      <th>431</th>\n      <td>chr1</td>\n      <td>12894999</td>\n      <td>12940000</td>\n      <td>ISCA-46308</td>\n      <td>1p36.21 population region (gnomAD-SV_v2.1_DUP_...</td>\n      <td>HNRNPCL1,PRAMEF2,PRAMEF4</td>\n    </tr>\n    <tr>\n      <th>184</th>\n      <td>chr1</td>\n      <td>12897609</td>\n      <td>12922238</td>\n      <td>ISCA-46557</td>\n      <td>1p36.21 population region (DGV_Gold_Standard_J...</td>\n      <td>HNRNPCL1,PRAMEF2</td>\n    </tr>\n    <tr>\n      <th>418</th>\n      <td>chr1</td>\n      <td>104190999</td>\n      <td>104303200</td>\n      <td>ISCA-46321</td>\n      <td>1p21.1 population region (gnomAD-SV_v2.1_DUP_1...</td>\n      <td>AMY1C,AMY1A,AMY1B</td>\n    </tr>\n  </tbody>\n</table>\n</div>"
     },
     "execution_count": 236,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "uts_region['genes'] = uts_region.apply(\n",
    "    lambda row: ','.join(gene.loc[\n",
    "        (gene['chrom'] == row['chrom'])\n",
    "        & (gene['txEnd'] >= row['start'])\n",
    "        & (gene['txStart'] <= row['end']),\n",
    "        'name2'\n",
    "    ]),\n",
    "    axis=1\n",
    ")\n",
    "uts_region.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 237,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-05-11T06:01:38.926049200Z",
     "start_time": "2024-05-11T06:01:38.893166100Z"
    }
   },
   "outputs": [],
   "source": [
    "uts_region.rename(columns={\n",
    "    'chrom': '#chrom',\n",
    "    '#ISCA ID': 'isca_id', 'ISCA Region Name': 'name'\n",
    "}).to_csv(uts_region_file, sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 238,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-05-11T06:01:40.926034100Z",
     "start_time": "2024-05-11T06:01:38.971926800Z"
    }
   },
   "outputs": [],
   "source": [
    "!bgzip -cf {uts_region_file} > {uts_region_file}.gz\n",
    "!tabix -fp bed {uts_region_file}.gz\n",
    "!rm {uts_region_file}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 239,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-05-11T06:01:43.760533300Z",
     "start_time": "2024-05-11T06:01:40.936733600Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": "      chrom                                               info       id  type  \\\n12     chr1  ID=gssvG9;Name=gssvG9;variant_type=CNV;variant...   gssvG9  Gain   \n54     chr1  ID=gssvG28;Name=gssvG28;variant_type=CNV;varia...  gssvG28  Gain   \n18651  chr1  ID=gssvL18;Name=gssvL18;variant_type=CNV;varia...  gssvL18  Loss   \n18627  chr1  ID=gssvL4;Name=gssvL4;variant_type=CNV;variant...   gssvL4  Loss   \n48     chr1  ID=gssvG24;Name=gssvG24;variant_type=CNV;varia...  gssvG24  Gain   \n\n        start     end  freq  sample      af  \n12      49911  222421  1.13    1149  0.0113  \n54      60905   97505  0.61    2609  0.0061  \n18651   60905   97505  0.61    2609  0.0061  \n18627   61724  346583  0.30   12364  0.0030  \n48     125400  176984  0.73    1510  0.0073  ",
      "text/html": "<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>chrom</th>\n      <th>info</th>\n      <th>id</th>\n      <th>type</th>\n      <th>start</th>\n      <th>end</th>\n      <th>freq</th>\n      <th>sample</th>\n      <th>af</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>12</th>\n      <td>chr1</td>\n      <td>ID=gssvG9;Name=gssvG9;variant_type=CNV;variant...</td>\n      <td>gssvG9</td>\n      <td>Gain</td>\n      <td>49911</td>\n      <td>222421</td>\n      <td>1.13</td>\n      <td>1149</td>\n      <td>0.0113</td>\n    </tr>\n    <tr>\n      <th>54</th>\n      <td>chr1</td>\n      <td>ID=gssvG28;Name=gssvG28;variant_type=CNV;varia...</td>\n      <td>gssvG28</td>\n      <td>Gain</td>\n      <td>60905</td>\n      <td>97505</td>\n      <td>0.61</td>\n      <td>2609</td>\n      <td>0.0061</td>\n    </tr>\n    <tr>\n      <th>18651</th>\n      <td>chr1</td>\n      <td>ID=gssvL18;Name=gssvL18;variant_type=CNV;varia...</td>\n      <td>gssvL18</td>\n      <td>Loss</td>\n      <td>60905</td>\n      <td>97505</td>\n      <td>0.61</td>\n      <td>2609</td>\n      <td>0.0061</td>\n    </tr>\n    <tr>\n      <th>18627</th>\n      <td>chr1</td>\n      <td>ID=gssvL4;Name=gssvL4;variant_type=CNV;variant...</td>\n      <td>gssvL4</td>\n      <td>Loss</td>\n      <td>61724</td>\n      <td>346583</td>\n      <td>0.30</td>\n      <td>12364</td>\n      <td>0.0030</td>\n    </tr>\n    <tr>\n      <th>48</th>\n      <td>chr1</td>\n      <td>ID=gssvG24;Name=gssvG24;variant_type=CNV;varia...</td>\n      <td>gssvG24</td>\n      <td>Gain</td>\n      <td>125400</td>\n      <td>176984</td>\n      <td>0.73</td>\n      <td>1510</td>\n      <td>0.0073</td>\n    </tr>\n  </tbody>\n</table>\n</div>"
     },
     "execution_count": 239,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dgv = pd.read_csv(\n",
    "    dgv_ori_file,\n",
    "    sep='\\t', names=['chrom', 'info'], usecols=[0, 8]\n",
    ").drop_duplicates('info')\n",
    "info = dgv['info'].str.extract(\n",
    "    r'ID=(?P<id>[^;]+).*variant_sub_type=(?P<type>[^;]+).*inner_start=(?P<start>[^;]+).*inner_end=(?P<end>[^;]+).*Frequency=(?P<freq>\\S+?)%;.*num_unique_samples_tested=(?P<sample>[^;]+)'\n",
    ").astype({'start': int, 'end': int, 'sample': int, 'freq': float})\n",
    "dgv = dgv.merge(info, left_index=True, right_index=True)\n",
    "dgv['af'] = dgv['freq'] / 100\n",
    "dgv = dgv[dgv['sample'] >= 1000].sort_values(['chrom', 'start', 'end'])\n",
    "dgv.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 240,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-05-11T06:01:43.773498300Z",
     "start_time": "2024-05-11T06:01:43.763526600Z"
    }
   },
   "outputs": [],
   "source": [
    "def fetch_gene(gene, chrom, start, end):\n",
    "    return ','.join(gene.loc[\n",
    "        (gene['chrom'] == chrom) & (gene['txEnd'] >= start) & (gene['txStart'] <= end), 'name2'\n",
    "    ])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 241,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-05-11T06:01:51.886934900Z",
     "start_time": "2024-05-11T06:01:43.768511500Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": "      chrom                                               info       id  type  \\\n12     chr1  ID=gssvG9;Name=gssvG9;variant_type=CNV;variant...   gssvG9  Gain   \n54     chr1  ID=gssvG28;Name=gssvG28;variant_type=CNV;varia...  gssvG28  Gain   \n18651  chr1  ID=gssvL18;Name=gssvL18;variant_type=CNV;varia...  gssvL18  Loss   \n18627  chr1  ID=gssvL4;Name=gssvL4;variant_type=CNV;variant...   gssvL4  Loss   \n48     chr1  ID=gssvG24;Name=gssvG24;variant_type=CNV;varia...  gssvG24  Gain   \n\n        start     end  freq  sample      af  genes  \n12      49911  222421  1.13    1149  0.0113  OR4F5  \n54      60905   97505  0.61    2609  0.0061  OR4F5  \n18651   60905   97505  0.61    2609  0.0061  OR4F5  \n18627   61724  346583  0.30   12364  0.0030  OR4F5  \n48     125400  176984  0.73    1510  0.0073         ",
      "text/html": "<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>chrom</th>\n      <th>info</th>\n      <th>id</th>\n      <th>type</th>\n      <th>start</th>\n      <th>end</th>\n      <th>freq</th>\n      <th>sample</th>\n      <th>af</th>\n      <th>genes</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>12</th>\n      <td>chr1</td>\n      <td>ID=gssvG9;Name=gssvG9;variant_type=CNV;variant...</td>\n      <td>gssvG9</td>\n      <td>Gain</td>\n      <td>49911</td>\n      <td>222421</td>\n      <td>1.13</td>\n      <td>1149</td>\n      <td>0.0113</td>\n      <td>OR4F5</td>\n    </tr>\n    <tr>\n      <th>54</th>\n      <td>chr1</td>\n      <td>ID=gssvG28;Name=gssvG28;variant_type=CNV;varia...</td>\n      <td>gssvG28</td>\n      <td>Gain</td>\n      <td>60905</td>\n      <td>97505</td>\n      <td>0.61</td>\n      <td>2609</td>\n      <td>0.0061</td>\n      <td>OR4F5</td>\n    </tr>\n    <tr>\n      <th>18651</th>\n      <td>chr1</td>\n      <td>ID=gssvL18;Name=gssvL18;variant_type=CNV;varia...</td>\n      <td>gssvL18</td>\n      <td>Loss</td>\n      <td>60905</td>\n      <td>97505</td>\n      <td>0.61</td>\n      <td>2609</td>\n      <td>0.0061</td>\n      <td>OR4F5</td>\n    </tr>\n    <tr>\n      <th>18627</th>\n      <td>chr1</td>\n      <td>ID=gssvL4;Name=gssvL4;variant_type=CNV;variant...</td>\n      <td>gssvL4</td>\n      <td>Loss</td>\n      <td>61724</td>\n      <td>346583</td>\n      <td>0.30</td>\n      <td>12364</td>\n      <td>0.0030</td>\n      <td>OR4F5</td>\n    </tr>\n    <tr>\n      <th>48</th>\n      <td>chr1</td>\n      <td>ID=gssvG24;Name=gssvG24;variant_type=CNV;varia...</td>\n      <td>gssvG24</td>\n      <td>Gain</td>\n      <td>125400</td>\n      <td>176984</td>\n      <td>0.73</td>\n      <td>1510</td>\n      <td>0.0073</td>\n      <td></td>\n    </tr>\n  </tbody>\n</table>\n</div>"
     },
     "execution_count": 241,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "with Pool(processes=7) as pool:\n",
    "    dgv['genes'] = pool.starmap(fetch_gene, (\n",
    "        (gene, row['chrom'], row['start'], row['end']) for _, row in dgv.iterrows()\n",
    "    ), chunksize=70)\n",
    "dgv.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 242,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-05-11T06:01:52.013590700Z",
     "start_time": "2024-05-11T06:01:51.887733Z"
    }
   },
   "outputs": [],
   "source": [
    "dgv.loc[\n",
    "    dgv['type'] == 'Gain', ['chrom', 'start', 'end', 'id', 'genes', 'af']\n",
    "].rename(columns={'chrom': '#chrom'}).to_csv(dgv_gain_file, sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 243,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-05-11T06:01:53.957482700Z",
     "start_time": "2024-05-11T06:01:51.947747800Z"
    }
   },
   "outputs": [],
   "source": [
    "!bgzip -cf {dgv_gain_file} > {dgv_gain_file}.gz\n",
    "!tabix -fp bed {dgv_gain_file}.gz\n",
    "!rm {dgv_gain_file}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 244,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-05-11T06:01:54.121635400Z",
     "start_time": "2024-05-11T06:01:53.963466500Z"
    }
   },
   "outputs": [],
   "source": [
    "dgv.loc[\n",
    "    dgv['type'] == 'Loss', ['chrom', 'start', 'end', 'id', 'genes', 'af']\n",
    "].rename(columns={'chrom': '#chrom'}).to_csv(dgv_loss_file, sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 245,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-05-11T06:01:56.300644500Z",
     "start_time": "2024-05-11T06:01:54.200313200Z"
    }
   },
   "outputs": [],
   "source": [
    "!bgzip -cf {dgv_loss_file} > {dgv_loss_file}.gz\n",
    "!tabix -fp bed {dgv_loss_file}.gz\n",
    "!rm {dgv_loss_file}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 246,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-05-11T06:02:00.304051100Z",
     "start_time": "2024-05-11T06:01:56.313618Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": "   #chrom   start     end                    name svtype  \\\n6    chr1   62398   62489  gnomAD-SV_v2.1_DEL_1_3    DEL   \n7    chr1   66349   66427  gnomAD-SV_v2.1_DEL_1_4    DEL   \n15   chr1  136999  155500  gnomAD-SV_v2.1_DUP_1_3    DUP   \n16   chr1  141474  155000  gnomAD-SV_v2.1_DUP_1_4    DUP   \n24   chr1  319999  327650  gnomAD-SV_v2.1_DEL_1_9    DEL   \n\n                       AF N_BI_GENOS               AFR_AF AFR_N_BI_GENOS  \\\n6   9.600000339560211e-05       5192                  0.0           2014   \n7   0.0004830000107176602       5177                  0.0           2009   \n15      0.138032004237175       5122  0.10211800038814545           1983   \n16   0.006382000166922808       4544  0.01448499970138073           1864   \n24     0.1377120018005371       5192  0.29294899106025696           2014   \n\n                   AMR_AF AMR_N_BI_GENOS                 EAS_AF  \\\n6                     0.0            609  0.0005840000230818987   \n7   0.0024709999561309814            607                    0.0   \n15     0.1157890036702156            570     0.3907710015773773   \n16                    0.0            503   0.001905999961309135   \n24   0.020524999126791954            609    0.08235999941825867   \n\n   EAS_N_BI_GENOS                  EUR_AF EUR_N_BI_GENOS FILTER  \n6             856                     0.0           1660   PASS  \n7             856   0.0006050000083632767           1652   PASS  \n15            856     0.05963899940252304           1660   PASS  \n16            787  0.00037200000951997936           1343   PASS  \n24            856    0.022590000182390213           1660   PASS  ",
      "text/html": "<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>#chrom</th>\n      <th>start</th>\n      <th>end</th>\n      <th>name</th>\n      <th>svtype</th>\n      <th>AF</th>\n      <th>N_BI_GENOS</th>\n      <th>AFR_AF</th>\n      <th>AFR_N_BI_GENOS</th>\n      <th>AMR_AF</th>\n      <th>AMR_N_BI_GENOS</th>\n      <th>EAS_AF</th>\n      <th>EAS_N_BI_GENOS</th>\n      <th>EUR_AF</th>\n      <th>EUR_N_BI_GENOS</th>\n      <th>FILTER</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>6</th>\n      <td>chr1</td>\n      <td>62398</td>\n      <td>62489</td>\n      <td>gnomAD-SV_v2.1_DEL_1_3</td>\n      <td>DEL</td>\n      <td>9.600000339560211e-05</td>\n      <td>5192</td>\n      <td>0.0</td>\n      <td>2014</td>\n      <td>0.0</td>\n      <td>609</td>\n      <td>0.0005840000230818987</td>\n      <td>856</td>\n      <td>0.0</td>\n      <td>1660</td>\n      <td>PASS</td>\n    </tr>\n    <tr>\n      <th>7</th>\n      <td>chr1</td>\n      <td>66349</td>\n      <td>66427</td>\n      <td>gnomAD-SV_v2.1_DEL_1_4</td>\n      <td>DEL</td>\n      <td>0.0004830000107176602</td>\n      <td>5177</td>\n      <td>0.0</td>\n      <td>2009</td>\n      <td>0.0024709999561309814</td>\n      <td>607</td>\n      <td>0.0</td>\n      <td>856</td>\n      <td>0.0006050000083632767</td>\n      <td>1652</td>\n      <td>PASS</td>\n    </tr>\n    <tr>\n      <th>15</th>\n      <td>chr1</td>\n      <td>136999</td>\n      <td>155500</td>\n      <td>gnomAD-SV_v2.1_DUP_1_3</td>\n      <td>DUP</td>\n      <td>0.138032004237175</td>\n      <td>5122</td>\n      <td>0.10211800038814545</td>\n      <td>1983</td>\n      <td>0.1157890036702156</td>\n      <td>570</td>\n      <td>0.3907710015773773</td>\n      <td>856</td>\n      <td>0.05963899940252304</td>\n      <td>1660</td>\n      <td>PASS</td>\n    </tr>\n    <tr>\n      <th>16</th>\n      <td>chr1</td>\n      <td>141474</td>\n      <td>155000</td>\n      <td>gnomAD-SV_v2.1_DUP_1_4</td>\n      <td>DUP</td>\n      <td>0.006382000166922808</td>\n      <td>4544</td>\n      <td>0.01448499970138073</td>\n      <td>1864</td>\n      <td>0.0</td>\n      <td>503</td>\n      <td>0.001905999961309135</td>\n      <td>787</td>\n      <td>0.00037200000951997936</td>\n      <td>1343</td>\n      <td>PASS</td>\n    </tr>\n    <tr>\n      <th>24</th>\n      <td>chr1</td>\n      <td>319999</td>\n      <td>327650</td>\n      <td>gnomAD-SV_v2.1_DEL_1_9</td>\n      <td>DEL</td>\n      <td>0.1377120018005371</td>\n      <td>5192</td>\n      <td>0.29294899106025696</td>\n      <td>2014</td>\n      <td>0.020524999126791954</td>\n      <td>609</td>\n      <td>0.08235999941825867</td>\n      <td>856</td>\n      <td>0.022590000182390213</td>\n      <td>1660</td>\n      <td>PASS</td>\n    </tr>\n  </tbody>\n</table>\n</div>"
     },
     "execution_count": 246,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gnomad = pd.read_csv(\n",
    "    gnomad_control_ori_file, sep='\\t',  dtype=str,\n",
    "    usecols=[0, 1, 2, 3, 4, 37, 38, 73, 74, 107, 108, 141, 142, 175, 176, 241]\n",
    ")\n",
    "gnomad = gnomad[\n",
    "    (gnomad['FILTER'] == 'PASS') & gnomad['svtype'].isin(['DEL', 'DUP'])\n",
    "]\n",
    "gnomad = gnomad[\n",
    "    (gnomad['N_BI_GENOS'].astype(int) >= 1000) |\n",
    "    (gnomad['AFR_N_BI_GENOS'].astype(int) >= 1000) |\n",
    "    (gnomad['AMR_N_BI_GENOS'].astype(int) >= 1000) |\n",
    "    (gnomad['EAS_N_BI_GENOS'].astype(int) >= 1000) |\n",
    "    (gnomad['EUR_N_BI_GENOS'].astype(int) >= 1000)\n",
    "]\n",
    "gnomad['#chrom'] = 'chr' + gnomad['#chrom']\n",
    "gnomad['start'] = gnomad['start'].astype(int)\n",
    "gnomad['end'] = gnomad['end'].astype(int)\n",
    "gnomad.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 247,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-05-11T06:02:24.426845Z",
     "start_time": "2024-05-11T06:02:00.308039800Z"
    }
   },
   "outputs": [],
   "source": [
    "with Pool(processes=30) as pool:\n",
    "    gnomad['genes'] = pool.starmap(fetch_gene, (\n",
    "        (gene, row['#chrom'], row['start'], row['end']) for _, row in gnomad.iterrows()\n",
    "    ), chunksize=70)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 248,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-05-11T06:05:01.630858900Z",
     "start_time": "2024-05-11T06:02:24.454588100Z"
    }
   },
   "outputs": [
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001B[0;31m---------------------------------------------------------------------------\u001B[0m",
      "\u001B[0;31mKeyboardInterrupt\u001B[0m                         Traceback (most recent call last)",
      "Cell \u001B[0;32mIn[248], line 4\u001B[0m\n\u001B[1;32m      1\u001B[0m \u001B[38;5;66;03m# gnomad['genes'] = pool.starmap(fetch_gene, (\u001B[39;00m\n\u001B[1;32m      2\u001B[0m \u001B[38;5;66;03m#         (gene, row['#chrom'], row['start'], row['end']) for _, row in gnomad.iterrows()\u001B[39;00m\n\u001B[1;32m      3\u001B[0m \u001B[38;5;66;03m#     ), chunksize=70)\u001B[39;00m\n\u001B[0;32m----> 4\u001B[0m gnomad[\u001B[38;5;124m'\u001B[39m\u001B[38;5;124mgene\u001B[39m\u001B[38;5;124m'\u001B[39m] \u001B[38;5;241m=\u001B[39m \u001B[43mgnomad\u001B[49m\u001B[38;5;241;43m.\u001B[39;49m\u001B[43mapply\u001B[49m\u001B[43m(\u001B[49m\u001B[38;5;28;43;01mlambda\u001B[39;49;00m\u001B[43m \u001B[49m\u001B[43mrow\u001B[49m\u001B[43m:\u001B[49m\u001B[43m \u001B[49m\u001B[43mfetch_gene\u001B[49m\u001B[43m(\u001B[49m\u001B[43mgene\u001B[49m\u001B[43m,\u001B[49m\u001B[43m \u001B[49m\u001B[43mrow\u001B[49m\u001B[43m[\u001B[49m\u001B[38;5;124;43m'\u001B[39;49m\u001B[38;5;124;43m#chrom\u001B[39;49m\u001B[38;5;124;43m'\u001B[39;49m\u001B[43m]\u001B[49m\u001B[43m,\u001B[49m\u001B[43m \u001B[49m\u001B[43mrow\u001B[49m\u001B[43m[\u001B[49m\u001B[38;5;124;43m'\u001B[39;49m\u001B[38;5;124;43mstart\u001B[39;49m\u001B[38;5;124;43m'\u001B[39;49m\u001B[43m]\u001B[49m\u001B[43m,\u001B[49m\u001B[43m \u001B[49m\u001B[43mrow\u001B[49m\u001B[43m[\u001B[49m\u001B[38;5;124;43m'\u001B[39;49m\u001B[38;5;124;43mend\u001B[39;49m\u001B[38;5;124;43m'\u001B[39;49m\u001B[43m]\u001B[49m\u001B[43m)\u001B[49m\u001B[43m,\u001B[49m\u001B[43m \u001B[49m\u001B[43maxis\u001B[49m\u001B[38;5;241;43m=\u001B[39;49m\u001B[38;5;241;43m1\u001B[39;49m\u001B[43m)\u001B[49m\n\u001B[1;32m      5\u001B[0m gnomad\u001B[38;5;241m.\u001B[39mhead()\n",
      "File \u001B[0;32m/staging/software/lib/python3.9/site-packages/pandas/core/frame.py:9568\u001B[0m, in \u001B[0;36mDataFrame.apply\u001B[0;34m(self, func, axis, raw, result_type, args, **kwargs)\u001B[0m\n\u001B[1;32m   9557\u001B[0m \u001B[38;5;28;01mfrom\u001B[39;00m \u001B[38;5;21;01mpandas\u001B[39;00m\u001B[38;5;21;01m.\u001B[39;00m\u001B[38;5;21;01mcore\u001B[39;00m\u001B[38;5;21;01m.\u001B[39;00m\u001B[38;5;21;01mapply\u001B[39;00m \u001B[38;5;28;01mimport\u001B[39;00m frame_apply\n\u001B[1;32m   9559\u001B[0m op \u001B[38;5;241m=\u001B[39m frame_apply(\n\u001B[1;32m   9560\u001B[0m     \u001B[38;5;28mself\u001B[39m,\n\u001B[1;32m   9561\u001B[0m     func\u001B[38;5;241m=\u001B[39mfunc,\n\u001B[0;32m   (...)\u001B[0m\n\u001B[1;32m   9566\u001B[0m     kwargs\u001B[38;5;241m=\u001B[39mkwargs,\n\u001B[1;32m   9567\u001B[0m )\n\u001B[0;32m-> 9568\u001B[0m \u001B[38;5;28;01mreturn\u001B[39;00m \u001B[43mop\u001B[49m\u001B[38;5;241;43m.\u001B[39;49m\u001B[43mapply\u001B[49m\u001B[43m(\u001B[49m\u001B[43m)\u001B[49m\u001B[38;5;241m.\u001B[39m__finalize__(\u001B[38;5;28mself\u001B[39m, method\u001B[38;5;241m=\u001B[39m\u001B[38;5;124m\"\u001B[39m\u001B[38;5;124mapply\u001B[39m\u001B[38;5;124m\"\u001B[39m)\n",
      "File \u001B[0;32m/staging/software/lib/python3.9/site-packages/pandas/core/apply.py:764\u001B[0m, in \u001B[0;36mFrameApply.apply\u001B[0;34m(self)\u001B[0m\n\u001B[1;32m    761\u001B[0m \u001B[38;5;28;01melif\u001B[39;00m \u001B[38;5;28mself\u001B[39m\u001B[38;5;241m.\u001B[39mraw:\n\u001B[1;32m    762\u001B[0m     \u001B[38;5;28;01mreturn\u001B[39;00m \u001B[38;5;28mself\u001B[39m\u001B[38;5;241m.\u001B[39mapply_raw()\n\u001B[0;32m--> 764\u001B[0m \u001B[38;5;28;01mreturn\u001B[39;00m \u001B[38;5;28;43mself\u001B[39;49m\u001B[38;5;241;43m.\u001B[39;49m\u001B[43mapply_standard\u001B[49m\u001B[43m(\u001B[49m\u001B[43m)\u001B[49m\n",
      "File \u001B[0;32m/staging/software/lib/python3.9/site-packages/pandas/core/apply.py:891\u001B[0m, in \u001B[0;36mFrameApply.apply_standard\u001B[0;34m(self)\u001B[0m\n\u001B[1;32m    890\u001B[0m \u001B[38;5;28;01mdef\u001B[39;00m \u001B[38;5;21mapply_standard\u001B[39m(\u001B[38;5;28mself\u001B[39m):\n\u001B[0;32m--> 891\u001B[0m     results, res_index \u001B[38;5;241m=\u001B[39m \u001B[38;5;28;43mself\u001B[39;49m\u001B[38;5;241;43m.\u001B[39;49m\u001B[43mapply_series_generator\u001B[49m\u001B[43m(\u001B[49m\u001B[43m)\u001B[49m\n\u001B[1;32m    893\u001B[0m     \u001B[38;5;66;03m# wrap results\u001B[39;00m\n\u001B[1;32m    894\u001B[0m     \u001B[38;5;28;01mreturn\u001B[39;00m \u001B[38;5;28mself\u001B[39m\u001B[38;5;241m.\u001B[39mwrap_results(results, res_index)\n",
      "File \u001B[0;32m/staging/software/lib/python3.9/site-packages/pandas/core/apply.py:907\u001B[0m, in \u001B[0;36mFrameApply.apply_series_generator\u001B[0;34m(self)\u001B[0m\n\u001B[1;32m    904\u001B[0m \u001B[38;5;28;01mwith\u001B[39;00m option_context(\u001B[38;5;124m\"\u001B[39m\u001B[38;5;124mmode.chained_assignment\u001B[39m\u001B[38;5;124m\"\u001B[39m, \u001B[38;5;28;01mNone\u001B[39;00m):\n\u001B[1;32m    905\u001B[0m     \u001B[38;5;28;01mfor\u001B[39;00m i, v \u001B[38;5;129;01min\u001B[39;00m \u001B[38;5;28menumerate\u001B[39m(series_gen):\n\u001B[1;32m    906\u001B[0m         \u001B[38;5;66;03m# ignore SettingWithCopy here in case the user mutates\u001B[39;00m\n\u001B[0;32m--> 907\u001B[0m         results[i] \u001B[38;5;241m=\u001B[39m \u001B[38;5;28;43mself\u001B[39;49m\u001B[38;5;241;43m.\u001B[39;49m\u001B[43mf\u001B[49m\u001B[43m(\u001B[49m\u001B[43mv\u001B[49m\u001B[43m)\u001B[49m\n\u001B[1;32m    908\u001B[0m         \u001B[38;5;28;01mif\u001B[39;00m \u001B[38;5;28misinstance\u001B[39m(results[i], ABCSeries):\n\u001B[1;32m    909\u001B[0m             \u001B[38;5;66;03m# If we have a view on v, we need to make a copy because\u001B[39;00m\n\u001B[1;32m    910\u001B[0m             \u001B[38;5;66;03m#  series_generator will swap out the underlying data\u001B[39;00m\n\u001B[1;32m    911\u001B[0m             results[i] \u001B[38;5;241m=\u001B[39m results[i]\u001B[38;5;241m.\u001B[39mcopy(deep\u001B[38;5;241m=\u001B[39m\u001B[38;5;28;01mFalse\u001B[39;00m)\n",
      "Cell \u001B[0;32mIn[248], line 4\u001B[0m, in \u001B[0;36m<lambda>\u001B[0;34m(row)\u001B[0m\n\u001B[1;32m      1\u001B[0m \u001B[38;5;66;03m# gnomad['genes'] = pool.starmap(fetch_gene, (\u001B[39;00m\n\u001B[1;32m      2\u001B[0m \u001B[38;5;66;03m#         (gene, row['#chrom'], row['start'], row['end']) for _, row in gnomad.iterrows()\u001B[39;00m\n\u001B[1;32m      3\u001B[0m \u001B[38;5;66;03m#     ), chunksize=70)\u001B[39;00m\n\u001B[0;32m----> 4\u001B[0m gnomad[\u001B[38;5;124m'\u001B[39m\u001B[38;5;124mgene\u001B[39m\u001B[38;5;124m'\u001B[39m] \u001B[38;5;241m=\u001B[39m gnomad\u001B[38;5;241m.\u001B[39mapply(\u001B[38;5;28;01mlambda\u001B[39;00m row: \u001B[43mfetch_gene\u001B[49m\u001B[43m(\u001B[49m\u001B[43mgene\u001B[49m\u001B[43m,\u001B[49m\u001B[43m \u001B[49m\u001B[43mrow\u001B[49m\u001B[43m[\u001B[49m\u001B[38;5;124;43m'\u001B[39;49m\u001B[38;5;124;43m#chrom\u001B[39;49m\u001B[38;5;124;43m'\u001B[39;49m\u001B[43m]\u001B[49m\u001B[43m,\u001B[49m\u001B[43m \u001B[49m\u001B[43mrow\u001B[49m\u001B[43m[\u001B[49m\u001B[38;5;124;43m'\u001B[39;49m\u001B[38;5;124;43mstart\u001B[39;49m\u001B[38;5;124;43m'\u001B[39;49m\u001B[43m]\u001B[49m\u001B[43m,\u001B[49m\u001B[43m \u001B[49m\u001B[43mrow\u001B[49m\u001B[43m[\u001B[49m\u001B[38;5;124;43m'\u001B[39;49m\u001B[38;5;124;43mend\u001B[39;49m\u001B[38;5;124;43m'\u001B[39;49m\u001B[43m]\u001B[49m\u001B[43m)\u001B[49m, axis\u001B[38;5;241m=\u001B[39m\u001B[38;5;241m1\u001B[39m)\n\u001B[1;32m      5\u001B[0m gnomad\u001B[38;5;241m.\u001B[39mhead()\n",
      "Cell \u001B[0;32mIn[240], line 3\u001B[0m, in \u001B[0;36mfetch_gene\u001B[0;34m(gene, chrom, start, end)\u001B[0m\n\u001B[1;32m      1\u001B[0m \u001B[38;5;28;01mdef\u001B[39;00m \u001B[38;5;21mfetch_gene\u001B[39m(gene, chrom, start, end):\n\u001B[1;32m      2\u001B[0m     \u001B[38;5;28;01mreturn\u001B[39;00m \u001B[38;5;124m'\u001B[39m\u001B[38;5;124m,\u001B[39m\u001B[38;5;124m'\u001B[39m\u001B[38;5;241m.\u001B[39mjoin(gene\u001B[38;5;241m.\u001B[39mloc[\n\u001B[0;32m----> 3\u001B[0m         (\u001B[43mgene\u001B[49m\u001B[43m[\u001B[49m\u001B[38;5;124;43m'\u001B[39;49m\u001B[38;5;124;43mchrom\u001B[39;49m\u001B[38;5;124;43m'\u001B[39;49m\u001B[43m]\u001B[49m\u001B[43m \u001B[49m\u001B[38;5;241;43m==\u001B[39;49m\u001B[43m \u001B[49m\u001B[43mchrom\u001B[49m) \u001B[38;5;241m&\u001B[39m (gene[\u001B[38;5;124m'\u001B[39m\u001B[38;5;124mtxEnd\u001B[39m\u001B[38;5;124m'\u001B[39m] \u001B[38;5;241m>\u001B[39m\u001B[38;5;241m=\u001B[39m start) \u001B[38;5;241m&\u001B[39m (gene[\u001B[38;5;124m'\u001B[39m\u001B[38;5;124mtxStart\u001B[39m\u001B[38;5;124m'\u001B[39m] \u001B[38;5;241m<\u001B[39m\u001B[38;5;241m=\u001B[39m end), \u001B[38;5;124m'\u001B[39m\u001B[38;5;124mname2\u001B[39m\u001B[38;5;124m'\u001B[39m\n\u001B[1;32m      4\u001B[0m     ])\n",
      "File \u001B[0;32m/staging/software/lib/python3.9/site-packages/pandas/core/ops/common.py:72\u001B[0m, in \u001B[0;36m_unpack_zerodim_and_defer.<locals>.new_method\u001B[0;34m(self, other)\u001B[0m\n\u001B[1;32m     68\u001B[0m             \u001B[38;5;28;01mreturn\u001B[39;00m \u001B[38;5;28mNotImplemented\u001B[39m\n\u001B[1;32m     70\u001B[0m other \u001B[38;5;241m=\u001B[39m item_from_zerodim(other)\n\u001B[0;32m---> 72\u001B[0m \u001B[38;5;28;01mreturn\u001B[39;00m \u001B[43mmethod\u001B[49m\u001B[43m(\u001B[49m\u001B[38;5;28;43mself\u001B[39;49m\u001B[43m,\u001B[49m\u001B[43m \u001B[49m\u001B[43mother\u001B[49m\u001B[43m)\u001B[49m\n",
      "File \u001B[0;32m/staging/software/lib/python3.9/site-packages/pandas/core/arraylike.py:42\u001B[0m, in \u001B[0;36mOpsMixin.__eq__\u001B[0;34m(self, other)\u001B[0m\n\u001B[1;32m     40\u001B[0m \u001B[38;5;129m@unpack_zerodim_and_defer\u001B[39m(\u001B[38;5;124m\"\u001B[39m\u001B[38;5;124m__eq__\u001B[39m\u001B[38;5;124m\"\u001B[39m)\n\u001B[1;32m     41\u001B[0m \u001B[38;5;28;01mdef\u001B[39;00m \u001B[38;5;21m__eq__\u001B[39m(\u001B[38;5;28mself\u001B[39m, other):\n\u001B[0;32m---> 42\u001B[0m     \u001B[38;5;28;01mreturn\u001B[39;00m \u001B[38;5;28;43mself\u001B[39;49m\u001B[38;5;241;43m.\u001B[39;49m\u001B[43m_cmp_method\u001B[49m\u001B[43m(\u001B[49m\u001B[43mother\u001B[49m\u001B[43m,\u001B[49m\u001B[43m \u001B[49m\u001B[43moperator\u001B[49m\u001B[38;5;241;43m.\u001B[39;49m\u001B[43meq\u001B[49m\u001B[43m)\u001B[49m\n",
      "File \u001B[0;32m/staging/software/lib/python3.9/site-packages/pandas/core/series.py:6243\u001B[0m, in \u001B[0;36mSeries._cmp_method\u001B[0;34m(self, other, op)\u001B[0m\n\u001B[1;32m   6240\u001B[0m rvalues \u001B[38;5;241m=\u001B[39m extract_array(other, extract_numpy\u001B[38;5;241m=\u001B[39m\u001B[38;5;28;01mTrue\u001B[39;00m, extract_range\u001B[38;5;241m=\u001B[39m\u001B[38;5;28;01mTrue\u001B[39;00m)\n\u001B[1;32m   6242\u001B[0m \u001B[38;5;28;01mwith\u001B[39;00m np\u001B[38;5;241m.\u001B[39merrstate(\u001B[38;5;28mall\u001B[39m\u001B[38;5;241m=\u001B[39m\u001B[38;5;124m\"\u001B[39m\u001B[38;5;124mignore\u001B[39m\u001B[38;5;124m\"\u001B[39m):\n\u001B[0;32m-> 6243\u001B[0m     res_values \u001B[38;5;241m=\u001B[39m \u001B[43mops\u001B[49m\u001B[38;5;241;43m.\u001B[39;49m\u001B[43mcomparison_op\u001B[49m\u001B[43m(\u001B[49m\u001B[43mlvalues\u001B[49m\u001B[43m,\u001B[49m\u001B[43m \u001B[49m\u001B[43mrvalues\u001B[49m\u001B[43m,\u001B[49m\u001B[43m \u001B[49m\u001B[43mop\u001B[49m\u001B[43m)\u001B[49m\n\u001B[1;32m   6245\u001B[0m \u001B[38;5;28;01mreturn\u001B[39;00m \u001B[38;5;28mself\u001B[39m\u001B[38;5;241m.\u001B[39m_construct_result(res_values, name\u001B[38;5;241m=\u001B[39mres_name)\n",
      "File \u001B[0;32m/staging/software/lib/python3.9/site-packages/pandas/core/ops/array_ops.py:287\u001B[0m, in \u001B[0;36mcomparison_op\u001B[0;34m(left, right, op)\u001B[0m\n\u001B[1;32m    284\u001B[0m     \u001B[38;5;28;01mreturn\u001B[39;00m invalid_comparison(lvalues, rvalues, op)\n\u001B[1;32m    286\u001B[0m \u001B[38;5;28;01melif\u001B[39;00m is_object_dtype(lvalues\u001B[38;5;241m.\u001B[39mdtype) \u001B[38;5;129;01mor\u001B[39;00m \u001B[38;5;28misinstance\u001B[39m(rvalues, \u001B[38;5;28mstr\u001B[39m):\n\u001B[0;32m--> 287\u001B[0m     res_values \u001B[38;5;241m=\u001B[39m \u001B[43mcomp_method_OBJECT_ARRAY\u001B[49m\u001B[43m(\u001B[49m\u001B[43mop\u001B[49m\u001B[43m,\u001B[49m\u001B[43m \u001B[49m\u001B[43mlvalues\u001B[49m\u001B[43m,\u001B[49m\u001B[43m \u001B[49m\u001B[43mrvalues\u001B[49m\u001B[43m)\u001B[49m\n\u001B[1;32m    289\u001B[0m \u001B[38;5;28;01melse\u001B[39;00m:\n\u001B[1;32m    290\u001B[0m     res_values \u001B[38;5;241m=\u001B[39m _na_arithmetic_op(lvalues, rvalues, op, is_cmp\u001B[38;5;241m=\u001B[39m\u001B[38;5;28;01mTrue\u001B[39;00m)\n",
      "File \u001B[0;32m/staging/software/lib/python3.9/site-packages/pandas/core/ops/array_ops.py:75\u001B[0m, in \u001B[0;36mcomp_method_OBJECT_ARRAY\u001B[0;34m(op, x, y)\u001B[0m\n\u001B[1;32m     73\u001B[0m     result \u001B[38;5;241m=\u001B[39m libops\u001B[38;5;241m.\u001B[39mvec_compare(x\u001B[38;5;241m.\u001B[39mravel(), y\u001B[38;5;241m.\u001B[39mravel(), op)\n\u001B[1;32m     74\u001B[0m \u001B[38;5;28;01melse\u001B[39;00m:\n\u001B[0;32m---> 75\u001B[0m     result \u001B[38;5;241m=\u001B[39m \u001B[43mlibops\u001B[49m\u001B[38;5;241;43m.\u001B[39;49m\u001B[43mscalar_compare\u001B[49m\u001B[43m(\u001B[49m\u001B[43mx\u001B[49m\u001B[38;5;241;43m.\u001B[39;49m\u001B[43mravel\u001B[49m\u001B[43m(\u001B[49m\u001B[43m)\u001B[49m\u001B[43m,\u001B[49m\u001B[43m \u001B[49m\u001B[43my\u001B[49m\u001B[43m,\u001B[49m\u001B[43m \u001B[49m\u001B[43mop\u001B[49m\u001B[43m)\u001B[49m\n\u001B[1;32m     76\u001B[0m \u001B[38;5;28;01mreturn\u001B[39;00m result\u001B[38;5;241m.\u001B[39mreshape(x\u001B[38;5;241m.\u001B[39mshape)\n",
      "\u001B[0;31mKeyboardInterrupt\u001B[0m: "
     ]
    }
   ],
   "source": [
    "# gnomad['genes'] = pool.starmap(fetch_gene, (\n",
    "#         (gene, row['#chrom'], row['start'], row['end']) for _, row in gnomad.iterrows()\n",
    "#     ), chunksize=70)\n",
    "gnomad['gene'] = gnomad.apply(lambda row: fetch_gene(gene, row['#chrom'], row['start'], row['end']), axis=1)\n",
    "gnomad.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gnomad[\n",
    "    gnomad['svtype'] == 'DEL'\n",
    "][['#chrom', 'start', 'end', 'name', 'genes', 'AF', 'AFR_AF', 'AMR_AF', 'EAS_AF', 'EUR_AF']].rename(columns={\n",
    "    'AF': 'af', 'AFR_AF': 'af_afr', 'AMR_AF': 'af_amr', 'EAS_AF': 'af_eas', 'EUR_AF': 'af_eur'\n",
    "}).to_csv(gnomad_del_file, sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "!bgzip -cf {gnomad_del_file} > {gnomad_del_file}.gz\n",
    "!tabix -fp bed {gnomad_del_file}.gz\n",
    "!rm {gnomad_del_file}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gnomad.loc[\n",
    "    gnomad['svtype'] == 'DUP',\n",
    "    ['#chrom', 'start', 'end', 'name', 'genes', 'AF', 'AFR_AF', 'AMR_AF', 'EAS_AF', 'EUR_AF']\n",
    "].rename(columns={\n",
    "    'AF': 'af', 'AFR_AF': 'af_afr', 'AMR_AF': 'af_amr', 'EAS_AF': 'af_eas', 'EUR_AF': 'af_eur'\n",
    "}).to_csv(gnomad_dup_file, sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [],
   "source": [
    "!bgzip -cf {gnomad_dup_file} > {gnomad_dup_file}.gz\n",
    "!tabix -fp bed {gnomad_dup_file}.gz\n",
    "!rm {gnomad_dup_file}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 256,
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    },
    "ExecuteTime": {
     "end_time": "2024-05-11T06:19:53.468253400Z",
     "start_time": "2024-05-11T06:19:37.092669200Z"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:root:Extracted GTF attributes: ['gene_id', 'transcript_id', 'gene_name', 'exon_number', 'exon_id']\n"
     ]
    },
    {
     "data": {
      "text/plain": "  seqname   source     feature  start    end  score strand  frame  \\\n0    chr1  refGene  transcript  11869  14362    NaN      +      0   \n1    chr1  refGene        exon  11869  12227    NaN      +      0   \n2    chr1  refGene        exon  12613  12721    NaN      +      0   \n3    chr1  refGene        exon  13221  14362    NaN      +      0   \n4    chr1  refGene  transcript  11874  14409    NaN      +      0   \n\n        gene_id transcript_id     gene_name exon_number      exon_id  \n0  LOC102725121     NR_148357  LOC102725121                           \n1  LOC102725121     NR_148357  LOC102725121           1  NR_148357.1  \n2  LOC102725121     NR_148357  LOC102725121           2  NR_148357.2  \n3  LOC102725121     NR_148357  LOC102725121           3  NR_148357.3  \n4       DDX11L1     NR_046018       DDX11L1                           ",
      "text/html": "<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>seqname</th>\n      <th>source</th>\n      <th>feature</th>\n      <th>start</th>\n      <th>end</th>\n      <th>score</th>\n      <th>strand</th>\n      <th>frame</th>\n      <th>gene_id</th>\n      <th>transcript_id</th>\n      <th>gene_name</th>\n      <th>exon_number</th>\n      <th>exon_id</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>0</th>\n      <td>chr1</td>\n      <td>refGene</td>\n      <td>transcript</td>\n      <td>11869</td>\n      <td>14362</td>\n      <td>NaN</td>\n      <td>+</td>\n      <td>0</td>\n      <td>LOC102725121</td>\n      <td>NR_148357</td>\n      <td>LOC102725121</td>\n      <td></td>\n      <td></td>\n    </tr>\n    <tr>\n      <th>1</th>\n      <td>chr1</td>\n      <td>refGene</td>\n      <td>exon</td>\n      <td>11869</td>\n      <td>12227</td>\n      <td>NaN</td>\n      <td>+</td>\n      <td>0</td>\n      <td>LOC102725121</td>\n      <td>NR_148357</td>\n      <td>LOC102725121</td>\n      <td>1</td>\n      <td>NR_148357.1</td>\n    </tr>\n    <tr>\n      <th>2</th>\n      <td>chr1</td>\n      <td>refGene</td>\n      <td>exon</td>\n      <td>12613</td>\n      <td>12721</td>\n      <td>NaN</td>\n      <td>+</td>\n      <td>0</td>\n      <td>LOC102725121</td>\n      <td>NR_148357</td>\n      <td>LOC102725121</td>\n      <td>2</td>\n      <td>NR_148357.2</td>\n    </tr>\n    <tr>\n      <th>3</th>\n      <td>chr1</td>\n      <td>refGene</td>\n      <td>exon</td>\n      <td>13221</td>\n      <td>14362</td>\n      <td>NaN</td>\n      <td>+</td>\n      <td>0</td>\n      <td>LOC102725121</td>\n      <td>NR_148357</td>\n      <td>LOC102725121</td>\n      <td>3</td>\n      <td>NR_148357.3</td>\n    </tr>\n    <tr>\n      <th>4</th>\n      <td>chr1</td>\n      <td>refGene</td>\n      <td>transcript</td>\n      <td>11874</td>\n      <td>14409</td>\n      <td>NaN</td>\n      <td>+</td>\n      <td>0</td>\n      <td>DDX11L1</td>\n      <td>NR_046018</td>\n      <td>DDX11L1</td>\n      <td></td>\n      <td></td>\n    </tr>\n  </tbody>\n</table>\n</div>"
     },
     "execution_count": 256,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "refGene_df = read_gtf(exon_ori_file, result_type='pandas')\n",
    "# refGene_df = refGene_df.applymap(lambda x: x.strip('\"') if '\"' in str(x) else x)\n",
    "refGene_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 257,
   "outputs": [
    {
     "data": {
      "text/plain": "  seqname   source feature  start    end  score strand  frame       gene_id  \\\n1    chr1  refGene    exon  11869  12227    NaN      +      0  LOC102725121   \n2    chr1  refGene    exon  12613  12721    NaN      +      0  LOC102725121   \n3    chr1  refGene    exon  13221  14362    NaN      +      0  LOC102725121   \n5    chr1  refGene    exon  11874  12227    NaN      +      0       DDX11L1   \n6    chr1  refGene    exon  12613  12721    NaN      +      0       DDX11L1   \n\n  transcript_id     gene_name exon_number      exon_id  \n1     NR_148357  LOC102725121           1  NR_148357.1  \n2     NR_148357  LOC102725121           2  NR_148357.2  \n3     NR_148357  LOC102725121           3  NR_148357.3  \n5     NR_046018       DDX11L1           1  NR_046018.1  \n6     NR_046018       DDX11L1           2  NR_046018.2  ",
      "text/html": "<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>seqname</th>\n      <th>source</th>\n      <th>feature</th>\n      <th>start</th>\n      <th>end</th>\n      <th>score</th>\n      <th>strand</th>\n      <th>frame</th>\n      <th>gene_id</th>\n      <th>transcript_id</th>\n      <th>gene_name</th>\n      <th>exon_number</th>\n      <th>exon_id</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>1</th>\n      <td>chr1</td>\n      <td>refGene</td>\n      <td>exon</td>\n      <td>11869</td>\n      <td>12227</td>\n      <td>NaN</td>\n      <td>+</td>\n      <td>0</td>\n      <td>LOC102725121</td>\n      <td>NR_148357</td>\n      <td>LOC102725121</td>\n      <td>1</td>\n      <td>NR_148357.1</td>\n    </tr>\n    <tr>\n      <th>2</th>\n      <td>chr1</td>\n      <td>refGene</td>\n      <td>exon</td>\n      <td>12613</td>\n      <td>12721</td>\n      <td>NaN</td>\n      <td>+</td>\n      <td>0</td>\n      <td>LOC102725121</td>\n      <td>NR_148357</td>\n      <td>LOC102725121</td>\n      <td>2</td>\n      <td>NR_148357.2</td>\n    </tr>\n    <tr>\n      <th>3</th>\n      <td>chr1</td>\n      <td>refGene</td>\n      <td>exon</td>\n      <td>13221</td>\n      <td>14362</td>\n      <td>NaN</td>\n      <td>+</td>\n      <td>0</td>\n      <td>LOC102725121</td>\n      <td>NR_148357</td>\n      <td>LOC102725121</td>\n      <td>3</td>\n      <td>NR_148357.3</td>\n    </tr>\n    <tr>\n      <th>5</th>\n      <td>chr1</td>\n      <td>refGene</td>\n      <td>exon</td>\n      <td>11874</td>\n      <td>12227</td>\n      <td>NaN</td>\n      <td>+</td>\n      <td>0</td>\n      <td>DDX11L1</td>\n      <td>NR_046018</td>\n      <td>DDX11L1</td>\n      <td>1</td>\n      <td>NR_046018.1</td>\n    </tr>\n    <tr>\n      <th>6</th>\n      <td>chr1</td>\n      <td>refGene</td>\n      <td>exon</td>\n      <td>12613</td>\n      <td>12721</td>\n      <td>NaN</td>\n      <td>+</td>\n      <td>0</td>\n      <td>DDX11L1</td>\n      <td>NR_046018</td>\n      <td>DDX11L1</td>\n      <td>2</td>\n      <td>NR_046018.2</td>\n    </tr>\n  </tbody>\n</table>\n</div>"
     },
     "execution_count": 257,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "refGene_df = refGene_df[refGene_df['feature'] == 'exon']\n",
    "refGene_df.head()"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2024-05-11T06:20:03.836331Z",
     "start_time": "2024-05-11T06:20:03.692797800Z"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 272,
   "outputs": [
    {
     "data": {
      "text/plain": "  #chrom   start     end  gene_id  symbol exon_number    transcript\n0   chr1   69091   70008    79501   OR4F5           1  NM_001005484\n1   chr1  367659  368597   729759  OR4F29           1  NM_001005221\n2   chr1  367659  368597    81399  OR4F16           1  NM_001005277\n3   chr1  879583  880180    26155   NOC2L           1     NM_015658\n4   chr1  880437  880526    26155   NOC2L           2     NM_015658",
      "text/html": "<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>#chrom</th>\n      <th>start</th>\n      <th>end</th>\n      <th>gene_id</th>\n      <th>symbol</th>\n      <th>exon_number</th>\n      <th>transcript</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>0</th>\n      <td>chr1</td>\n      <td>69091</td>\n      <td>70008</td>\n      <td>79501</td>\n      <td>OR4F5</td>\n      <td>1</td>\n      <td>NM_001005484</td>\n    </tr>\n    <tr>\n      <th>1</th>\n      <td>chr1</td>\n      <td>367659</td>\n      <td>368597</td>\n      <td>729759</td>\n      <td>OR4F29</td>\n      <td>1</td>\n      <td>NM_001005221</td>\n    </tr>\n    <tr>\n      <th>2</th>\n      <td>chr1</td>\n      <td>367659</td>\n      <td>368597</td>\n      <td>81399</td>\n      <td>OR4F16</td>\n      <td>1</td>\n      <td>NM_001005277</td>\n    </tr>\n    <tr>\n      <th>3</th>\n      <td>chr1</td>\n      <td>879583</td>\n      <td>880180</td>\n      <td>26155</td>\n      <td>NOC2L</td>\n      <td>1</td>\n      <td>NM_015658</td>\n    </tr>\n    <tr>\n      <th>4</th>\n      <td>chr1</td>\n      <td>880437</td>\n      <td>880526</td>\n      <td>26155</td>\n      <td>NOC2L</td>\n      <td>2</td>\n      <td>NM_015658</td>\n    </tr>\n  </tbody>\n</table>\n</div>"
     },
     "execution_count": 272,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_db = pd.read_csv(settings.GENE_DATABASE, sep='\\t')\n",
    "merge_df = gene_db.merge(refGene_df, left_on='transcript', right_on='transcript_id')[[\n",
    "    '#chrom', 'start_y', 'end_y', 'gene_id_x', 'symbol', 'exon_number', 'transcript'\n",
    "]]\n",
    "merge_df.columns = [x.strip('_x').strip('_y') for x in merge_df.columns]\n",
    "merge_df = merge_df.sort_values(['#chrom', 'start', 'end'])\n",
    "merge_df.head()"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2024-05-11T06:50:44.968596Z",
     "start_time": "2024-05-11T06:50:44.413737700Z"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 273,
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/staging/software/lib/python3.9/site-packages/autocnv-1-py3.9.egg/data/exon.sorted.bed True\n"
     ]
    }
   ],
   "source": [
    "merge_df.to_csv(settings.GENE_EXON_DATABASE.strip('.gz'), sep='\\t', index=False)\n",
    "print(settings.GENE_EXON_DATABASE.strip('.gz'), os.path.exists(settings.GENE_EXON_DATABASE.strip('.gz')))"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2024-05-11T06:50:50.349488900Z",
     "start_time": "2024-05-11T06:50:49.767532300Z"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 274,
   "outputs": [],
   "source": [
    "!bgzip -cf {settings.GENE_EXON_DATABASE.strip('.gz')} > {settings.GENE_EXON_DATABASE}"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    },
    "ExecuteTime": {
     "end_time": "2024-05-11T06:50:54.011031Z",
     "start_time": "2024-05-11T06:50:52.772345200Z"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 275,
   "outputs": [],
   "source": [
    "!tabix -fp bed {settings.GENE_EXON_DATABASE}\n",
    "!rm {settings.GENE_EXON_DATABASE.strip('.gz')}"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2024-05-11T06:50:59.122931100Z",
     "start_time": "2024-05-11T06:50:57.477319800Z"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "# cnv_syndrome_loss cnv_syndrome_gain"
   ],
   "metadata": {
    "collapsed": false
   }
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "a51fe8dc5e4b3f3bbe9cb07bc9df4f2d9831135c008ea9f53caf8f1c6eba4359"
  },
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
